// cellar-door-entry/dist/index.js
// src/types.ts
var ENTRY_CONTEXT_V1 = "https://cellar-door.dev/entry/v1";

// src/verify-departure.ts
import { verifyMarker, fromJSON } from "cellar-door-exit";
function verifyDeparture(exitMarker) {
  const result = verifyMarker(exitMarker);
  return { valid: result.valid, errors: [...result.errors] };
}
function verifyDepartureJSON(exitMarkerJson) {
  try {
    const marker = fromJSON(exitMarkerJson);
    const result = verifyDeparture(marker);
    return { marker, result, parsed: true };
  } catch (e) {
    return {
      marker: null,
      result: { valid: false, errors: [e.message] },
      parsed: false
    };
  }
}

// src/arrival.ts
import { sha256 } from "@noble/hashes/sha256";
import _canonicalize from "canonicalize";
function canonicalize(obj) {
  const result = _canonicalize(obj);
  if (result === void 0) return "null";
  return result;
}
function computeArrivalId(marker) {
  const canonical = canonicalize(marker);
  const hash = sha256(new TextEncoder().encode(canonical));
  return Array.from(hash).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function createArrivalMarker(exitMarker, destination, opts) {
  const exit = JSON.parse(JSON.stringify(exitMarker));
  const verificationResult = verifyDeparture(exit);
  const admissionType = opts?.admissionType ?? (verificationResult.valid ? "automatic" : "reviewed");
  const timestamp = opts?.timestamp ?? (/* @__PURE__ */ new Date()).toISOString();
  const body = {
    "@context": ENTRY_CONTEXT_V1,
    type: "ArrivalMarker",
    departureRef: exit.id,
    departureOrigin: exit.origin,
    destination,
    subject: exit.subject,
    timestamp,
    admissionType,
    ...opts?.conditions ? { conditions: opts.conditions } : {},
    verificationResult,
    ...opts?.probation ? { probation: opts.probation } : {},
    ...opts?.capabilityScope ? { capabilityScope: opts.capabilityScope } : {}
  };
  const id = `urn:entry:${computeArrivalId(body)}`;
  return { ...body, id };
}

// src/sign.ts
import {
  sign,
  verify,
  didFromPublicKey,
  publicKeyFromDid,
  signP256,
  verifyP256,
  didFromP256PublicKey,
  publicKeyFromP256Did,
  algorithmFromDid
} from "cellar-door-exit";
var DOMAIN_PREFIX = "entry-marker-v1.0:";
function signArrivalMarker(marker, privateKey, publicKey, algorithm = "Ed25519") {
  const { proof: _proof, ...rest } = marker;
  const canonical = canonicalize(rest);
  const data = new TextEncoder().encode(DOMAIN_PREFIX + canonical);
  let did;
  let signature;
  let proofType;
  if (algorithm === "P-256") {
    did = didFromP256PublicKey(publicKey);
    const sig = signP256(data, privateKey);
    signature = new Uint8Array(sig);
    proofType = "EcdsaP256Signature2019";
  } else {
    did = didFromPublicKey(publicKey);
    signature = sign(data, privateKey);
    proofType = "Ed25519Signature2020";
  }
  const proofValue = btoa(String.fromCharCode(...signature));
  const proof = {
    type: proofType,
    created: (/* @__PURE__ */ new Date()).toISOString(),
    verificationMethod: did,
    proofValue
  };
  return { ...marker, proof };
}
function verifyArrivalMarker(marker) {
  const errors = [];
  if (!marker.proof) {
    return { valid: false, errors: ["Missing proof"] };
  }
  const supportedTypes = ["Ed25519Signature2020", "EcdsaP256Signature2019"];
  if (!supportedTypes.includes(marker.proof.type)) {
    return { valid: false, errors: [`Unsupported proof type: ${marker.proof.type}`] };
  }
  if (!marker.proof.verificationMethod || !marker.proof.proofValue) {
    return { valid: false, errors: ["Incomplete proof"] };
  }
  try {
    const did = marker.proof.verificationMethod;
    const alg = algorithmFromDid(did);
    const publicKey = alg === "P-256" ? publicKeyFromP256Did(did) : publicKeyFromDid(did);
    const { proof: _proof, ...rest } = marker;
    const canonical = canonicalize(rest);
    const data = new TextEncoder().encode(DOMAIN_PREFIX + canonical);
    const sigStr = atob(marker.proof.proofValue);
    const signature = new Uint8Array(sigStr.length);
    for (let i = 0; i < sigStr.length; i++) signature[i] = sigStr.charCodeAt(i);
    const valid = alg === "P-256" ? verifyP256(data, signature, publicKey) : verify(data, signature, publicKey);
    if (!valid) {
      errors.push("Signature verification failed");
    }
  } catch (e) {
    errors.push(`Verification error: ${e.message}`);
  }
  return { valid: errors.length === 0, errors };
}

// src/continuity.ts
import { verifyMarker as verifyMarker2 } from "cellar-door-exit";
function verifyContinuity(exitMarker, arrivalMarker) {
  const errors = [];
  if (arrivalMarker.departureRef !== exitMarker.id) {
    errors.push(
      `Departure reference mismatch: arrival references "${arrivalMarker.departureRef}" but exit marker has id "${exitMarker.id}"`
    );
  }
  if (arrivalMarker.subject !== exitMarker.subject) {
    errors.push(
      `Subject mismatch: arrival subject "${arrivalMarker.subject}" does not match departure subject "${exitMarker.subject}"`
    );
  }
  if (arrivalMarker.departureOrigin !== exitMarker.origin) {
    errors.push(
      `Origin mismatch: arrival claims origin "${arrivalMarker.departureOrigin}" but exit marker has origin "${exitMarker.origin}"`
    );
  }
  const exitTime = new Date(exitMarker.timestamp).getTime();
  const arrivalTime = new Date(arrivalMarker.timestamp).getTime();
  if (!isNaN(exitTime) && !isNaN(arrivalTime) && arrivalTime < exitTime) {
    errors.push(
      `Temporal violation: arrival timestamp (${arrivalMarker.timestamp}) is before departure timestamp (${exitMarker.timestamp})`
    );
  }
  const exitVerification = verifyMarker2(exitMarker);
  if (!exitVerification.valid) {
    errors.push(`Departure marker is invalid: ${exitVerification.errors.join("; ")}`);
  }
  return { valid: errors.length === 0, errors };
}

// src/convenience.ts
import {
  fromJSON as fromJSON2,
  generateKeyPair,
  generateP256KeyPair
} from "cellar-door-exit";

// src/admit.ts
import { createHash, randomUUID } from "crypto";
import {
  addCounterSignature,
  deriveStatusConfirmation as deriveStatusConfirmation3,
  verifyCounterSignature,
  publicKeyFromDid as publicKeyFromDid2,
  publicKeyFromP256Did as publicKeyFromP256Did2,
  algorithmFromDid as algorithmFromDid2
} from "cellar-door-exit";

// src/policy-builder.ts
import { StatusConfirmation as StatusConfirmation3 } from "cellar-door-exit";

// src/conflict.ts
function resolveConflict(results, strategy = "deny-overrides", defaultDecision = "deny") {
  const reasons = [];
  const conditions = [];
  for (const r of results) {
    if (r.conditions) conditions.push(...r.conditions);
  }
  const decisive = results.filter((r) => r.decision !== "abstain");
  if (decisive.length === 0) {
    const admitted = defaultDecision === "admit";
    return {
      admitted,
      reasons: [admitted ? "No rules produced a decision \u2014 default admit" : "No rules produced a decision \u2014 default deny"],
      conditions,
      ruleResults: results
    };
  }
  switch (strategy) {
    case "deny-overrides": {
      const denied = decisive.filter((r) => r.decision === "deny");
      if (denied.length > 0) {
        return {
          admitted: false,
          reasons: denied.map((r) => `[${r.name}] ${r.reason}`),
          conditions,
          ruleResults: results
        };
      }
      const admitted = decisive.filter((r) => r.decision === "admit");
      return {
        admitted: true,
        reasons: admitted.map((r) => `[${r.name}] ${r.reason}`),
        conditions,
        ruleResults: results
      };
    }
    case "first-match": {
      const first = decisive[0];
      return {
        admitted: first.decision === "admit",
        reasons: [`[${first.name}] ${first.reason}`],
        conditions: first.conditions ?? [],
        ruleResults: results
      };
    }
    case "permit-overrides": {
      const admitted = decisive.filter((r) => r.decision === "admit");
      if (admitted.length > 0) {
        return {
          admitted: true,
          reasons: admitted.map((r) => `[${r.name}] ${r.reason}`),
          conditions,
          ruleResults: results
        };
      }
      const denied = decisive.filter((r) => r.decision === "deny");
      return {
        admitted: false,
        reasons: denied.map((r) => `[${r.name}] ${r.reason}`),
        conditions,
        ruleResults: results
      };
    }
  }
}

// src/admission-policy.ts
function parseDuration(dur) {
  if (typeof dur === "number") return dur;
  const match = dur.match(/^(\d+)(ms|s|m|h|d)$/);
  if (!match) throw new Error(`Invalid duration: ${dur}`);
  const val = parseInt(match[1], 10);
  const unit = match[2];
  const multipliers = { ms: 1, s: 1e3, m: 6e4, h: 36e5, d: 864e5 };
  return val * multipliers[unit];
}
var MODULE_KEYS = ["lineage", "stateSnapshot", "dispute", "economic", "metadata", "crossDomain"];
function evaluateAdmission(exitMarker, policy, now) {
  if ("evaluate" in policy && typeof policy.evaluate === "function") {
    const result = policy.evaluate(exitMarker, { now: now ?? /* @__PURE__ */ new Date() });
    return {
      admitted: result.admitted,
      conditions: result.conditions,
      reasons: result.reasons
    };
  }
  return evaluateFlat(exitMarker, policy, now);
}
function evaluateFlat(exitMarker, policy, now) {
  const reasons = [];
  const conditions = [];
  const currentTime = (now ?? /* @__PURE__ */ new Date()).getTime();
  if (policy.allowedExitTypes && policy.allowedExitTypes.length > 0) {
    if (!policy.allowedExitTypes.includes(exitMarker.exitType)) {
      reasons.push(`Exit type "${exitMarker.exitType}" is not in allowed types: ${policy.allowedExitTypes.join(", ")}`);
    }
  }
  if (policy.maxDepartureAge != null) {
    const departureTime = new Date(exitMarker.timestamp).getTime();
    const age = currentTime - departureTime;
    if (age > policy.maxDepartureAge) {
      reasons.push(`Departure is ${age}ms old, exceeds maximum of ${policy.maxDepartureAge}ms`);
    }
  }
  if (policy.requiredModules && policy.requiredModules.length > 0) {
    for (const mod of policy.requiredModules) {
      const key = mod;
      if (MODULE_KEYS.includes(key) && !exitMarker[key]) {
        reasons.push(`Required module "${mod}" is missing`);
      }
    }
  }
  if (policy.requireVerifiedDeparture) {
    if (!exitMarker.proof || !exitMarker.proof.proofValue) {
      reasons.push("Departure marker has no proof/signature");
    }
    conditions.push("departure-verified");
  }
  return {
    admitted: reasons.length === 0,
    conditions,
    reasons
  };
}
var OPEN_DOOR = {
  requireVerifiedDeparture: true
};
var STRICT = {
  requireVerifiedDeparture: true,
  maxDepartureAge: 864e5,
  // 24h
  allowedExitTypes: ["voluntary"],
  requiredModules: ["lineage", "stateSnapshot"]
};
var EMERGENCY_ONLY = {
  requireVerifiedDeparture: true,
  allowedExitTypes: ["emergency"]
};

// src/rules/confirmation-rule.ts
import { StatusConfirmation, deriveStatusConfirmation } from "cellar-door-exit";
var LEVEL_ORDER = [
  StatusConfirmation.SelfOnly,
  StatusConfirmation.OriginOnly,
  StatusConfirmation.Mutual,
  StatusConfirmation.Witnessed
];
function levelIndex(level) {
  const idx = LEVEL_ORDER.indexOf(level);
  return idx >= 0 ? idx : 0;
}
function createConfirmationRule(config) {
  return {
    name: "confirmation-level",
    evaluate(marker, ctx) {
      if (ctx.isMinting || !marker) {
        return { name: "confirmation-level", decision: "abstain", reason: "No marker to check confirmation" };
      }
      const actual = deriveStatusConfirmation(marker);
      const actualIdx = levelIndex(actual);
      const requiredIdx = levelIndex(config.requiredLevel);
      if (actualIdx >= requiredIdx) {
        return { name: "confirmation-level", decision: "admit", reason: `Confirmation level ${actual} meets minimum ${config.requiredLevel}` };
      }
      return { name: "confirmation-level", decision: "deny", reason: `Confirmation level ${actual} below required ${config.requiredLevel}` };
    },
    toJSON() {
      return { type: "confirmation-level", requiredLevel: config.requiredLevel };
    }
  };
}

// src/rules/origin-rule.ts
function createOriginRule(config) {
  return {
    name: "origin-check",
    evaluate(marker, ctx) {
      if (ctx.isMinting || !marker) {
        return { name: "origin-check", decision: "abstain", reason: "No marker to check origin" };
      }
      const origin = marker.origin;
      if (config.blockedOrigins) {
        const now = ctx.now.getTime();
        const blocked = config.blockedOrigins.find((b) => {
          if (b.origin !== origin) return false;
          if (b.expiresAt && new Date(b.expiresAt).getTime() <= now) return false;
          return true;
        });
        if (blocked) {
          const reasonStr = typeof blocked.reason === "string" ? blocked.reason : `[${blocked.reason.category}] ${blocked.reason.detail}`;
          return { name: "origin-check", decision: "deny", reason: `Origin "${origin}" is blocked: ${reasonStr}` };
        }
      }
      if (config.allowedOrigins && config.allowedOrigins.length > 0) {
        if (!config.allowedOrigins.includes(origin)) {
          return { name: "origin-check", decision: "deny", reason: `Origin "${origin}" not in allowed list` };
        }
      }
      return { name: "origin-check", decision: "admit", reason: `Origin "${origin}" is acceptable` };
    },
    toJSON() {
      return { type: "origin-check", ...config };
    }
  };
}

// src/rules/condition.ts
function formatCondition(c) {
  if (!c.params || Object.keys(c.params).length === 0) return c.type;
  return Object.entries(c.params).map(([k, v]) => `${c.type}-${k}:${v}`).join(",");
}
function parseCondition(s) {
  const match = s.match(/^([a-z-]+?)-([a-z]+):(.+)$/);
  if (match) {
    return { type: match[1], params: { [match[2]]: match[3] } };
  }
  return { type: s };
}

// src/rules/dispute-rule.ts
function isDisputed(marker) {
  return !!(marker.dispute?.disputes && marker.dispute.disputes.length > 0);
}
function createDisputeRule(config) {
  return {
    name: "dispute-check",
    evaluate(marker, ctx) {
      if (ctx.isMinting || !marker) {
        return { name: "dispute-check", decision: "abstain", reason: "No marker to check disputes" };
      }
      if (!isDisputed(marker)) {
        return { name: "dispute-check", decision: "abstain", reason: "Marker is not disputed" };
      }
      switch (config.action) {
        case "reject":
          return { name: "dispute-check", decision: "deny", reason: "Disputed marker rejected by policy" };
        case "quarantine":
          return {
            name: "dispute-check",
            decision: "admit",
            reason: "Disputed marker quarantined for review",
            conditions: [
              formatCondition({ type: "quarantine" }),
              ...config.maxDuration ? [formatCondition({ type: "quarantine", params: { max: config.maxDuration } })] : [],
              ...config.reviewRequired ? [formatCondition({ type: "review" })] : []
            ]
          };
        case "manual_review":
          return {
            name: "dispute-check",
            decision: "admit",
            reason: "Disputed marker admitted pending manual review",
            conditions: ["manual-review-required"]
          };
      }
    },
    toJSON() {
      return { type: "dispute-check", ...config };
    }
  };
}

// src/rules/trust-level-rule.ts
import { StatusConfirmation as StatusConfirmation2, deriveStatusConfirmation as deriveStatusConfirmation2 } from "cellar-door-exit";
function createTrustLevelRule(config) {
  return {
    name: "trust-level",
    evaluate(marker, ctx) {
      if (ctx.isMinting || !marker) {
        return { name: "trust-level", decision: "abstain", reason: "No marker to check trust level" };
      }
      const level = deriveStatusConfirmation2(marker);
      if (level !== StatusConfirmation2.SelfOnly) {
        return { name: "trust-level", decision: "abstain", reason: `Trust level ${level} is not self_only` };
      }
      switch (config.action) {
        case "reject":
          return { name: "trust-level", decision: "deny", reason: "Self-only markers rejected by policy" };
        case "probation":
          return {
            name: "trust-level",
            decision: "admit",
            reason: "Self-only marker admitted with probation",
            conditions: [
              formatCondition({ type: "probation" }),
              ...config.probationDuration ? [formatCondition({ type: "probation", params: { duration: config.probationDuration } })] : []
            ]
          };
        case "admit":
          return { name: "trust-level", decision: "admit", reason: "Self-only marker admitted" };
      }
    },
    toJSON() {
      return { type: "trust-level", ...config };
    }
  };
}

// src/rules/minting-rule.ts
function createMintingRule(config) {
  return {
    name: "minting",
    evaluate(_marker, ctx) {
      if (!ctx.isMinting) {
        return { name: "minting", decision: "abstain", reason: "Not a minting operation" };
      }
      if (!config.allowed) {
        return { name: "minting", decision: "deny", reason: "Minting is not allowed by policy" };
      }
      if (config.requireJustification && !ctx.mintingJustification) {
        return { name: "minting", decision: "deny", reason: "Minting requires a justification" };
      }
      const conditions = [];
      if (config.enhancedProbation) {
        conditions.push("probation", `probation-duration:${config.enhancedProbation}`);
      }
      return {
        name: "minting",
        decision: "admit",
        reason: "Minting allowed by policy",
        conditions
      };
    },
    toJSON() {
      return { type: "minting", ...config };
    }
  };
}

// src/rules/custom-rule.ts
function createCustomRule(name, fn) {
  return {
    name,
    evaluate(marker, ctx) {
      const result = fn(marker, ctx);
      return {
        name,
        decision: result.admitted ? "admit" : "deny",
        reason: result.reason ?? (result.admitted ? "Custom rule passed" : "Custom rule failed"),
        conditions: result.conditions
      };
    }
    // No toJSON — custom rules are not serializable
  };
}

// src/policy-builder.ts
function buildContext(partial) {
  return {
    isMinting: partial?.isMinting ?? false,
    now: partial?.now ?? /* @__PURE__ */ new Date(),
    mintingJustification: partial?.mintingJustification
  };
}
var PolicyBuilder = class _PolicyBuilder {
  _name;
  _rules = [];
  _strategy = "deny-overrides";
  _defaultDecision = "deny";
  // Track serializable config for reconstruction
  _ruleConfigs = [];
  _warnings = [];
  constructor(name) {
    this._name = name;
  }
  requireConfirmation(level) {
    const resolved = typeof level === "string" ? this._resolveConfirmation(level) : level;
    this._rules.push(createConfirmationRule({ requiredLevel: resolved }));
    this._ruleConfigs.push({ type: "confirmation-level", config: { requiredLevel: resolved } });
    return this;
  }
  _resolveConfirmation(level) {
    const map = {
      self_only: StatusConfirmation3.SelfOnly,
      origin_only: StatusConfirmation3.OriginOnly,
      mutual: StatusConfirmation3.Mutual,
      witnessed: StatusConfirmation3.Witnessed
    };
    return map[level] ?? StatusConfirmation3.SelfOnly;
  }
  requireVerifiedDeparture() {
    this._rules.push({
      name: "verified-departure",
      evaluate(marker, ctx) {
        if (ctx.isMinting || !marker) {
          return { name: "verified-departure", decision: "abstain", reason: "No marker to verify" };
        }
        if (!marker.proof?.proofValue) {
          return { name: "verified-departure", decision: "deny", reason: "Departure marker has no proof" };
        }
        return { name: "verified-departure", decision: "admit", reason: "Departure is signed" };
      },
      toJSON() {
        return { type: "verified-departure" };
      }
    });
    this._ruleConfigs.push({ type: "verified-departure", config: {} });
    return this;
  }
  maxAge(duration) {
    const maxMs = typeof duration === "string" ? parseDuration(duration) : duration;
    const durStr = typeof duration === "string" ? duration : `${duration}ms`;
    this._rules.push({
      name: "max-age",
      evaluate(marker, ctx) {
        if (ctx.isMinting || !marker) {
          return { name: "max-age", decision: "abstain", reason: "No marker to check age" };
        }
        const age = ctx.now.getTime() - new Date(marker.timestamp).getTime();
        if (age > maxMs) {
          return { name: "max-age", decision: "deny", reason: `Marker is ${age}ms old, exceeds ${durStr}` };
        }
        return { name: "max-age", decision: "admit", reason: `Marker age ${age}ms within ${durStr}` };
      },
      toJSON() {
        return { type: "max-age", duration: durStr };
      }
    });
    this._ruleConfigs.push({ type: "max-age", config: { duration: durStr } });
    return this;
  }
  allowExitTypes(types) {
    this._rules.push({
      name: "exit-type",
      evaluate(marker, ctx) {
        if (ctx.isMinting || !marker) {
          return { name: "exit-type", decision: "abstain", reason: "No marker to check exit type" };
        }
        if (types.includes(marker.exitType)) {
          return { name: "exit-type", decision: "admit", reason: `Exit type "${marker.exitType}" is allowed` };
        }
        return { name: "exit-type", decision: "deny", reason: `Exit type "${marker.exitType}" not in ${types.join(", ")}` };
      },
      toJSON() {
        return { type: "exit-type", allowedTypes: types };
      }
    });
    this._ruleConfigs.push({ type: "exit-type", config: { allowedTypes: types } });
    return this;
  }
  blockOrigins(origins, opts) {
    this._rules.push(createOriginRule({
      blockedOrigins: origins.map((o) => ({ origin: o, reason: opts.reason }))
    }));
    this._ruleConfigs.push({ type: "origin-check", config: { blockedOrigins: origins.map((o) => ({ origin: o, reason: opts.reason })) } });
    return this;
  }
  allowOrigins(origins) {
    this._rules.push(createOriginRule({ allowedOrigins: origins }));
    this._ruleConfigs.push({ type: "origin-check", config: { allowedOrigins: origins } });
    return this;
  }
  requireModules(modules) {
    this._rules.push({
      name: "required-modules",
      evaluate(marker, ctx) {
        if (ctx.isMinting || !marker) {
          return { name: "required-modules", decision: "abstain", reason: "No marker to check modules" };
        }
        const missing = [];
        for (const mod of modules) {
          if (!marker[mod]) missing.push(mod);
        }
        if (missing.length > 0) {
          return { name: "required-modules", decision: "deny", reason: `Missing modules: ${missing.join(", ")}` };
        }
        return { name: "required-modules", decision: "admit", reason: "All required modules present" };
      },
      toJSON() {
        return { type: "required-modules", modules };
      }
    });
    this._ruleConfigs.push({ type: "required-modules", config: { modules } });
    return this;
  }
  onDisputed(action, opts) {
    this._rules.push(createDisputeRule({ action, maxDuration: opts?.maxDuration, reviewRequired: opts?.reviewRequired }));
    this._ruleConfigs.push({ type: "dispute-check", config: { action, ...opts } });
    return this;
  }
  onSelfOnly(action, opts) {
    this._rules.push(createTrustLevelRule({ action, probationDuration: opts?.duration }));
    this._ruleConfigs.push({ type: "trust-level", config: { action, probationDuration: opts?.duration } });
    return this;
  }
  allowMinting(opts) {
    this._rules.push(createMintingRule({
      allowed: true,
      requireJustification: opts?.requireJustification,
      enhancedProbation: opts?.enhancedProbation
    }));
    this._ruleConfigs.push({ type: "minting", config: { allowed: true, ...opts } });
    return this;
  }
  custom(name, fn) {
    this._rules.push(createCustomRule(name, fn));
    this._ruleConfigs.push(null);
    return this;
  }
  conflictResolution(strategy) {
    this._strategy = strategy;
    return this;
  }
  /**
   * Set the default decision when no rules produce a decisive result.
   * Defaults to 'deny' (fail-closed). Use 'admit' for explicit open-door policies.
   */
  defaultDecision(decision) {
    this._defaultDecision = decision;
    return this;
  }
  build() {
    const rules = [...this._rules];
    const strategy = this._strategy;
    const name = this._name;
    const defaultDec = this._defaultDecision;
    const ruleConfigs = [...this._ruleConfigs];
    const warnings = [...this._warnings];
    function evaluate(marker, ctxPartial) {
      const ctx = buildContext(ctxPartial);
      const results = rules.map((r) => r.evaluate(marker, ctx));
      return resolveConflict(results, strategy, defaultDec);
    }
    function dryRun(marker, ctxPartial) {
      const ctx = buildContext(ctxPartial);
      const results = rules.map((r) => r.evaluate(marker, ctx));
      const decision = resolveConflict(results, strategy, defaultDec);
      return { policyName: name, rules: results, decision, timestamp: (/* @__PURE__ */ new Date()).toISOString() };
    }
    function toJSON() {
      const serializable = [];
      const customOmitted = [];
      for (let i = 0; i < rules.length; i++) {
        const cfg = ruleConfigs[i];
        if (cfg) {
          serializable.push(cfg.config ? { ...cfg.config, type: cfg.type } : { type: cfg.type });
        } else {
          customOmitted.push(rules[i].name);
        }
      }
      return { name, conflictStrategy: strategy, rules: serializable, customRulesOmitted: customOmitted };
    }
    return { name, rules, conflictStrategy: strategy, defaultDecision: defaultDec, _warnings: warnings.length > 0 ? warnings : void 0, evaluate, dryRun, toJSON };
  }
  /**
   * Reconstruct a PolicyBuilder from JSON.
   * Custom rules are NOT restored (they are not serializable).
   */
  static fromJSON(json) {
    const builder = new _PolicyBuilder(json.name);
    builder._strategy = json.conflictStrategy;
    for (const rule of json.rules) {
      switch (rule.type) {
        case "confirmation-level":
          builder.requireConfirmation(rule.requiredLevel);
          break;
        case "verified-departure":
          builder.requireVerifiedDeparture();
          break;
        case "max-age":
          builder.maxAge(rule.duration);
          break;
        case "exit-type":
          builder.allowExitTypes(rule.allowedTypes);
          break;
        case "origin-check":
          if (rule.blockedOrigins) {
            const blocked = rule.blockedOrigins;
            builder.blockOrigins(blocked.map((b) => b.origin), { reason: blocked[0]?.reason ?? "Blocked" });
          }
          if (rule.allowedOrigins) {
            builder.allowOrigins(rule.allowedOrigins);
          }
          break;
        case "required-modules":
          builder.requireModules(rule.modules);
          break;
        case "dispute-check":
          builder.onDisputed(rule.action, { maxDuration: rule.maxDuration, reviewRequired: rule.reviewRequired });
          break;
        case "trust-level":
          builder.onSelfOnly(rule.action, { duration: rule.probationDuration });
          break;
        case "minting":
          builder.allowMinting({ requireJustification: rule.requireJustification, enhancedProbation: rule.enhancedProbation });
          break;
        default:
          builder._warnings.push(`Unknown rule type: ${rule.type}`);
          break;
      }
    }
    return builder;
  }
};
function createPolicy(name) {
  return new PolicyBuilder(name);
}
var REQUIRE_MUTUAL = createPolicy("require-mutual").requireConfirmation("mutual").requireVerifiedDeparture().build();
var REQUIRE_MUTUAL_WITH_ONRAMP = createPolicy("require-mutual-with-onramp").requireConfirmation("mutual").requireVerifiedDeparture().allowMinting({ requireJustification: true, enhancedProbation: "14d" }).build();
var CAUTIOUS = createPolicy("cautious").requireVerifiedDeparture().onSelfOnly("probation", { duration: "7d" }).onDisputed("quarantine", { maxDuration: "7d", reviewRequired: true }).allowMinting({ requireJustification: true, enhancedProbation: "14d" }).build();
var OPEN_DOOR_V2 = createPolicy("open-door").defaultDecision("admit").build();
var QUARANTINE_UNKNOWN = createPolicy("quarantine-unknown").requireVerifiedDeparture().onDisputed("quarantine", { maxDuration: "7d", reviewRequired: true }).build();
var PERMISSIVE = createPolicy("permissive").defaultDecision("admit").onSelfOnly("probation", { duration: "7d" }).allowMinting().build();
var LOCKDOWN = createPolicy("lockdown").build();

// src/admit.ts
function isAdmissionStore(store) {
  return typeof store.putAdmission === "function";
}
function generateRecordId() {
  return `admission-${randomUUID()}`;
}
async function persistRejection(record, store) {
  if (store && isAdmissionStore(store)) {
    await store.putAdmission(record);
  }
}
function computePolicySnapshot(policy) {
  if (!policy.toJSON) return void 0;
  try {
    return JSON.stringify(policy.toJSON());
  } catch {
    return void 0;
  }
}
function computePolicyVersion(policy) {
  try {
    const json = JSON.stringify(policy.toJSON());
    return createHash("sha256").update(json).digest("hex").slice(0, 16);
  } catch {
    return "unknown";
  }
}
async function admit(exitMarkerInput, opts = {}) {
  let exitMarker = exitMarkerInput ? JSON.parse(JSON.stringify(exitMarkerInput)) : exitMarkerInput;
  const policy = opts.policy ?? CAUTIOUS;
  const destination = opts.destination ?? "https://destination.example.com";
  const timestamp = opts.timestamp ?? (/* @__PURE__ */ new Date()).toISOString();
  const isMinting = !exitMarker;
  if (isMinting && !opts.mintingSubject) {
    throw new Error("mintingSubject is required when minting (no exit marker)");
  }
  const pendingClaimId = `pending-${randomUUID()}`;
  if (opts.store && exitMarker) {
    const subjectDid = exitMarker.subject ?? "unknown";
    const claimed = await opts.store.claim(exitMarker.id, pendingClaimId, subjectDid);
    if (!claimed) {
      const record2 = {
        id: generateRecordId(),
        exitMarkerId: exitMarker.id,
        arrivalMarkerId: "",
        subjectDid: exitMarker.subject ?? "unknown",
        originDid: exitMarker.origin,
        policyApplied: policy.name,
        policyVersion: computePolicyVersion(policy),
        admitted: false,
        admissionType: "standard",
        counterSigned: false,
        reasonCodes: ["replay-detected"],
        conditions: [],
        timestamp,
        policySnapshot: computePolicySnapshot(policy)
      };
      await persistRejection(record2, opts.store);
      opts.emitter?.emit("agent:rejected", { record: record2, exitMarkerId: exitMarker.id });
      return {
        counterSignedExitMarker: exitMarker,
        arrivalMarker: null,
        admission: record2
      };
    }
  }
  let verified = true;
  if (exitMarker) {
    const verificationResult = verifyDeparture(exitMarker);
    verified = verificationResult.valid;
    if (!verified && exitMarker.dispute?.counterpartyAcks?.length) {
      const strippedDispute = { ...exitMarker.dispute };
      delete strippedDispute.counterpartyAcks;
      const hasOtherDispute = Object.keys(strippedDispute).length > 0;
      const stripped = hasOtherDispute ? { ...exitMarker, dispute: strippedDispute } : (() => {
        const { dispute: _, ...rest } = exitMarker;
        return rest;
      })();
      const strippedResult = verifyDeparture(stripped);
      if (strippedResult.valid) verified = true;
    }
    if (exitMarker.dispute?.counterpartyAcks?.length) {
      const markerForVerify = exitMarker;
      const verifiedAcks = exitMarker.dispute.counterpartyAcks.filter((ack, i) => {
        try {
          const algo = algorithmFromDid2(ack.verificationMethod);
          const pubKey = algo === "P-256" ? publicKeyFromP256Did2(ack.verificationMethod) : publicKeyFromDid2(ack.verificationMethod);
          const result = verifyCounterSignature(markerForVerify, pubKey, i);
          return result.valid;
        } catch {
          return false;
        }
      });
      if (verifiedAcks.length !== exitMarker.dispute.counterpartyAcks.length) {
        const cleanedDispute = { ...exitMarker.dispute };
        if (verifiedAcks.length > 0) {
          cleanedDispute.counterpartyAcks = verifiedAcks;
        } else {
          delete cleanedDispute.counterpartyAcks;
        }
        const remainingKeys = Object.keys(cleanedDispute).filter((k) => cleanedDispute[k] !== void 0);
        if (remainingKeys.length === 0) {
          const { dispute: _, ...rest } = exitMarker;
          exitMarker = rest;
        } else {
          exitMarker = { ...exitMarker, dispute: cleanedDispute };
        }
      }
    }
    const policyRequiresVerification = policy.rules.some((r) => r.name === "verified-departure");
    if (!verified && policyRequiresVerification) {
      const record2 = {
        id: generateRecordId(),
        exitMarkerId: exitMarker.id,
        arrivalMarkerId: "",
        subjectDid: exitMarker.subject ?? "unknown",
        originDid: exitMarker.origin,
        policyApplied: policy.name,
        policyVersion: computePolicyVersion(policy),
        admitted: false,
        admissionType: "standard",
        counterSigned: false,
        reasonCodes: ["departure-verification-failed"],
        conditions: [],
        timestamp,
        policySnapshot: computePolicySnapshot(policy)
      };
      if (opts.store) await opts.store.revoke(pendingClaimId).catch((e) => {
        if (typeof console !== "undefined") console.warn("[cellar-door-entry] Failed to release pre-claim:", e);
      });
      await persistRejection(record2, opts.store);
      opts.emitter?.emit("agent:rejected", { record: record2, exitMarkerId: exitMarker.id });
      return {
        counterSignedExitMarker: exitMarker,
        arrivalMarker: null,
        admission: record2
      };
    }
  }
  const policyResult = policy.evaluate(exitMarker ?? null, {
    isMinting,
    now: new Date(timestamp),
    mintingJustification: opts.mintingJustification,
    verified
  });
  if (!policyResult.admitted) {
    const record2 = {
      id: generateRecordId(),
      exitMarkerId: exitMarker?.id,
      arrivalMarkerId: "",
      subjectDid: exitMarker?.subject ?? "unknown",
      originDid: exitMarker?.origin,
      confirmationLevel: exitMarker ? deriveStatusConfirmation3(exitMarker) : void 0,
      policyApplied: policy.name,
      policyVersion: computePolicyVersion(policy),
      admitted: false,
      admissionType: isMinting ? "minted" : "standard",
      counterSigned: false,
      reasonCodes: policyResult.reasons,
      conditions: policyResult.conditions,
      timestamp,
      policySnapshot: computePolicySnapshot(policy)
    };
    if (opts.store && exitMarker) await opts.store.revoke(pendingClaimId).catch((e) => {
      if (typeof console !== "undefined") console.warn("[cellar-door-entry] Failed to release pre-claim:", e);
    });
    await persistRejection(record2, opts.store);
    opts.emitter?.emit("agent:rejected", { record: record2, exitMarkerId: exitMarker?.id });
    return {
      counterSignedExitMarker: exitMarker ?? void 0,
      arrivalMarker: null,
      admission: record2
    };
  }
  const isQuarantined = policyResult.conditions.some((c) => c === "quarantine");
  let quarantineExpires;
  if (isQuarantined) {
    const maxDurCondition = policyResult.conditions.find((c) => c.startsWith("quarantine-max:"));
    if (maxDurCondition) {
      const durStr = maxDurCondition.split(":")[1];
      const ms = parseDuration(durStr);
      quarantineExpires = new Date(new Date(timestamp).getTime() + ms).toISOString();
    }
  }
  let counterSignedExit = exitMarker ?? void 0;
  let didCounterSign = false;
  if (exitMarker && opts.platformIdentity && opts.counterSign !== false) {
    counterSignedExit = addCounterSignature(
      exitMarker,
      opts.platformIdentity.privateKey,
      opts.platformIdentity.publicKey
    );
    didCounterSign = true;
  }
  let arrivalMarker;
  if (exitMarker) {
    arrivalMarker = createArrivalMarker(exitMarker, destination, {
      admissionType: isQuarantined ? "conditional" : "automatic",
      conditions: policyResult.conditions.length > 0 ? policyResult.conditions : void 0,
      timestamp
    });
  } else {
    arrivalMarker = {
      "@context": "https://cellar-door.dev/entry/v1",
      id: `urn:entry:minted-${randomUUID()}`,
      departureRef: "urn:entry:minted",
      departureOrigin: "urn:entry:minted",
      destination,
      subject: opts.mintingSubject,
      timestamp,
      admissionType: "reviewed",
      verificationResult: { valid: false, errors: ["minted-no-departure"] }
    };
  }
  if (opts.platformIdentity) {
    arrivalMarker = signArrivalMarker(
      arrivalMarker,
      opts.platformIdentity.privateKey,
      opts.platformIdentity.publicKey
    );
  }
  if (opts.store) {
    if (isMinting) {
      const subjectDid = arrivalMarker.subject;
      await opts.store.claim(arrivalMarker.id, arrivalMarker.id, subjectDid);
    }
  }
  const record = {
    id: generateRecordId(),
    exitMarkerId: exitMarker?.id,
    arrivalMarkerId: arrivalMarker.id,
    subjectDid: exitMarker?.subject ?? arrivalMarker.subject,
    originDid: exitMarker?.origin,
    confirmationLevel: exitMarker ? deriveStatusConfirmation3(exitMarker) : void 0,
    policyApplied: policy.name,
    policyVersion: computePolicyVersion(policy),
    admitted: true,
    admissionType: opts.admissionType ?? (isMinting ? "minted" : "standard"),
    counterSigned: didCounterSign,
    counterSignMeaning: didCounterSign ? opts.counterSignMeaning ?? "receipt_only" : void 0,
    reasonCodes: policyResult.reasons,
    conditions: policyResult.conditions,
    timestamp,
    quarantined: isQuarantined || void 0,
    quarantineExpires,
    mintingJustification: isMinting ? opts.mintingJustification : void 0,
    policySnapshot: computePolicySnapshot(policy)
  };
  if (opts.store && isAdmissionStore(opts.store)) {
    await opts.store.putAdmission(record);
  }
  if (opts.emitter) {
    if (isQuarantined) {
      opts.emitter.emit("agent:quarantined", { record });
    } else if (isMinting) {
      opts.emitter.emit("agent:minted", { record });
    } else {
      opts.emitter.emit("agent:admitted", { record });
    }
  }
  return {
    counterSignedExitMarker: counterSignedExit,
    arrivalMarker,
    admission: record
  };
}

// src/convenience.ts
function quickEntry(exitMarkerJson, destination, opts) {
  const algorithm = opts?.algorithm ?? "Ed25519";
  const exitMarker = fromJSON2(exitMarkerJson);
  const arrival = createArrivalMarker(exitMarker, destination, opts);
  if (typeof console !== "undefined" && console.warn) {
    console.warn(
      "[cellar-door-entry] quickEntry() uses an ephemeral keypair that is discarded after signing. The arrival marker cannot be revoked or updated later. For production, use createArrivalMarker() + signArrivalMarker() with a persistent key."
    );
  }
  let publicKey;
  let privateKey;
  if (algorithm === "P-256") {
    const kp = generateP256KeyPair();
    publicKey = kp.publicKey;
    privateKey = kp.privateKey;
  } else {
    const kp = generateKeyPair();
    publicKey = kp.publicKey;
    privateKey = kp.privateKey;
  }
  const signed = signArrivalMarker(arrival, privateKey, publicKey, algorithm);
  const continuity = verifyContinuity(exitMarker, signed);
  return {
    arrivalMarker: signed,
    exitMarker,
    continuity
  };
}
async function quickAdmit(exitMarker, opts) {
  let marker = null;
  if (typeof exitMarker === "string") {
    marker = fromJSON2(exitMarker);
  } else if (exitMarker) {
    marker = exitMarker;
  }
  const platformIdentity = opts?.platformIdentity ?? (() => {
    if (typeof console !== "undefined" && console.warn) {
      console.warn(
        "[cellar-door-entry] quickAdmit() is generating an ephemeral keypair. Counter-signatures and arrival proofs will reference an unverifiable DID. For production, pass platformIdentity with persistent keys."
      );
    }
    const kp = generateKeyPair();
    return { privateKey: kp.privateKey, publicKey: kp.publicKey };
  })();
  return admit(marker, {
    platformIdentity,
    policy: opts?.policy ?? CAUTIOUS,
    counterSignMeaning: opts?.counterSignMeaning ?? "receipt_only",
    destination: opts?.destination ?? "https://destination.example.com",
    counterSign: opts?.counterSign,
    mintingSubject: opts?.mintingSubject,
    mintingJustification: opts?.mintingJustification
  });
}
function quickEntryP256(exitMarkerJson, destination, opts) {
  return quickEntry(exitMarkerJson, destination, { ...opts, algorithm: "P-256" });
}

// src/probation.ts
var MAX_PROBATION_DURATION_MS = 180 * 24 * 60 * 60 * 1e3;
function createProbationaryArrival(exitMarker, destination, probation, opts) {
  const maxDuration = opts?.maxProbationDuration ?? MAX_PROBATION_DURATION_MS;
  if (probation.duration > maxDuration) {
    console.warn(
      `[cellar-door-entry] Probation duration ${probation.duration}ms exceeds soft cap of ${maxDuration}ms (${Math.round(maxDuration / 864e5)} days). Proceeding anyway.`
    );
  }
  const timestamp = opts?.timestamp ?? (/* @__PURE__ */ new Date()).toISOString();
  const probationInfo = {
    duration: probation.duration,
    restrictions: probation.restrictions,
    reviewRequired: probation.reviewRequired,
    startedAt: timestamp
  };
  return createArrivalMarker(exitMarker, destination, {
    ...opts,
    timestamp,
    admissionType: "conditional",
    conditions: [
      ...opts?.conditions ?? [],
      `probation-${probation.duration}ms`,
      ...probation.restrictions.map((r) => `restriction:${r}`)
    ],
    probation: probationInfo
  });
}
function isProbationComplete(arrival, now) {
  if (!arrival.probation) return true;
  const start = new Date(arrival.probation.startedAt).getTime();
  const currentTime = (now ?? /* @__PURE__ */ new Date()).getTime();
  return currentTime >= start + arrival.probation.duration;
}

// src/capability-scope.ts
function scopeFromExitMarker(marker) {
  const allowed = [];
  if (marker.lineage) allowed.push("identity-continuity");
  if (marker.stateSnapshot) allowed.push("state-portability");
  if (marker.dispute) allowed.push("dispute-context");
  if (marker.economic) allowed.push("economic-portability");
  if (marker.metadata) allowed.push("metadata-access");
  if (marker.crossDomain) allowed.push("cross-domain-reference");
  allowed.push("basic-interaction");
  return { allowed, denied: [] };
}
function createRestrictedScope(allowed, denied, expires) {
  return { allowed, denied, ...expires ? { expires } : {} };
}
function mergeScopes(a, b) {
  const allAllowed = [.../* @__PURE__ */ new Set([...a.allowed, ...b.allowed])];
  const allDenied = [.../* @__PURE__ */ new Set([...a.denied, ...b.denied])];
  const allowed = allAllowed.filter((cap) => !allDenied.includes(cap));
  let expires;
  if (a.expires && b.expires) {
    expires = a.expires < b.expires ? a.expires : b.expires;
  } else {
    expires = a.expires ?? b.expires;
  }
  return { allowed, denied: allDenied, ...expires ? { expires } : {} };
}

// src/claim-tracking.ts
var InMemoryClaimStore = class {
  claims = /* @__PURE__ */ new Map();
  // exitMarkerId → arrivalMarkerId
  reverseIndex = /* @__PURE__ */ new Map();
  // arrivalMarkerId → exitMarkerId
  subjectIndex = /* @__PURE__ */ new Map();
  // subjectDid → Set<exitMarkerId>
  /**
   * Claim an EXIT marker for an arrival. Returns true if successful (unclaimed),
   * false if already claimed.
   */
  async claim(exitMarkerId, arrivalMarkerId, subjectDid) {
    if (this.claims.has(exitMarkerId)) return false;
    this.claims.set(exitMarkerId, arrivalMarkerId);
    this.reverseIndex.set(arrivalMarkerId, exitMarkerId);
    if (subjectDid) {
      let set = this.subjectIndex.get(subjectDid);
      if (!set) {
        set = /* @__PURE__ */ new Set();
        this.subjectIndex.set(subjectDid, set);
      }
      set.add(exitMarkerId);
    }
    return true;
  }
  /** Check if an EXIT marker has been claimed. */
  async isClaimed(exitMarkerId) {
    return this.claims.has(exitMarkerId);
  }
  /** Get the arrival marker ID that claimed this exit marker. */
  async getArrivalId(exitMarkerId) {
    return this.claims.get(exitMarkerId);
  }
  /**
   * Revoke an arrival, unclaiming the associated EXIT marker.
   * Returns true if found and revoked.
   */
  async revoke(arrivalMarkerId) {
    const exitId = this.reverseIndex.get(arrivalMarkerId);
    if (!exitId) return false;
    this.claims.delete(exitId);
    this.reverseIndex.delete(arrivalMarkerId);
    return true;
  }
  /**
   * GDPR Art. 17 — delete all claims involving a given subject DID.
   * Returns the number of claims deleted.
   */
  async deleteBySubject(subjectDid) {
    const exitIds = this.subjectIndex.get(subjectDid);
    if (!exitIds || exitIds.size === 0) return 0;
    let count = 0;
    for (const exitId of exitIds) {
      const arrivalId = this.claims.get(exitId);
      if (arrivalId) {
        this.reverseIndex.delete(arrivalId);
      }
      this.claims.delete(exitId);
      count++;
    }
    this.subjectIndex.delete(subjectDid);
    return count;
  }
  /** Number of active claims. */
  get size() {
    return this.claims.size;
  }
  /** Clear all claims. */
  clear() {
    this.claims.clear();
    this.reverseIndex.clear();
    this.subjectIndex.clear();
  }
};

// src/revocation.ts
import {
  sign as sign2,
  verify as verify2,
  didFromPublicKey as didFromPublicKey2,
  publicKeyFromDid as publicKeyFromDid3,
  signP256 as signP2562,
  verifyP256 as verifyP2562,
  didFromP256PublicKey as didFromP256PublicKey2,
  publicKeyFromP256Did as publicKeyFromP256Did3,
  algorithmFromDid as algorithmFromDid3
} from "cellar-door-exit";
var DOMAIN_PREFIX2 = "entry-revocation-v1.0:";
function createRevocationMarker(arrivalMarker, reason, privateKey, publicKey, algorithm = "Ed25519") {
  const did = algorithm === "P-256" ? didFromP256PublicKey2(publicKey) : didFromPublicKey2(publicKey);
  if (arrivalMarker.proof?.verificationMethod && arrivalMarker.proof.verificationMethod !== did) {
    throw new Error(
      `Authority mismatch: revoker DID ${did} does not match arrival signer ${arrivalMarker.proof.verificationMethod}. Only the destination platform that signed the arrival can revoke it.`
    );
  }
  const body = {
    type: "RevocationMarker",
    arrivalRef: arrivalMarker.id,
    subject: arrivalMarker.subject,
    authority: did,
    reason,
    timestamp: (/* @__PURE__ */ new Date()).toISOString()
  };
  const canonical = canonicalize(body);
  const data = new TextEncoder().encode(DOMAIN_PREFIX2 + canonical);
  const signature = algorithm === "P-256" ? signP2562(data, privateKey) : sign2(data, privateKey);
  const proofValue = btoa(String.fromCharCode(...signature));
  const proofType = algorithm === "P-256" ? "EcdsaP256Signature2019" : "Ed25519Signature2020";
  return {
    ...body,
    proof: {
      type: proofType,
      created: (/* @__PURE__ */ new Date()).toISOString(),
      verificationMethod: did,
      proofValue
    }
  };
}
function verifyRevocationMarker(marker, arrivalMarker) {
  const errors = [];
  if (!marker.proof?.proofValue) {
    return { valid: false, errors: ["Missing proof"] };
  }
  if (arrivalMarker?.proof?.verificationMethod) {
    if (marker.authority !== arrivalMarker.proof.verificationMethod) {
      errors.push(
        `Authority mismatch: revocation authority ${marker.authority} does not match arrival signer ${arrivalMarker.proof.verificationMethod}`
      );
    }
  }
  if (marker.authority && marker.proof.verificationMethod !== marker.authority) {
    errors.push(
      `Proof signer ${marker.proof.verificationMethod} does not match claimed authority ${marker.authority}`
    );
  }
  try {
    const did = marker.proof.verificationMethod;
    const alg = algorithmFromDid3(did);
    const publicKey = alg === "P-256" ? publicKeyFromP256Did3(did) : publicKeyFromDid3(did);
    const { proof: _proof, ...rest } = marker;
    const canonical = canonicalize(rest);
    const data = new TextEncoder().encode(DOMAIN_PREFIX2 + canonical);
    const sigStr = atob(marker.proof.proofValue);
    const signature = new Uint8Array(sigStr.length);
    for (let i = 0; i < sigStr.length; i++) signature[i] = sigStr.charCodeAt(i);
    const valid = alg === "P-256" ? verifyP2562(data, signature, publicKey) : verify2(data, signature, publicKey);
    if (!valid) {
      errors.push("Revocation signature verification failed");
    }
  } catch (e) {
    errors.push(`Revocation verification error: ${e.message}`);
  }
  return { valid: errors.length === 0, errors };
}
function isRevoked(arrivalId, revocations) {
  return revocations.some((r) => r.arrivalRef === arrivalId);
}

// src/transfer.ts
import { verifyMarker as verifyMarker3 } from "cellar-door-exit";
function verifyTransfer(exitMarker, arrivalMarker) {
  const errors = [];
  const exitResult = verifyMarker3(exitMarker);
  if (!exitResult.valid) {
    errors.push(...exitResult.errors.map((e) => `EXIT: ${e}`));
  }
  const arrivalResult = verifyArrivalMarker(arrivalMarker);
  if (!arrivalResult.valid) {
    errors.push(...arrivalResult.errors.map((e) => `ARRIVAL: ${e}`));
  }
  const continuity = verifyContinuity(exitMarker, arrivalMarker);
  if (!continuity.valid) {
    errors.push(...continuity.errors.map((e) => `CONTINUITY: ${e}`));
  }
  const exitTime = new Date(exitMarker.timestamp).getTime();
  const arrivalTime = new Date(arrivalMarker.timestamp).getTime();
  const transferTime = arrivalTime - exitTime;
  return {
    exit: exitMarker,
    arrival: arrivalMarker,
    continuity,
    transferTime,
    verified: errors.length === 0,
    errors
  };
}

// src/mint.ts
async function mintAgent(opts) {
  const result = await admit(null, {
    platformIdentity: opts.platformIdentity,
    policy: opts.policy,
    store: opts.store,
    destination: opts.destination,
    emitter: opts.emitter,
    timestamp: opts.timestamp,
    mintingJustification: opts.justification,
    mintingSubject: opts.subjectDid,
    admissionType: opts.admissionType
  });
  return { ...result, subjectDid: opts.subjectDid };
}
async function bulkMint(agents, opts = {}) {
  const results = [];
  let succeeded = 0;
  let failed = 0;
  for (const agent of agents) {
    const result = await mintAgent({
      ...opts,
      subjectDid: agent.subjectDid,
      justification: agent.justification,
      admissionType: "migration"
    });
    results.push(result);
    if (result.admission.admitted) {
      succeeded++;
    } else {
      failed++;
    }
  }
  return { results, succeeded, failed };
}

// src/events.ts
var AdmissionEventEmitter = class {
  handlers = /* @__PURE__ */ new Map();
  allHandlers = [];
  errorHandler;
  constructor(opts) {
    this.errorHandler = opts?.onError ?? ((err, event) => {
      console.error(`AdmissionEventEmitter: handler error for ${event.type}:`, err);
    });
  }
  on(type, handler) {
    const list = this.handlers.get(type) ?? [];
    list.push(handler);
    this.handlers.set(type, list);
    return this;
  }
  onAny(handler) {
    this.allHandlers.push(handler);
    return this;
  }
  off(type, handler) {
    const list = this.handlers.get(type);
    if (list) {
      this.handlers.set(type, list.filter((h) => h !== handler));
    }
    return this;
  }
  offAny(handler) {
    this.allHandlers = this.allHandlers.filter((h) => h !== handler);
    return this;
  }
  emit(type, data = {}) {
    const event = {
      type,
      timestamp: (/* @__PURE__ */ new Date()).toISOString(),
      data
    };
    const list = this.handlers.get(type) ?? [];
    for (const handler of list) {
      try {
        handler(event);
      } catch (err) {
        this.errorHandler(err, event);
      }
    }
    for (const handler of this.allHandlers) {
      try {
        handler(event);
      } catch (err) {
        this.errorHandler(err, event);
      }
    }
  }
  removeAllListeners() {
    this.handlers.clear();
    this.allHandlers = [];
  }
  listenerCount(type) {
    return (this.handlers.get(type) ?? []).length;
  }
};

// src/sqlite-store.ts
var SqliteClaimStore = class {
  db;
  dbPath;
  opts;
  initialized = false;
  constructor(dbPath, opts) {
    this.dbPath = dbPath;
    this.opts = opts ?? {};
  }
  assertInit() {
    if (!this.initialized) {
      throw new Error("SqliteClaimStore.init() must be called before use");
    }
  }
  /**
   * Initialize the database: create tables, indexes, set pragmas.
   */
  async init() {
    let BetterSqlite3;
    try {
      BetterSqlite3 = (await import("better-sqlite3")).default;
    } catch {
      throw new Error(
        "better-sqlite3 is required for SqliteClaimStore. Install it: npm i better-sqlite3"
      );
    }
    this.db = new BetterSqlite3(this.dbPath);
    this.db.pragma("journal_mode = WAL");
    this.db.pragma("busy_timeout = 5000");
    this.db.pragma("foreign_keys = ON");
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS claims (
        exit_marker_id TEXT PRIMARY KEY,
        arrival_marker_id TEXT NOT NULL,
        subject_did TEXT,
        origin_did TEXT,
        timestamp TEXT DEFAULT (datetime('now'))
      );

      CREATE TABLE IF NOT EXISTS admissions (
        id TEXT PRIMARY KEY,
        exit_marker_id TEXT,
        arrival_marker_id TEXT NOT NULL,
        subject_did TEXT NOT NULL,
        origin_did TEXT,
        confirmation_level TEXT,
        policy_name TEXT NOT NULL,
        policy_version TEXT,
        admitted INTEGER NOT NULL DEFAULT 1,
        admission_type TEXT NOT NULL DEFAULT 'standard',
        counter_signed INTEGER NOT NULL DEFAULT 0,
        counter_sign_meaning TEXT,
        reason_codes TEXT,
        conditions TEXT,
        timestamp TEXT NOT NULL,
        quarantined INTEGER DEFAULT 0,
        quarantine_expires TEXT,
        quarantine_resolution TEXT,
        minting_justification TEXT,
        retention_expires TEXT,
        contested INTEGER DEFAULT 0,
        contest_reason TEXT,
        policy_snapshot TEXT
      );

      CREATE INDEX IF NOT EXISTS idx_claims_subject ON claims(subject_did);
      CREATE INDEX IF NOT EXISTS idx_claims_arrival ON claims(arrival_marker_id);
      CREATE INDEX IF NOT EXISTS idx_admissions_subject ON admissions(subject_did);
      CREATE INDEX IF NOT EXISTS idx_admissions_origin ON admissions(origin_did);
      CREATE INDEX IF NOT EXISTS idx_admissions_policy_time ON admissions(policy_name, timestamp);
      CREATE INDEX IF NOT EXISTS idx_admissions_quarantine ON admissions(quarantined, quarantine_expires);
      CREATE INDEX IF NOT EXISTS idx_admissions_confirmation ON admissions(confirmation_level);
    `);
    this.initialized = true;
  }
  // ─── ClaimStore Interface ──────────────────────────────────────────────────
  async claim(exitMarkerId, arrivalMarkerId, subjectDid) {
    this.assertInit();
    try {
      this.db.prepare(
        "INSERT INTO claims (exit_marker_id, arrival_marker_id, subject_did) VALUES (?, ?, ?)"
      ).run(exitMarkerId, arrivalMarkerId, subjectDid ?? null);
      return true;
    } catch (e) {
      if (e.code === "SQLITE_CONSTRAINT_PRIMARYKEY" || e.message?.includes("UNIQUE")) return false;
      throw e;
    }
  }
  async isClaimed(exitMarkerId) {
    this.assertInit();
    const row = this.db.prepare("SELECT 1 FROM claims WHERE exit_marker_id = ?").get(exitMarkerId);
    return !!row;
  }
  async getArrivalId(exitMarkerId) {
    this.assertInit();
    const row = this.db.prepare("SELECT arrival_marker_id FROM claims WHERE exit_marker_id = ?").get(exitMarkerId);
    return row?.arrival_marker_id;
  }
  async revoke(arrivalMarkerId) {
    this.assertInit();
    const result = this.db.prepare("DELETE FROM claims WHERE arrival_marker_id = ?").run(arrivalMarkerId);
    return result.changes > 0;
  }
  async deleteBySubject(subjectDid) {
    this.assertInit();
    const result = this.db.prepare("DELETE FROM claims WHERE subject_did = ?").run(subjectDid);
    return result.changes;
  }
  // ─── Admission Records ────────────────────────────────────────────────────
  async putAdmission(record) {
    this.assertInit();
    const retentionExpires = this.opts.retentionPeriod ? new Date(new Date(record.timestamp).getTime() + this.opts.retentionPeriod).toISOString() : record.retentionExpires ?? null;
    this.db.prepare(`
      INSERT INTO admissions (
        id, exit_marker_id, arrival_marker_id, subject_did, origin_did,
        confirmation_level, policy_name, policy_version, admitted, admission_type,
        counter_signed, counter_sign_meaning, reason_codes, conditions,
        timestamp, quarantined, quarantine_expires, quarantine_resolution,
        minting_justification, retention_expires, policy_snapshot
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).run(
      record.id,
      record.exitMarkerId ?? null,
      record.arrivalMarkerId,
      record.subjectDid,
      record.originDid ?? null,
      record.confirmationLevel ?? null,
      record.policyApplied,
      record.policyVersion ?? null,
      record.admitted ? 1 : 0,
      record.admissionType,
      record.counterSigned ? 1 : 0,
      record.counterSignMeaning ?? null,
      JSON.stringify(record.reasonCodes),
      JSON.stringify(record.conditions),
      record.timestamp,
      record.quarantined ? 1 : 0,
      record.quarantineExpires ?? null,
      record.quarantineResolution ?? null,
      record.mintingJustification ?? null,
      retentionExpires,
      record.policySnapshot ?? null
    );
  }
  async getAdmission(id) {
    this.assertInit();
    const row = this.db.prepare("SELECT * FROM admissions WHERE id = ?").get(id);
    return row ? this._rowToRecord(row) : void 0;
  }
  async getAdmissionHistory(subjectDid) {
    this.assertInit();
    const rows = this.db.prepare(
      "SELECT * FROM admissions WHERE subject_did = ? ORDER BY timestamp DESC"
    ).all(subjectDid);
    return rows.map((r) => this._rowToRecord(r));
  }
  async getAdmissionsByPolicy(policyName, opts) {
    this.assertInit();
    const limit = opts?.limit ?? 100;
    const offset = opts?.offset ?? 0;
    const rows = this.db.prepare(
      "SELECT * FROM admissions WHERE policy_name = ? ORDER BY timestamp DESC LIMIT ? OFFSET ?"
    ).all(policyName, limit, offset);
    return rows.map((r) => this._rowToRecord(r));
  }
  async getQuarantined() {
    this.assertInit();
    const rows = this.db.prepare(
      "SELECT * FROM admissions WHERE quarantined = 1 AND quarantine_resolution IS NULL ORDER BY timestamp DESC"
    ).all();
    return rows.map((r) => this._rowToRecord(r));
  }
  async getExpiredQuarantines(now) {
    const ts = (now ?? /* @__PURE__ */ new Date()).toISOString();
    const rows = this.db.prepare(
      "SELECT * FROM admissions WHERE quarantined = 1 AND quarantine_resolution IS NULL AND quarantine_expires IS NOT NULL AND quarantine_expires <= ?"
    ).all(ts);
    return rows.map((r) => this._rowToRecord(r));
  }
  /**
   * Contest any admission record (quarantined, rejected, or admitted).
   * Sets contested=1 and contest_reason. For quarantined records, also sets quarantine_resolution='contested'.
   */
  async contestAdmission(recordId, reason) {
    this.assertInit();
    const row = this.db.prepare("SELECT quarantined FROM admissions WHERE id = ?").get(recordId);
    if (!row) return false;
    if (row.quarantined) {
      this.db.prepare(
        "UPDATE admissions SET contested = 1, contest_reason = ?, quarantine_resolution = 'contested' WHERE id = ?"
      ).run(reason, recordId);
    } else {
      this.db.prepare(
        "UPDATE admissions SET contested = 1, contest_reason = ? WHERE id = ?"
      ).run(reason, recordId);
    }
    return true;
  }
  /**
   * Delete all records where retention has expired.
   */
  async purgeExpiredRecords() {
    this.assertInit();
    const now = (/* @__PURE__ */ new Date()).toISOString();
    const claimsResult = this.db.prepare(
      "DELETE FROM claims WHERE exit_marker_id IN (SELECT exit_marker_id FROM admissions WHERE retention_expires IS NOT NULL AND retention_expires <= ?)"
    ).run(now);
    const admResult = this.db.prepare(
      "DELETE FROM admissions WHERE retention_expires IS NOT NULL AND retention_expires <= ?"
    ).run(now);
    return claimsResult.changes + admResult.changes;
  }
  /**
   * Atomically erase all data for a subject (GDPR right-to-erasure).
   */
  async eraseSubject(subjectDid) {
    this.assertInit();
    const tx = this.db.transaction(() => {
      const claims = this.db.prepare("DELETE FROM claims WHERE subject_did = ?").run(subjectDid).changes;
      const admissions = this.db.prepare("DELETE FROM admissions WHERE subject_did = ?").run(subjectDid).changes;
      return { claims, admissions };
    });
    return tx();
  }
  /**
   * Resolve expired quarantines with a default resolution.
   */
  async resolveExpiredQuarantines(defaultResolution) {
    this.assertInit();
    const now = (/* @__PURE__ */ new Date()).toISOString();
    const resolution = defaultResolution === "admit" ? "admitted" : "rejected";
    const result = this.db.prepare(
      "UPDATE admissions SET quarantine_resolution = ? WHERE quarantined = 1 AND quarantine_resolution IS NULL AND quarantine_expires IS NOT NULL AND quarantine_expires <= ?"
    ).run(resolution, now);
    return result.changes;
  }
  async deleteAdmissionsBySubject(subjectDid) {
    this.assertInit();
    const result = this.db.prepare("DELETE FROM admissions WHERE subject_did = ?").run(subjectDid);
    return result.changes;
  }
  async stats() {
    this.assertInit();
    const totalClaims = this.db.prepare("SELECT COUNT(*) as c FROM claims").get().c;
    const totalAdmissions = this.db.prepare("SELECT COUNT(*) as c FROM admissions").get().c;
    const admitted = this.db.prepare("SELECT COUNT(*) as c FROM admissions WHERE admitted = 1").get().c;
    const rejected = this.db.prepare("SELECT COUNT(*) as c FROM admissions WHERE admitted = 0").get().c;
    const quarantined = this.db.prepare("SELECT COUNT(*) as c FROM admissions WHERE quarantined = 1 AND quarantine_resolution IS NULL").get().c;
    const minted = this.db.prepare("SELECT COUNT(*) as c FROM admissions WHERE admission_type = 'minted' OR admission_type = 'migration'").get().c;
    const policyRows = this.db.prepare("SELECT policy_name, COUNT(*) as c FROM admissions GROUP BY policy_name").all();
    const byPolicy = {};
    for (const r of policyRows) byPolicy[r.policy_name] = r.c;
    const confRows = this.db.prepare("SELECT confirmation_level, COUNT(*) as c FROM admissions WHERE confirmation_level IS NOT NULL GROUP BY confirmation_level").all();
    const byConfirmation = {};
    for (const r of confRows) byConfirmation[r.confirmation_level] = r.c;
    return { totalClaims, totalAdmissions, admitted, rejected, quarantined, minted, byPolicy, byConfirmation };
  }
  /**
   * Close the database connection.
   */
  close() {
    this.db?.close();
  }
  _rowToRecord(row) {
    return {
      id: row.id,
      exitMarkerId: row.exit_marker_id ?? void 0,
      arrivalMarkerId: row.arrival_marker_id,
      subjectDid: row.subject_did,
      originDid: row.origin_did ?? void 0,
      confirmationLevel: row.confirmation_level ?? void 0,
      policyApplied: row.policy_name,
      policyVersion: row.policy_version ?? "",
      admitted: !!row.admitted,
      admissionType: row.admission_type,
      counterSigned: !!row.counter_signed,
      counterSignMeaning: row.counter_sign_meaning ?? void 0,
      reasonCodes: row.reason_codes ? JSON.parse(row.reason_codes) : [],
      conditions: row.conditions ? JSON.parse(row.conditions) : [],
      timestamp: row.timestamp,
      quarantined: row.quarantined ? true : void 0,
      quarantineExpires: row.quarantine_expires ?? void 0,
      quarantineResolution: row.quarantine_resolution ?? void 0,
      mintingJustification: row.minting_justification ?? void 0,
      retentionExpires: row.retention_expires ?? void 0,
      contested: row.contested ? true : void 0,
      contestReason: row.contest_reason ?? void 0,
      policySnapshot: row.policy_snapshot ?? void 0
    };
  }
};

// src/validation.ts
var MAX_MARKER_SIZE = 8192;
var DID_REGEX = /^did:[a-z0-9]+:[a-zA-Z0-9._:%-]+$/;
var ISO_TIMESTAMP_REGEX = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z$/;
function validateArrivalMarker(marker) {
  const errors = [];
  if (!marker || typeof marker !== "object") {
    return { valid: false, errors: ["Marker must be a non-null object"] };
  }
  const m = marker;
  try {
    const serialized = JSON.stringify(m);
    const byteLength = new TextEncoder().encode(serialized).byteLength;
    if (byteLength > MAX_MARKER_SIZE) {
      errors.push(`Marker exceeds maximum size of ${MAX_MARKER_SIZE} bytes (got ${byteLength})`);
    }
  } catch {
    errors.push("Marker cannot be serialized to JSON");
  }
  for (const field of ["subject", "departureRef", "departureOrigin", "destination", "id"]) {
    if (typeof m[field] === "string" && /[\x00-\x08\x0b\x0c\x0e-\x1f]/.test(m[field])) {
      errors.push(`${field} contains invalid control characters`);
    }
  }
  if (m["@context"] !== ENTRY_CONTEXT_V1) {
    errors.push(`Invalid @context: expected "${ENTRY_CONTEXT_V1}", got "${m["@context"]}"`);
  }
  if (typeof m.id !== "string" || !m.id.startsWith("urn:entry:")) {
    errors.push(`Invalid id: must be a string starting with "urn:entry:", got "${m.id}"`);
  }
  if (typeof m.departureRef !== "string" || !m.departureRef) {
    errors.push("Missing or invalid departureRef");
  }
  if (typeof m.departureOrigin !== "string" || !m.departureOrigin) {
    errors.push("Missing or invalid departureOrigin");
  }
  if (typeof m.destination !== "string" || !m.destination) {
    errors.push("Missing or invalid destination");
  }
  if (typeof m.subject !== "string") {
    errors.push("Missing subject");
  } else if (!DID_REGEX.test(m.subject)) {
    errors.push(`Invalid subject DID format: "${m.subject}"`);
  }
  if (typeof m.timestamp !== "string") {
    errors.push("Missing timestamp");
  } else if (!ISO_TIMESTAMP_REGEX.test(m.timestamp)) {
    errors.push(`Invalid timestamp format (must be ISO 8601 UTC): "${m.timestamp}"`);
  } else {
    const ts = new Date(m.timestamp).getTime();
    if (isNaN(ts)) {
      errors.push(`Timestamp parses to invalid date: "${m.timestamp}"`);
    }
  }
  const validAdmissionTypes = ["automatic", "reviewed", "conditional"];
  if (!validAdmissionTypes.includes(m.admissionType)) {
    errors.push(`Invalid admissionType: "${m.admissionType}"`);
  }
  if (!m.verificationResult || typeof m.verificationResult !== "object") {
    errors.push("Missing verificationResult");
  } else {
    const vr = m.verificationResult;
    if (typeof vr.valid !== "boolean") {
      errors.push("verificationResult.valid must be a boolean");
    }
    if (!Array.isArray(vr.errors)) {
      errors.push("verificationResult.errors must be an array");
    }
  }
  if (m.conditions !== void 0) {
    if (!Array.isArray(m.conditions) || !m.conditions.every((c) => typeof c === "string")) {
      errors.push("conditions must be an array of strings");
    }
  }
  return { valid: errors.length === 0, errors };
}
export {
  AdmissionEventEmitter,
  CAUTIOUS,
  EMERGENCY_ONLY,
  ENTRY_CONTEXT_V1,
  InMemoryClaimStore,
  LOCKDOWN,
  MAX_MARKER_SIZE,
  OPEN_DOOR,
  OPEN_DOOR_V2,
  PERMISSIVE,
  PolicyBuilder,
  QUARANTINE_UNKNOWN,
  REQUIRE_MUTUAL,
  REQUIRE_MUTUAL_WITH_ONRAMP,
  STRICT,
  SqliteClaimStore,
  admit,
  bulkMint,
  canonicalize,
  computeArrivalId,
  createArrivalMarker,
  createConfirmationRule,
  createCustomRule,
  createDisputeRule,
  createMintingRule,
  createOriginRule,
  createPolicy,
  createProbationaryArrival,
  createRestrictedScope,
  createRevocationMarker,
  createTrustLevelRule,
  evaluateAdmission,
  formatCondition,
  isAdmissionStore,
  isProbationComplete,
  isRevoked,
  mergeScopes,
  mintAgent,
  parseCondition,
  parseDuration,
  quickAdmit,
  quickEntry,
  quickEntryP256,
  resolveConflict,
  scopeFromExitMarker,
  signArrivalMarker,
  validateArrivalMarker,
  verifyArrivalMarker,
  verifyContinuity,
  verifyDeparture,
  verifyDepartureJSON,
  verifyRevocationMarker,
  verifyTransfer
};
