// cellar-door-exit/dist/chunk-KKNMKJSS.js
import {
  computeAnchorHash
} from "./chunk-PXQJCDFL.js";

// src/tsa.ts
import { randomBytes } from "crypto";
var DEFAULT_TSA_URL = "https://freetsa.org/tsr";
var MAX_TSR_SIZE = 1048576;
var DEFAULT_TIMEOUT_MS = 3e4;
var SHA256_OID = Buffer.from([
  96,
  134,
  72,
  1,
  101,
  3,
  4,
  2,
  1
]);
function derLength(len) {
  if (len < 128) return Buffer.from([len]);
  if (len < 256) return Buffer.from([129, len]);
  return Buffer.from([130, len >> 8 & 255, len & 255]);
}
function derWrap(tag, content) {
  return Buffer.concat([Buffer.from([tag]), derLength(content.length), content]);
}
function buildTimestampRequest(hashHex, nonce) {
  const hashBytes = Buffer.from(hashHex, "hex");
  if (hashBytes.length !== 32) {
    throw new Error(`Expected 32-byte SHA-256 hash, got ${hashBytes.length} bytes`);
  }
  const version = derWrap(2, Buffer.from([1]));
  const oid = derWrap(6, SHA256_OID);
  const nullVal = Buffer.from([5, 0]);
  const algorithmId = derWrap(48, Buffer.concat([oid, nullVal]));
  const hashedMessage = derWrap(4, hashBytes);
  const messageImprint = derWrap(48, Buffer.concat([algorithmId, hashedMessage]));
  const nonceBytes = nonce ?? randomBytes(8);
  const noncePayload = nonceBytes[0] & 128 ? Buffer.concat([Buffer.from([0]), nonceBytes]) : nonceBytes;
  const nonceField = derWrap(2, noncePayload);
  const certReq = derWrap(1, Buffer.from([255]));
  return derWrap(48, Buffer.concat([version, messageImprint, nonceField, certReq]));
}
function extractTimestampFromTSR(tsr) {
  for (let i = 0; i < tsr.length - 2; i++) {
    if (tsr[i] === 24) {
      const len = tsr[i + 1];
      if (len >= 13 && len <= 19 && i + 2 + len <= tsr.length) {
        const timeStr = tsr.subarray(i + 2, i + 2 + len).toString("ascii");
        const match = timeStr.match(
          /^(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(?:\.(\d+))?Z$/
        );
        if (match) {
          const [, y, mo, d, h, mi, s, frac] = match;
          const iso = `${y}-${mo}-${d}T${h}:${mi}:${s}${frac ? "." + frac : ""}Z`;
          return iso;
        }
      }
    }
  }
  return null;
}
function checkTSRStatus(tsr) {
  try {
    let pos = 0;
    if (tsr[pos] !== 48) return { ok: false, status: -1 };
    pos++;
    if (tsr[pos] & 128) pos += (tsr[pos] & 127) + 1;
    else pos++;
    if (tsr[pos] !== 48) return { ok: false, status: -1 };
    pos++;
    if (tsr[pos] & 128) pos += (tsr[pos] & 127) + 1;
    else pos++;
    if (tsr[pos] !== 2) return { ok: false, status: -1 };
    pos++;
    const intLen = tsr[pos];
    pos++;
    let status = 0;
    for (let j = 0; j < intLen; j++) {
      status = status << 8 | tsr[pos + j];
    }
    return { ok: status <= 1, status };
  } catch {
    return { ok: false, status: -1 };
  }
}
async function requestTimestamp(hash, tsaUrl = DEFAULT_TSA_URL) {
  const nonce = randomBytes(8);
  const reqDer = buildTimestampRequest(hash, nonce);
  const response = await fetch(tsaUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/timestamp-query"
    },
    body: Uint8Array.from(reqDer),
    signal: AbortSignal.timeout(DEFAULT_TIMEOUT_MS)
  });
  if (!response.ok) {
    throw new Error(`TSA request failed: HTTP ${response.status} ${response.statusText}`);
  }
  const contentLength = response.headers.get("content-length");
  if (contentLength && parseInt(contentLength, 10) > MAX_TSR_SIZE) {
    throw new Error(`TSA response too large: ${contentLength} bytes (max ${MAX_TSR_SIZE})`);
  }
  const arrayBuf = await response.arrayBuffer();
  if (arrayBuf.byteLength > MAX_TSR_SIZE) {
    throw new Error(`TSA response too large: ${arrayBuf.byteLength} bytes (max ${MAX_TSR_SIZE})`);
  }
  const tsrBuffer = Buffer.from(arrayBuf);
  const { ok, status } = checkTSRStatus(tsrBuffer);
  if (!ok) {
    throw new Error(`TSA returned error status: ${status}`);
  }
  const timestamp = extractTimestampFromTSR(tsrBuffer) ?? (/* @__PURE__ */ new Date()).toISOString();
  return {
    tsaUrl,
    hash,
    timestamp,
    receipt: tsrBuffer.toString("base64"),
    requestNonce: nonce.toString("hex")
  };
}
async function anchorWithTSA(marker, tsaUrl) {
  const hash = computeAnchorHash(marker);
  return requestTimestamp(hash, tsaUrl);
}
async function checkTSAReceiptStructure(receipt, hash) {
  if (receipt.hash !== hash) return false;
  let tsrBuffer;
  try {
    tsrBuffer = Buffer.from(receipt.receipt, "base64");
  } catch {
    return false;
  }
  if (tsrBuffer.length < 4 || tsrBuffer[0] !== 48) return false;
  const { ok } = checkTSRStatus(tsrBuffer);
  if (!ok) return false;
  const hashBytes = Buffer.from(hash, "hex");
  const hashPresent = tsrBuffer.includes(hashBytes);
  if (!hashPresent) return false;
  const ts = extractTimestampFromTSR(tsrBuffer);
  if (!ts) return false;
  return true;
}

export {
  buildTimestampRequest,
  extractTimestampFromTSR,
  checkTSRStatus,
  requestTimestamp,
  anchorWithTSA,
  checkTSAReceiptStructure
};
