// cellar-door-exit/dist/chunk-CFRFGOQV.js
import {
  createAnchorRecord
} from "./chunk-PXQJCDFL.js";

// src/git-ledger.ts
import { execFile as execFileCb } from "child_process";
import { promisify } from "util";
import { mkdir, writeFile, readFile, readdir } from "fs/promises";
import { join } from "path";
var execFile = promisify(execFileCb);
function branch(config) {
  return config.branch ?? "exit-ledger";
}
function remote(config) {
  return config.remoteName ?? "origin";
}
async function git(config, ...args) {
  const { stdout } = await execFile("git", ["-C", config.repoPath, ...args]);
  return stdout.trim();
}
async function branchExists(config) {
  try {
    await git(config, "rev-parse", "--verify", branch(config));
    return true;
  } catch {
    return false;
  }
}
async function initLedger(config) {
  await mkdir(config.repoPath, { recursive: true });
  try {
    await git(config, "rev-parse", "--git-dir");
  } catch {
    await execFile("git", ["init", config.repoPath]);
  }
  try {
    await git(config, "config", "user.email");
  } catch {
    await git(config, "config", "user.email", "exit-ledger@cellar-door.local");
    await git(config, "config", "user.name", "EXIT Ledger");
  }
  if (await branchExists(config)) {
    return;
  }
  await git(config, "checkout", "--orphan", branch(config));
  try {
    await git(config, "rm", "-rf", "--cached", ".");
  } catch {
  }
  const ledgerDir = join(config.repoPath, "ledger");
  await mkdir(ledgerDir, { recursive: true });
  await writeFile(join(ledgerDir, ".gitkeep"), "");
  await git(config, "add", "ledger/.gitkeep");
  await git(config, "commit", "-m", "exit-ledger: init");
}
async function anchorToGit(marker, config) {
  await git(config, "checkout", branch(config));
  const record = createAnchorRecord(marker);
  const filePath = join("ledger", `${record.hash}.json`).replace(/\\/g, "/");
  const absPath = join(config.repoPath, filePath);
  const committedAt = (/* @__PURE__ */ new Date()).toISOString();
  const content = {
    ...record,
    committedAt
  };
  await mkdir(join(config.repoPath, "ledger"), { recursive: true });
  await writeFile(absPath, JSON.stringify(content, null, 2) + "\n");
  await git(config, "add", filePath);
  await git(config, "commit", "-m", `exit-ledger: anchor ${record.hash.slice(0, 12)}`);
  const commitHash = await git(config, "rev-parse", "HEAD");
  content.commitHash = commitHash;
  await writeFile(absPath, JSON.stringify(content, null, 2) + "\n");
  await git(config, "add", filePath);
  await git(config, "commit", "--amend", "--no-edit");
  if (config.autoPush) {
    await git(config, "push", remote(config), branch(config));
  }
  return {
    hash: record.hash,
    timestamp: record.timestamp,
    filePath,
    commitHash
  };
}
async function verifyLedgerEntry(hash, config) {
  const absPath = join(config.repoPath, "ledger", `${hash}.json`);
  try {
    const raw = await readFile(absPath, "utf-8");
    const data = JSON.parse(raw);
    return data.hash === hash;
  } catch {
    return false;
  }
}
async function listLedgerEntries(config) {
  const ledgerDir = join(config.repoPath, "ledger");
  let files;
  try {
    files = await readdir(ledgerDir);
  } catch {
    return [];
  }
  const entries = [];
  for (const file of files) {
    if (!file.endsWith(".json")) continue;
    try {
      const raw = await readFile(join(ledgerDir, file), "utf-8");
      const data = JSON.parse(raw);
      entries.push({
        hash: data.hash,
        timestamp: data.timestamp,
        filePath: join("ledger", file),
        commitHash: data.commitHash
      });
    } catch {
    }
  }
  return entries;
}

export {
  initLedger,
  anchorToGit,
  verifyLedgerEntry,
  listLedgerEntries
};
