// cellar-door-exit/dist/chunk-LY5BBBIO.js
// src/visual.ts
function normalizeHash(hash) {
  return hash.replace(/^0x/, "").toLowerCase();
}
function hexAt(hash, index) {
  const h = normalizeHash(hash);
  return parseInt(h[index % h.length], 16) || 0;
}
function hashByte(h, byteIndex) {
  const i = byteIndex * 2 % h.length;
  const hi = parseInt(h[i] || "0", 16);
  const lo = parseInt(h[(i + 1) % h.length] || "0", 16);
  return hi << 4 | lo;
}
var CC = {
  // Frame characters — box drawing vertical lines
  frame_v: ["\u2502", "\u2503", "\u2551", "\u254E", "\u254F", "\u2506", "\u2507", "\u250A"],
  frame_h: ["\u2500", "\u2501", "\u2550", "\u254C", "\u254D", "\u2504", "\u2505", "\u2508"],
  // Corners — grouped by style family (index 0=double, 1=heavy, 2=rounded, 3=light)
  corner_tl: ["\u2554", "\u250F", "\u256D", "\u250C"],
  corner_tr: ["\u2557", "\u2513", "\u256E", "\u2510"],
  corner_bl: ["\u255A", "\u2517", "\u2570", "\u2514"],
  corner_br: ["\u255D", "\u251B", "\u256F", "\u2518"],
  // Panel fills — Discord-safe only (box drawing U+2500-257F, block U+2580-259F, ASCII)
  // Encoding: 8 chars = 3 bits per fill position (log2(8))
  // Voluntary (light): airy, varied — 8 distinct chars for max hash encoding
  panel_light: ["\xB7", "\u2591", " ", "\u254C", "\u254D", "\u2504", "\u2508", "\u2505"],
  // Platform-initiated (heavy): institutional weight — 8 dense fill chars
  panel_medium: ["\u2592", "\u2591", "\u2593", "\u2592", "\u2591", "\u2592", "\u2593", "\u2591"],
  panel_heavy: ["\u2593", "\u2588", "\u2593", "\u2588", "\u2592", "\u2588", "\u2593", "\u2592"],
  // Emergency (distressed): cracked/broken aesthetic — lighter, fractured chars
  panel_emergency: ["\u2504", "\u254C", " ", "\u2508", "\u254D", "\xB7", "\u2505", "\u2591"],
  // Hinges — left and right structural markers
  hinge_l: ["\u255F", "\u2560", "\u251C", "\u255E", "\u251D", "\u2520", "\u2522", "\u251E"],
  hinge_r: ["\u2562", "\u2563", "\u2524", "\u2561", "\u2525", "\u2528", "\u252A", "\u2526"],
  // Handles — symmetric pairs, Discord-safe (█▓ contrast, no ▖▗▘▝▐▌)
  handle_l: ["\u2588", "\u2593", "\u2588", "\u2593", "\u2588", "\u2593", "\u2588", "\u2593"],
  handle_r: ["\u2593", "\u2588", "\u2593", "\u2588", "\u2593", "\u2588", "\u2593", "\u2588"],
  // Arch diagonal characters — hash-varied
  arch_diag_l: ["\u2571", "/", "\u2571", "/", "\u2571", "/", "\u2571", "/"],
  arch_diag_r: ["\u2572", "\\", "\u2572", "\\", "\u2572", "\\", "\u2572", "\\"],
  // Broken seam characters for emergency exits
  seam_broken: ["\u254F", "\u2506", "\u250A", "\u254E", "\u2502", "\u2507", "\u2506", "\u254F"],
  // Threshold — box drawing
  threshold: ["\u2550", "\u2501", "\u2500", "\u2501", "\u2550", "\u2500", "\u2501", "\u2550"],
  // Seam junctions
  seam_btm: ["\u2567", "\u253B", "\u2534", "\u2568"],
  // Crack/damage characters — box drawing only
  crack: ["\u2571", "\u2572", "\u2573", "\u254C", "\u254D", "\u2504", "\u2505", "\u2508"],
  broken: ["\u2504", "\u2506", "\u250A", "\u254C", "\u254D", "\u2505", "\u2508", "\u2507"]
};
function getStyleProfile(exitType, status) {
  const base = {
    familyIdx: 2,
    // rounded by default (elegant)
    panelClass: "panel_light",
    frameWeight: 0,
    damageRate: 0,
    gapRate: 0,
    asymmetric: false
  };
  switch (exitType) {
    case "voluntary":
      base.familyIdx = 2;
      base.panelClass = "panel_light";
      base.frameWeight = 0;
      break;
    case "platform_initiated":
      base.familyIdx = 0;
      base.panelClass = "panel_heavy";
      base.frameWeight = 2;
      break;
    case "emergency":
      base.familyIdx = 1;
      base.panelClass = "panel_emergency";
      base.frameWeight = 1;
      base.damageRate = 2;
      base.gapRate = 4;
      break;
  }
  switch (status) {
    case "pending":
      base.gapRate = Math.max(base.gapRate, 3);
      break;
    case "disputed":
      base.asymmetric = true;
      base.damageRate = Math.max(base.damageRate, 6);
      break;
  }
  return base;
}
function renderDoorASCII(hash, opts) {
  const h = normalizeHash(hash);
  const exitType = opts?.exitType ?? "voluntary";
  const status = opts?.status ?? "complete";
  const origin = opts?.origin;
  const isEntry = opts?.isEntry ?? false;
  const originShift = origin ? [...origin].reduce((a, c) => a + c.charCodeAt(0), 0) % 4 : 0;
  const profile = getStyleProfile(exitType, status);
  const fi = (profile.familyIdx + originShift) % 4;
  let byteIdx = 0;
  function nextByte() {
    return hashByte(h, byteIdx++);
  }
  function pick(cls, b) {
    return cls[b % cls.length];
  }
  const fillCls = CC[profile.panelClass];
  const rightFillCls = profile.asymmetric ? profile.panelClass === "panel_light" || profile.panelClass === "panel_emergency" ? CC.panel_heavy : CC.panel_light : fillCls;
  const emergencyCrackCls = [...CC.crack, ...CC.broken];
  const frameV = pick(CC.frame_v, profile.frameWeight * 2 + hexAt(h, 0) % 2);
  const seamCh = "\u2503";
  const archDiagL = pick(CC.arch_diag_l, hexAt(h, 7));
  const archDiagR = pick(CC.arch_diag_r, hexAt(h, 7));
  function seamAt(row) {
    if (exitType === "emergency") {
      const v = hashByte(h, row + 20);
      return pick(CC.seam_broken, v);
    }
    return seamCh;
  }
  const seamBtm = CC.seam_btm[fi];
  const threshCh = pick(CC.threshold, hexAt(h, 3));
  const handleL = pick(CC.handle_l, hexAt(h, 4));
  const handleR = pick(CC.handle_r, hexAt(h, 5));
  const hingeL = pick(CC.hinge_l, hexAt(h, 6));
  const hingeR = pick(CC.hinge_r, hexAt(h, 6));
  const cornerBL = CC.corner_bl[fi];
  const cornerBR = CC.corner_br[fi];
  function fill(row, col, isRight) {
    const b = nextByte();
    const cls = isRight ? rightFillCls : fillCls;
    let ch = pick(cls, b);
    if (exitType === "emergency") {
      const v = (row * 7 + col * 11 + hexAt(h, (row + col) % h.length)) % 3;
      if (v === 0) return pick(emergencyCrackCls, nextByte());
    } else if (profile.damageRate > 0) {
      const v = (row * 7 + col * 11 + hexAt(h, (row + col) % h.length)) % (profile.damageRate + 8);
      if (v < 1) {
        if (profile.gapRate > 0 && (row + col) % profile.gapRate === 0) return " ";
        return pick(CC.crack, nextByte());
      }
    }
    if (status === "pending") {
      const v = (row * 7 + col * 13) % 10;
      if (v < 2) return " ";
      if (v < 3) return "\xB7";
    }
    return ch;
  }
  const W = 21;
  const PW = 9;
  const HINGE_ROWS = [4, 7];
  const HANDLE_ROW = 5;
  const lines = [];
  const archDefs = [
    { leftEdge: "\u256D", rightEdge: "\u256E", innerHalf: 4, isCrown: true },
    { leftEdge: archDiagL, rightEdge: archDiagR, innerHalf: 6, isCrown: false },
    { leftEdge: archDiagL, rightEdge: archDiagR, innerHalf: 8, isCrown: false }
  ];
  for (let r = 0; r < archDefs.length; r++) {
    const { leftEdge, rightEdge, innerHalf, isCrown } = archDefs[r];
    const contentW = 2 + innerHalf * 2 + 1;
    const pad = Math.floor((W - contentW) / 2);
    let row = " ".repeat(pad) + leftEdge;
    if (isCrown) {
      const hCh = pick(CC.frame_h, hexAt(h, 1));
      row += hCh.repeat(innerHalf) + "\u252C" + hCh.repeat(innerHalf);
    } else {
      for (let c = 0; c < innerHalf; c++) {
        row += fill(r, c, false);
      }
      row += seamAt(r);
      for (let c = 0; c < innerHalf; c++) {
        row += fill(r, c, true);
      }
    }
    row += rightEdge;
    row += " ".repeat(Math.max(0, W - row.length));
    lines.push(row);
  }
  for (let r = 3; r <= 8; r++) {
    let line = "";
    const isHinge = HINGE_ROWS.includes(r);
    line += isHinge ? hingeL : frameV;
    for (let c = 0; c < PW; c++) {
      if (r === HANDLE_ROW && c === PW - 2) {
        line += handleL;
      } else {
        line += fill(r, c, false);
      }
    }
    line += seamAt(r);
    for (let c = 0; c < PW; c++) {
      if (r === HANDLE_ROW && c === 1) {
        line += handleR;
      } else {
        line += fill(r, c, true);
      }
    }
    line += isHinge ? hingeR : frameV;
    lines.push(line);
  }
  lines.push(cornerBL + threshCh.repeat(PW) + seamBtm + threshCh.repeat(PW) + cornerBR);
  if (isEntry) {
    for (let i = 3; i < 6 && i < lines.length; i++) {
      const arr = [...lines[i]];
      if (arr.length > 1) {
        arr[0] = "\u203A";
        arr[arr.length - 1] = "\u2039";
      }
      lines[i] = arr.join("");
    }
  }
  return lines.join("\n");
}
function hashToColors(hash) {
  const h = normalizeHash(hash);
  const colors = [];
  for (let i = 0; i < 5; i++) {
    const offset = i * 6;
    const r = hexAt(h, offset) * 16 + hexAt(h, offset + 1) & 255;
    const g = hexAt(h, offset + 2) * 16 + hexAt(h, offset + 3) & 255;
    const b = hexAt(h, offset + 4) * 16 + hexAt(h, offset + 5) & 255;
    colors.push(
      "#" + r.toString(16).padStart(2, "0") + g.toString(16).padStart(2, "0") + b.toString(16).padStart(2, "0")
    );
  }
  return colors;
}
function renderDoorSVG(hash, opts) {
  const w = opts?.width ?? 200;
  const ht = opts?.height ?? 340;
  const colors = hashToColors(hash);
  const h = normalizeHash(hash);
  const exitType = opts?.exitType ?? "voluntary";
  const status = opts?.status ?? "complete";
  const isEntry = opts?.isEntry ?? false;
  const handleDigit = hexAt(hash, 6);
  const handleSide = hexAt(hash, 19) & 1;
  const handleY = 200 + handleDigit % 5 * 6;
  const handleX = handleSide === 0 ? 90 : 110;
  const hingeW = 6 + hexAt(h, 8) % 4 * 2;
  const panelOpacity = status === "pending" ? 0.6 : 1;
  const doorTransform = status === "pending" ? "translate(3, 0)" : "";
  const disputedSkew = status === "disputed" ? "skewX(2)" : "";
  let damageElements = "";
  if (exitType === "platform_initiated") {
    damageElements = [
      `<line x1="60" y1="140" x2="95" y2="260" stroke="#444" stroke-width="1.5" stroke-dasharray="4,3" opacity="0.7"/>`,
      `<line x1="130" y1="150" x2="110" y2="280" stroke="#444" stroke-width="1" stroke-dasharray="3,4" opacity="0.5"/>`
    ].join("\n  ");
  } else if (exitType === "emergency") {
    damageElements = [
      `<line x1="55" y1="130" x2="100" y2="290" stroke="#666" stroke-width="2" stroke-dasharray="6,2" opacity="0.8"/>`,
      `<line x1="140" y1="140" x2="105" y2="285" stroke="#666" stroke-width="2" stroke-dasharray="5,3" opacity="0.8"/>`,
      `<line x1="75" y1="160" x2="60" y2="290" stroke="#555" stroke-width="1.5" opacity="0.6"/>`,
      `<rect x="95" y="280" width="10" height="20" fill="#1a1a2e" opacity="0.5"/>`
    ].join("\n  ");
  }
  const entryMarker = isEntry ? `<text x="100" y="325" text-anchor="middle" font-size="12" fill="${colors[3]}" font-family="monospace">\u25BC ARRIVAL \u25BC</text>` : "";
  return [
    `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 200 340" width="${w}" height="${ht}">`,
    `  <defs>`,
    `    <clipPath id="arch">`,
    `      <path d="M40,160 L40,300 L160,300 L160,160 A60,70 0 0,0 40,160 Z"/>`,
    `    </clipPath>`,
    `  </defs>`,
    `  <rect width="200" height="340" fill="#1a1a2e" rx="4"/>`,
    `  <g transform="${disputedSkew}">`,
    `  <path d="M35,160 L35,305 L165,305 L165,160 A65,75 0 0,0 35,160 Z" fill="${colors[4]}" stroke="#333" stroke-width="2"/>`,
    `  <g transform="${doorTransform}" opacity="${panelOpacity}">`,
    `    <rect x="42" y="130" width="55" height="168" fill="${colors[0]}" clip-path="url(#arch)" rx="2"/>`,
    `    <rect x="103" y="130" width="55" height="168" fill="${colors[1]}" clip-path="url(#arch)" rx="2"/>`,
    `  </g>`,
    `  <line x1="100" y1="95" x2="100" y2="300" stroke="${colors[2]}" stroke-width="2"/>`,
    `  <path d="M42,160 A58,67 0 0,1 158,160" fill="none" stroke="${colors[3]}" stroke-width="1.5"/>`,
    `  <rect x="38" y="170" width="${hingeW}" height="6" fill="#8b7355" rx="1"/>`,
    `  <rect x="38" y="250" width="${hingeW}" height="6" fill="#8b7355" rx="1"/>`,
    `  <rect x="${162 - hingeW}" y="170" width="${hingeW}" height="6" fill="#8b7355" rx="1"/>`,
    `  <rect x="${162 - hingeW}" y="250" width="${hingeW}" height="6" fill="#8b7355" rx="1"/>`,
    `  <circle cx="${handleX}" cy="${handleY}" r="5" fill="#ffd700" stroke="#b8860b" stroke-width="1.5"/>`,
    `  <circle cx="${handleX}" cy="${handleY + 14}" r="2.5" fill="#1a1a2e"/>`,
    `  <rect x="${handleX - 1}" y="${handleY + 14}" width="2" height="6" fill="#1a1a2e"/>`,
    damageElements ? `  ${damageElements}` : "",
    entryMarker ? `  ${entryMarker}` : "",
    `  </g>`,
    `  <line x1="20" y1="305" x2="180" y2="305" stroke="#555" stroke-width="1"/>`,
    `</svg>`
  ].filter(Boolean).join("\n");
}
function shortHash(hash, isEntry) {
  const h = normalizeHash(hash);
  const seg = h.slice(0, 12).padEnd(12, "0");
  const prefix = isEntry ? "\u{13278}\u279C" : "\u279C\u{13278}";
  return `${prefix} ${seg.slice(0, 4)}-${seg.slice(4, 8)}-${seg.slice(8, 12)}`;
}

export {
  renderDoorASCII,
  hashToColors,
  renderDoorSVG,
  shortHash
};
