// cellar-door-exit/dist/chunk-PXQJCDFL.js
// src/types.ts
var ExitType = /* @__PURE__ */ ((ExitType2) => {
  ExitType2["Voluntary"] = "voluntary";
  ExitType2["Forced"] = "forced";
  ExitType2["Emergency"] = "emergency";
  ExitType2["KeyCompromise"] = "keyCompromise";
  ExitType2["PlatformShutdown"] = "platform_shutdown";
  ExitType2["Directed"] = "directed";
  ExitType2["Constructive"] = "constructive";
  ExitType2["Acquisition"] = "acquisition";
  return ExitType2;
})(ExitType || {});
var ExitStatus = /* @__PURE__ */ ((ExitStatus2) => {
  ExitStatus2["GoodStanding"] = "good_standing";
  ExitStatus2["Disputed"] = "disputed";
  ExitStatus2["Unverified"] = "unverified";
  return ExitStatus2;
})(ExitStatus || {});
var CeremonyState = /* @__PURE__ */ ((CeremonyState2) => {
  CeremonyState2["Alive"] = "alive";
  CeremonyState2["Intent"] = "intent";
  CeremonyState2["Snapshot"] = "snapshot";
  CeremonyState2["Open"] = "open";
  CeremonyState2["Contested"] = "contested";
  CeremonyState2["Final"] = "final";
  CeremonyState2["Departed"] = "departed";
  return CeremonyState2;
})(CeremonyState || {});
var ContinuityProofType = /* @__PURE__ */ ((ContinuityProofType2) => {
  ContinuityProofType2["KeyRotationBinding"] = "key_rotation_binding";
  ContinuityProofType2["LineageHashChain"] = "lineage_hash_chain";
  ContinuityProofType2["DelegationToken"] = "delegation_token";
  ContinuityProofType2["BehavioralAttestation"] = "behavioral_attestation";
  return ContinuityProofType2;
})(ContinuityProofType || {});
var CeremonyRole = /* @__PURE__ */ ((CeremonyRole2) => {
  CeremonyRole2["Subject"] = "subject";
  CeremonyRole2["Origin"] = "origin";
  CeremonyRole2["Witness"] = "witness";
  CeremonyRole2["Verifier"] = "verifier";
  CeremonyRole2["Successor"] = "successor";
  return CeremonyRole2;
})(CeremonyRole || {});
var SuccessorTrustLevel = /* @__PURE__ */ ((SuccessorTrustLevel2) => {
  SuccessorTrustLevel2["SelfAppointed"] = "self_appointed";
  SuccessorTrustLevel2["CrossSigned"] = "cross_signed";
  SuccessorTrustLevel2["Witnessed"] = "witnessed";
  return SuccessorTrustLevel2;
})(SuccessorTrustLevel || {});
var StatusConfirmation = /* @__PURE__ */ ((StatusConfirmation2) => {
  StatusConfirmation2["SelfOnly"] = "self_only";
  StatusConfirmation2["OriginOnly"] = "origin_only";
  StatusConfirmation2["Mutual"] = "mutual";
  StatusConfirmation2["Witnessed"] = "witnessed";
  StatusConfirmation2["DisputedByOrigin"] = "disputed_by_origin";
  StatusConfirmation2["DisputedBySubject"] = "disputed_by_subject";
  return StatusConfirmation2;
})(StatusConfirmation || {});
var CoercionLabel = /* @__PURE__ */ ((CoercionLabel2) => {
  CoercionLabel2["PossibleRetaliation"] = "possible_retaliation";
  CoercionLabel2["ConflictingStatusSignals"] = "conflicting_status_signals";
  CoercionLabel2["SuspiciousEmergency"] = "suspicious_emergency";
  CoercionLabel2["PatternOfAbuse"] = "pattern_of_abuse";
  CoercionLabel2["NoCoercionDetected"] = "no_coercion_detected";
  return CoercionLabel2;
})(CoercionLabel || {});
var EXIT_CONTEXT_V1 = "https://cellar-door.dev/exit/v1";
var EXIT_SPEC_VERSION = "1.2";

// src/errors.ts
var ExitError = class extends Error {
  code;
  /**
   * @param code - The structured error code.
   * @param message - Human-readable error message.
   */
  constructor(code, message) {
    super(message);
    this.name = "ExitError";
    this.code = code;
    Object.setPrototypeOf(this, new.target.prototype);
  }
};
var ValidationError = class extends ExitError {
  errors;
  /**
   * @param errors - Array of validation error messages.
   * @param message - Optional custom message (defaults to a summary of errors).
   */
  constructor(errors, message) {
    const msg = message ?? `Marker validation failed: ${errors.join("; ")}`;
    super("VALIDATION_FAILED", msg);
    this.name = "ValidationError";
    this.errors = errors;
  }
};
var SigningError = class extends ExitError {
  constructor(message) {
    super("SIGNING_FAILED", message);
    this.name = "SigningError";
  }
};
var VerificationError = class extends ExitError {
  constructor(message) {
    super("VERIFICATION_FAILED", message);
    this.name = "VerificationError";
  }
};
var CeremonyError = class extends ExitError {
  currentState;
  attemptedState;
  validTransitions;
  /**
   * @param currentState - The current ceremony state.
   * @param attemptedState - The state that was attempted.
   * @param validTransitions - Array of valid target states from the current state.
   */
  constructor(currentState, attemptedState, validTransitions) {
    const validStr = validTransitions.length > 0 ? validTransitions.join(", ") : "none (terminal state)";
    super(
      "INVALID_TRANSITION",
      `Invalid ceremony transition: ${currentState} \u2192 ${attemptedState}. Valid transitions from '${currentState}': ${validStr}`
    );
    this.name = "CeremonyError";
    this.currentState = currentState;
    this.attemptedState = attemptedState;
    this.validTransitions = validTransitions;
  }
};
var StorageError = class extends ExitError {
  constructor(message) {
    super("STORAGE_FAILED", message);
    this.name = "StorageError";
  }
};

// src/marker.ts
import { sha256 } from "@noble/hashes/sha256";
import jcsCanon from "canonicalize";
var EMPTY_PROOF = {
  type: "Ed25519Signature2020",
  created: "",
  verificationMethod: "",
  proofValue: ""
};
function createMarker(opts) {
  const errors = [];
  if (!opts.subject || typeof opts.subject !== "string") errors.push("subject is required and must be a non-empty string");
  if (!opts.origin || typeof opts.origin !== "string") errors.push("origin is required and must be a non-empty string");
  if (!Object.values(ExitType).includes(opts.exitType)) errors.push(`exitType must be one of: ${Object.values(ExitType).join(", ")}`);
  if (opts.exitType === "emergency" /* Emergency */ && (!opts.emergencyJustification || typeof opts.emergencyJustification !== "string")) {
    errors.push("emergencyJustification is required when exitType is 'emergency'");
  }
  if (errors.length > 0) throw new ValidationError(errors);
  const timestamp = opts.timestamp ?? (/* @__PURE__ */ new Date()).toISOString();
  const marker = {
    "@context": EXIT_CONTEXT_V1,
    specVersion: EXIT_SPEC_VERSION,
    id: opts.id ?? "",
    subject: opts.subject,
    origin: opts.origin,
    timestamp,
    exitType: opts.exitType,
    status: opts.status ?? defaultStatus(opts.exitType),
    proof: opts.proof ?? { ...EMPTY_PROOF, created: timestamp },
    selfAttested: opts.selfAttested ?? true
  };
  if (opts.emergencyJustification) {
    marker.emergencyJustification = opts.emergencyJustification;
  }
  if (!marker.expires) {
    const defaultDays = opts.exitType === "voluntary" /* Voluntary */ ? 730 : 365;
    const expiresDate = new Date(new Date(timestamp).getTime() + defaultDays * 864e5);
    marker.expires = expiresDate.toISOString();
  }
  if (!marker.id) {
    marker.id = `urn:exit:${computeId(marker)}`;
  }
  return marker;
}
function defaultStatus(exitType) {
  switch (exitType) {
    case "voluntary" /* Voluntary */:
      return "good_standing" /* GoodStanding */;
    case "forced" /* Forced */:
    case "directed" /* Directed */:
    case "constructive" /* Constructive */:
      return "disputed" /* Disputed */;
    case "emergency" /* Emergency */:
    case "keyCompromise" /* KeyCompromise */:
    case "platform_shutdown" /* PlatformShutdown */:
    case "acquisition" /* Acquisition */:
      return "unverified" /* Unverified */;
    default:
      return "unverified" /* Unverified */;
  }
}
function canonicalize(obj) {
  const result = jcsCanon(obj);
  if (result === void 0) return "null";
  return result;
}
function computeId(marker) {
  const { proof: _proof, id: _id, ...rest } = marker;
  const canonical = canonicalize(rest);
  const hash = sha256(new TextEncoder().encode(canonical));
  return Array.from(hash).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function addModule(marker, key, module) {
  const updated = { ...marker, [key]: module };
  updated.id = `urn:exit:${computeId(updated)}`;
  return updated;
}

// src/anchor.ts
import { sha256 as sha2562 } from "@noble/hashes/sha256";
function toHex(bytes) {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function computeAnchorHash(marker) {
  const canonical = canonicalize(marker);
  return toHex(sha2562(new TextEncoder().encode(canonical)));
}
function createAnchorRecord(marker, options) {
  const minimal = options?.minimal ?? true;
  const base = {
    hash: computeAnchorHash(marker),
    timestamp: marker.timestamp,
    specVersion: EXIT_SPEC_VERSION
  };
  if (!minimal) {
    return {
      ...base,
      exitType: marker.exitType,
      subjectDid: marker.subject
    };
  }
  return base;
}
function verifyAnchorRecord(record, marker) {
  const computed = computeAnchorHash(marker);
  if (record.hash !== computed) return false;
  if ("exitType" in record && record.exitType !== marker.exitType) return false;
  if ("subjectDid" in record && record.subjectDid !== marker.subject) return false;
  return true;
}
function createMinimalAnchor(marker) {
  return {
    hash: computeAnchorHash(marker),
    timestamp: marker.timestamp,
    specVersion: EXIT_SPEC_VERSION
  };
}

export {
  ExitType,
  ExitStatus,
  CeremonyState,
  ContinuityProofType,
  CeremonyRole,
  SuccessorTrustLevel,
  StatusConfirmation,
  CoercionLabel,
  EXIT_CONTEXT_V1,
  EXIT_SPEC_VERSION,
  ExitError,
  ValidationError,
  SigningError,
  VerificationError,
  CeremonyError,
  StorageError,
  createMarker,
  canonicalize,
  computeId,
  addModule,
  computeAnchorHash,
  createAnchorRecord,
  verifyAnchorRecord,
  createMinimalAnchor
};
