// cellar-door-exit/dist/chunk-H5F44PDI.js
import {
  CeremonyError,
  EXIT_CONTEXT_V1,
  EXIT_SPEC_VERSION,
  ExitStatus,
  ExitType,
  SigningError,
  ValidationError,
  canonicalize,
  computeAnchorHash,
  computeId,
  createMarker
} from "./chunk-PXQJCDFL.js";
import {
  __require
} from "./chunk-3RG5ZIWI.js";

// src/crypto.ts
import { webcrypto } from "crypto";
import * as ed from "@noble/ed25519";
import { sha512 } from "@noble/hashes/sha512";
import { p256 } from "@noble/curves/nist.js";
if (!globalThis.crypto) {
  globalThis.crypto = webcrypto;
}
ed.etc.sha512Sync = (...m) => {
  const h = sha512.create();
  for (const msg of m) h.update(msg);
  return h.digest();
};
var ED25519_MULTICODEC = new Uint8Array([237, 1]);
var P256_MULTICODEC = new Uint8Array([128, 36]);
function generateKeyPair() {
  const privateKey = ed.utils.randomPrivateKey();
  const publicKey = ed.getPublicKey(privateKey);
  return { publicKey, privateKey };
}
function sign2(data, privateKey) {
  return ed.sign(data, privateKey);
}
function verify2(data, signature, publicKey) {
  try {
    return ed.verify(signature, data, publicKey);
  } catch {
    return false;
  }
}
function base58btcEncode(bytes) {
  const ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
  let num = 0n;
  for (const b of bytes) num = num * 256n + BigInt(b);
  let encoded = "";
  while (num > 0n) {
    encoded = ALPHABET[Number(num % 58n)] + encoded;
    num /= 58n;
  }
  for (const b of bytes) {
    if (b === 0) encoded = "1" + encoded;
    else break;
  }
  return encoded || "1";
}
function base58btcDecode(str) {
  const ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
  let num = 0n;
  for (const c of str) {
    const idx = ALPHABET.indexOf(c);
    if (idx === -1) throw new Error(`Invalid base58 character: ${c}`);
    num = num * 58n + BigInt(idx);
  }
  const hex = num.toString(16).padStart(2, "0");
  const paddedHex = hex.length % 2 ? "0" + hex : hex;
  const bytes = [];
  for (let i = 0; i < paddedHex.length; i += 2) {
    bytes.push(parseInt(paddedHex.slice(i, i + 2), 16));
  }
  let leadingZeros = 0;
  for (const c of str) {
    if (c === "1") leadingZeros++;
    else break;
  }
  const result = new Uint8Array(leadingZeros + bytes.length);
  result.set(new Uint8Array(bytes), leadingZeros);
  return result;
}
function didFromPublicKey(publicKey) {
  const multicodec = new Uint8Array(ED25519_MULTICODEC.length + publicKey.length);
  multicodec.set(ED25519_MULTICODEC);
  multicodec.set(publicKey, ED25519_MULTICODEC.length);
  return `did:key:z${base58btcEncode(multicodec)}`;
}
function publicKeyFromDid(did) {
  if (!did.startsWith("did:key:z")) {
    throw new Error("Invalid did:key format: must start with 'did:key:z'");
  }
  const encoded = did.slice("did:key:z".length);
  const decoded = base58btcDecode(encoded);
  if (decoded[0] === 237 && decoded[1] === 1) {
    const key = decoded.slice(2);
    if (key.length !== 32) {
      throw new Error(`Invalid Ed25519 public key length: expected 32 bytes, got ${key.length}`);
    }
    return key;
  }
  if (decoded[0] === 128 && decoded[1] === 36) {
    const key = decoded.slice(2);
    if (key.length !== 33) {
      throw new Error(`Invalid P-256 public key length: expected 33 bytes, got ${key.length}`);
    }
    return key;
  }
  throw new Error("Invalid multicodec prefix: expected Ed25519 (0xed01) or P-256 (0x8024)");
}
function generateP256KeyPair() {
  const privateKey = p256.utils.randomSecretKey();
  const publicKey = p256.getPublicKey(privateKey, true);
  return { publicKey, privateKey };
}
function signP256(data, privateKey) {
  const sig = p256.sign(data, privateKey, { prehash: true });
  return Uint8Array.from(sig);
}
function verifyP256(data, signature, publicKey) {
  try {
    return p256.verify(signature, data, publicKey, { prehash: true });
  } catch {
    return false;
  }
}
function didFromP256PublicKey(publicKey) {
  const multicodec = new Uint8Array(P256_MULTICODEC.length + publicKey.length);
  multicodec.set(P256_MULTICODEC);
  multicodec.set(publicKey, P256_MULTICODEC.length);
  return `did:key:z${base58btcEncode(multicodec)}`;
}
function publicKeyFromP256Did(did) {
  if (!did.startsWith("did:key:z")) {
    throw new Error("Invalid did:key format: must start with 'did:key:z'");
  }
  const encoded = did.slice("did:key:z".length);
  const decoded = base58btcDecode(encoded);
  if (decoded[0] !== 128 || decoded[1] !== 36) {
    throw new Error("Invalid multicodec prefix: expected P-256 (0x8024)");
  }
  return decoded.slice(2);
}
var AlgorithmRegistry = /* @__PURE__ */ new Map();
AlgorithmRegistry.set("ed25519", {
  generateKeyPair,
  sign: sign2,
  verify: verify2,
  didFromPublicKey,
  publicKeyFromDid
});
AlgorithmRegistry.set("p256", {
  generateKeyPair: generateP256KeyPair,
  sign: signP256,
  verify: verifyP256,
  didFromPublicKey: didFromP256PublicKey,
  publicKeyFromDid: publicKeyFromP256Did
});
function getAlgorithm(id) {
  const alg = AlgorithmRegistry.get(id);
  if (!alg) {
    throw new Error(`Unknown algorithm: "${id}". Registered: ${[...AlgorithmRegistry.keys()].join(", ")}`);
  }
  return alg;
}
function algorithmFromDid(did) {
  if (!did.startsWith("did:key:z")) {
    throw new Error("Invalid did:key format");
  }
  const encoded = did.slice("did:key:z".length);
  const decoded = base58btcDecode(encoded);
  if (decoded[0] === 237 && decoded[1] === 1) return "Ed25519";
  if (decoded[0] === 128 && decoded[1] === 36) return "P-256";
  throw new Error("Unknown multicodec prefix");
}

// src/signer.ts
var Ed25519Signer = class {
  algorithm = "Ed25519";
  #privateKey;
  #publicKey;
  constructor(privateKey, publicKey) {
    this.#privateKey = privateKey;
    this.#publicKey = publicKey;
  }
  sign(data) {
    return sign2(data, this.#privateKey);
  }
  verify(data, signature) {
    return verify2(data, signature, this.#publicKey);
  }
  did() {
    return didFromPublicKey(this.#publicKey);
  }
  publicKey() {
    return this.#publicKey;
  }
  /**
   * Zero the private key material (best-effort).
   *
   * **B2/B3:** In JavaScript, memory is managed by the GC and copies may exist
   * in optimized JIT code or V8 internals. This zeroing is best-effort — it
   * overwrites the backing ArrayBuffer but cannot guarantee all copies are erased.
   * For high-security contexts, consider hardware-backed signing (HSM/TPM).
   */
  destroy() {
    this.#privateKey.fill(0);
  }
};
var P256Signer = class {
  algorithm = "P-256";
  #privateKey;
  #publicKey;
  constructor(privateKey, publicKey) {
    this.#privateKey = privateKey;
    this.#publicKey = publicKey;
  }
  sign(data) {
    const sig = signP256(data, this.#privateKey);
    return new Uint8Array(sig);
  }
  verify(data, signature) {
    return verifyP256(data, signature, this.#publicKey);
  }
  did() {
    return didFromP256PublicKey(this.#publicKey);
  }
  publicKey() {
    return this.#publicKey;
  }
  /**
   * Zero the private key material (best-effort).
   * @see Ed25519Signer.destroy for caveats.
   */
  destroy() {
    this.#privateKey.fill(0);
  }
};
function generateKeyPairForAlgorithm(algorithm = "Ed25519") {
  if (algorithm === "P-256") return generateP256KeyPair();
  return generateKeyPair();
}
function createSigner(options) {
  const alg = options?.algorithm ?? "Ed25519";
  let privateKey = options?.privateKey;
  let publicKey = options?.publicKey;
  if (!privateKey) {
    const kp = generateKeyPairForAlgorithm(alg);
    privateKey = kp.privateKey;
    publicKey = kp.publicKey;
  }
  if (!publicKey) {
    throw new Error("publicKey is required when privateKey is provided");
  }
  if (alg === "P-256") return new P256Signer(privateKey, publicKey);
  return new Ed25519Signer(privateKey, publicKey);
}
function createVerifier(didStr, pubKey) {
  const alg = algorithmFromDid(didStr);
  return {
    algorithm: alg,
    verify(data, signature) {
      if (alg === "P-256") return verifyP256(data, signature, pubKey);
      return verify2(data, signature, pubKey);
    },
    did() {
      return didStr;
    },
    publicKey() {
      return pubKey;
    }
  };
}
function proofTypeForAlgorithm(alg) {
  if (alg === "P-256") return "EcdsaP256Signature2019";
  return "Ed25519Signature2020";
}
function algorithmFromProofType(proofType) {
  if (proofType === "Ed25519Signature2020") return "Ed25519";
  if (proofType === "EcdsaP256Signature2019") return "P-256";
  return null;
}

// src/validate.ts
var ISO_8601_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z?$/;
var DID_KEY_RE = /^did:key:z[1-9A-HJ-NP-Za-km-z]+$/;
var MAX_URI_LENGTH = 2048;
var MAX_LABEL_LENGTH = 256;
var MAX_STRING_LENGTH = 4096;
var HEX_HASH_RE = /^[0-9a-f]{64}$/i;
function isValidISO8601(s) {
  return ISO_8601_RE.test(s) && !isNaN(Date.parse(s));
}
function isValidDid(s) {
  return DID_KEY_RE.test(s) || s.startsWith("did:");
}
function containsControlChars(s) {
  return /[\x00-\x08\x0b\x0c\x0e-\x1f]/.test(s);
}
function validateMarker(marker) {
  const errors = [];
  const m = marker;
  if (!m || typeof m !== "object") {
    return { valid: false, errors: ["Marker must be an object"] };
  }
  try {
    const serialized = JSON.stringify(m);
    if (new TextEncoder().encode(serialized).byteLength > 8192) {
      return { valid: false, errors: ["Marker exceeds maximum size of 8192 bytes"] };
    }
  } catch {
    return { valid: false, errors: ["Marker cannot be serialized to JSON"] };
  }
  if (m["@context"] !== EXIT_CONTEXT_V1) {
    errors.push(`Invalid @context: expected "${EXIT_CONTEXT_V1}"`);
  }
  if (typeof m.specVersion !== "string" || !m.specVersion) {
    errors.push("Missing or invalid required field: specVersion");
  } else if (m.specVersion !== EXIT_SPEC_VERSION) {
    errors.push(`Unsupported specVersion: expected "${EXIT_SPEC_VERSION}", got "${m.specVersion}"`);
  }
  if (typeof m.id !== "string" || !m.id) {
    errors.push("Missing or invalid required field: id");
  }
  if (typeof m.subject !== "string" || !m.subject) {
    errors.push("Missing or invalid required field: subject");
  }
  if (typeof m.origin !== "string" || !m.origin) {
    errors.push("Missing or invalid required field: origin");
  }
  if (typeof m.timestamp !== "string" || !m.timestamp) {
    errors.push("Missing or invalid required field: timestamp");
  } else if (!isValidISO8601(m.timestamp)) {
    errors.push("Invalid timestamp: must be ISO-8601 format");
  }
  if (!Object.values(ExitType).includes(m.exitType)) {
    errors.push(`Invalid exitType: must be one of ${Object.values(ExitType).join(", ")}`);
  }
  if (!Object.values(ExitStatus).includes(m.status)) {
    errors.push(`Invalid status: must be one of ${Object.values(ExitStatus).join(", ")}`);
  }
  if (!m.proof || typeof m.proof !== "object") {
    errors.push("Missing or invalid required field: proof");
  } else {
    const p = m.proof;
    if (typeof p.type !== "string" || !p.type) errors.push("Proof missing type");
    if (typeof p.proofValue !== "string") errors.push("Proof missing proofValue");
    if (typeof p.verificationMethod !== "string") errors.push("Proof missing verificationMethod");
  }
  for (const field of ["subject", "origin", "id"]) {
    if (typeof m[field] === "string" && containsControlChars(m[field])) {
      errors.push(`${field} contains invalid control characters`);
    }
  }
  if (typeof m.subject === "string" && m.subject.startsWith("did:") && !isValidDid(m.subject)) {
    errors.push("Invalid subject DID format");
  }
  if (m.exitType === "emergency" /* Emergency */) {
    if (typeof m.emergencyJustification !== "string" || !m.emergencyJustification) {
      errors.push("emergencyJustification is required when exitType is 'emergency'");
    }
  }
  if (m.legalHold !== void 0) {
    if (typeof m.legalHold !== "object" || m.legalHold === null) {
      errors.push("legalHold must be an object");
    } else {
      const lh = m.legalHold;
      if (typeof lh.holdType !== "string" || !lh.holdType) errors.push("legalHold.holdType is required");
      if (typeof lh.authority !== "string" || !lh.authority) errors.push("legalHold.authority is required");
      if (typeof lh.reference !== "string" || !lh.reference) errors.push("legalHold.reference is required");
      if (typeof lh.dateIssued !== "string" || !lh.dateIssued) {
        errors.push("legalHold.dateIssued is required");
      } else if (!isValidISO8601(lh.dateIssued)) {
        errors.push("legalHold.dateIssued must be ISO-8601 format");
      }
      if (typeof lh.acknowledged !== "boolean") errors.push("legalHold.acknowledged must be a boolean");
    }
  }
  if (m.expires !== void 0) {
    if (typeof m.expires !== "string" || !isValidISO8601(m.expires)) {
      errors.push("expires must be a valid ISO-8601 date string");
    }
  }
  if (m.sequenceNumber !== void 0) {
    if (typeof m.sequenceNumber !== "number" || !Number.isInteger(m.sequenceNumber) || m.sequenceNumber < 0) {
      errors.push("sequenceNumber must be a non-negative integer");
    }
  }
  if (m.completenessAttestation !== void 0) {
    if (typeof m.completenessAttestation !== "object" || m.completenessAttestation === null) {
      errors.push("completenessAttestation must be an object");
    }
  }
  if (m.disputeExpiry !== void 0) {
    if (typeof m.disputeExpiry !== "string" || !isValidISO8601(m.disputeExpiry)) {
      errors.push("disputeExpiry must be a valid ISO-8601 date string");
    }
  }
  if (m.resolution !== void 0) {
    if (typeof m.resolution !== "string" || !m.resolution) {
      errors.push("resolution must be a non-empty string");
    }
  }
  if (m.arbiterDid !== void 0) {
    if (typeof m.arbiterDid !== "string" || !m.arbiterDid.startsWith("did:")) {
      errors.push('arbiterDid must be a string starting with "did:"');
    }
  }
  if (m.lineage !== void 0 && (typeof m.lineage !== "object" || m.lineage === null)) {
    errors.push("Module A (lineage) must be an object");
  }
  if (m.stateSnapshot !== void 0) {
    const ss = m.stateSnapshot;
    if (typeof ss !== "object" || ss === null) {
      errors.push("Module B (stateSnapshot) must be an object");
    } else if (typeof ss.stateHash !== "string" || !ss.stateHash) {
      errors.push("Module B (stateSnapshot) requires stateHash");
    }
  }
  if (m.trustEnhancers !== void 0) {
    if (typeof m.trustEnhancers !== "object" || m.trustEnhancers === null) {
      errors.push("trustEnhancers must be an object");
    } else {
      const te = m.trustEnhancers;
      if (te.timestamps !== void 0) {
        if (!Array.isArray(te.timestamps)) {
          errors.push("trustEnhancers.timestamps must be an array");
        } else {
          for (let i = 0; i < te.timestamps.length; i++) {
            const ts = te.timestamps[i];
            if (typeof ts.tsaUrl !== "string" || !ts.tsaUrl) {
              errors.push(`trustEnhancers.timestamps[${i}].tsaUrl is required`);
            }
            if (typeof ts.hash !== "string" || !/^[0-9a-f]{64}$/i.test(ts.hash)) {
              errors.push(`trustEnhancers.timestamps[${i}].hash must be a 64-char hex SHA-256`);
            }
            if (typeof ts.timestamp !== "string" || !isValidISO8601(ts.timestamp)) {
              errors.push(`trustEnhancers.timestamps[${i}].timestamp must be ISO-8601`);
            }
            if (typeof ts.receipt !== "string" || !ts.receipt) {
              errors.push(`trustEnhancers.timestamps[${i}].receipt is required`);
            }
          }
        }
      }
      if (te.witnesses !== void 0) {
        if (!Array.isArray(te.witnesses)) {
          errors.push("trustEnhancers.witnesses must be an array");
        } else {
          for (let i = 0; i < te.witnesses.length; i++) {
            const w = te.witnesses[i];
            if (typeof w.witnessDid !== "string" || !w.witnessDid) {
              errors.push(`trustEnhancers.witnesses[${i}].witnessDid is required`);
            }
            if (typeof w.attestation !== "string" || !w.attestation) {
              errors.push(`trustEnhancers.witnesses[${i}].attestation is required`);
            }
            if (typeof w.timestamp !== "string" || !isValidISO8601(w.timestamp)) {
              errors.push(`trustEnhancers.witnesses[${i}].timestamp must be ISO-8601`);
            }
            if (typeof w.signature !== "string" || !w.signature) {
              errors.push(`trustEnhancers.witnesses[${i}].signature is required`);
            }
            if (typeof w.signatureType !== "string" || !w.signatureType) {
              errors.push(`trustEnhancers.witnesses[${i}].signatureType is required`);
            }
          }
        }
      }
      if (te.identityClaims !== void 0) {
        if (!Array.isArray(te.identityClaims)) {
          errors.push("trustEnhancers.identityClaims must be an array");
        } else {
          for (let i = 0; i < te.identityClaims.length; i++) {
            const ic = te.identityClaims[i];
            if (typeof ic.scheme !== "string" || !ic.scheme) {
              errors.push(`trustEnhancers.identityClaims[${i}].scheme is required`);
            }
            if (typeof ic.value !== "string" || !ic.value) {
              errors.push(`trustEnhancers.identityClaims[${i}].value is required`);
            }
            if (typeof ic.issuedAt !== "string" || !isValidISO8601(ic.issuedAt)) {
              errors.push(`trustEnhancers.identityClaims[${i}].issuedAt must be ISO-8601`);
            }
            if (ic.expiresAt !== void 0 && (typeof ic.expiresAt !== "string" || !isValidISO8601(ic.expiresAt))) {
              errors.push(`trustEnhancers.identityClaims[${i}].expiresAt must be ISO-8601`);
            }
          }
        }
      }
    }
  }
  const warnings = [];
  for (const field of ["id", "origin"]) {
    if (typeof m[field] === "string" && m[field].length > MAX_URI_LENGTH) {
      warnings.push(`${field} exceeds recommended maximum length of ${MAX_URI_LENGTH} characters`);
    }
  }
  for (const field of ["subject"]) {
    if (typeof m[field] === "string" && m[field].length > MAX_URI_LENGTH) {
      warnings.push(`${field} exceeds recommended maximum length of ${MAX_URI_LENGTH} characters`);
    }
  }
  if (typeof m.coercionLabel === "string" && m.coercionLabel.length > MAX_LABEL_LENGTH) {
    warnings.push(`coercionLabel exceeds recommended maximum length of ${MAX_LABEL_LENGTH} characters`);
  }
  for (const field of ["reason", "narrative", "emergencyJustification", "resolution"]) {
    if (typeof m[field] === "string" && m[field].length > MAX_STRING_LENGTH) {
      warnings.push(`${field} exceeds recommended maximum length of ${MAX_STRING_LENGTH} characters`);
    }
  }
  if (typeof m.timestamp === "string" && ISO_8601_RE.test(m.timestamp) && !m.timestamp.endsWith("Z")) {
    warnings.push("timestamp missing 'Z' suffix \u2014 treating as UTC (B7)");
  }
  if (m.selfAttested !== void 0 && typeof m.selfAttested !== "boolean") {
    warnings.push("selfAttested should be a boolean");
  }
  if (m.coercionLabel !== void 0) {
    if (typeof m.coercionLabel !== "string") {
      warnings.push("coercionLabel should be a string");
    }
  }
  if (m.preRotationCommitment !== void 0) {
    if (typeof m.preRotationCommitment !== "string") {
      warnings.push("preRotationCommitment should be a string");
    } else if (!HEX_HASH_RE.test(m.preRotationCommitment)) {
      warnings.push("preRotationCommitment should be a 64-character hex SHA-256 hash");
    }
  }
  if (m.expires === void 0) {
    warnings.push("Missing required 'expires' field (\xA78.5). Implementations MUST populate with default: 730 days for voluntary exits, 365 days for involuntary exits.");
  }
  if (m.sunsetDate !== void 0) {
    if (typeof m.sunsetDate !== "string" || !isValidISO8601(m.sunsetDate)) {
      warnings.push("sunsetDate should be a valid ISO 8601 date string");
    }
  }
  if (m.completenessAttestation !== void 0 && typeof m.completenessAttestation === "object" && m.completenessAttestation !== null) {
    const ca = m.completenessAttestation;
    if (ca.attestedAt !== void 0 && (typeof ca.attestedAt !== "string" || !isValidISO8601(ca.attestedAt))) {
      warnings.push("completenessAttestation.attestedAt should be a valid ISO 8601 date");
    }
    if (ca.markerCount !== void 0 && (typeof ca.markerCount !== "number" || !Number.isInteger(ca.markerCount) || ca.markerCount < 0)) {
      warnings.push("completenessAttestation.markerCount should be a non-negative integer");
    }
    if (ca.signature !== void 0 && typeof ca.signature !== "string") {
      warnings.push("completenessAttestation.signature should be a string");
    }
  }
  if (m.intent !== void 0 && typeof m.intent === "object" && m.intent !== null) {
    const intent = m.intent;
    if (typeof intent.subject === "string" && typeof m.subject === "string" && intent.subject !== m.subject) {
      warnings.push("ExitIntent subject does not match marker subject");
    }
  }
  const result = { valid: errors.length === 0, errors };
  if (warnings.length > 0) result.warnings = warnings;
  return result;
}

// src/proof.ts
var DOMAIN_PREFIX = "exit-marker-v1.2:";
function signMarker(marker, privateKey, publicKey) {
  try {
    const did = didFromPublicKey(publicKey);
    const now = (/* @__PURE__ */ new Date()).toISOString();
    const { proof: _proof, id: _id, ...rest } = marker;
    const canonical = canonicalize(rest);
    const data = new TextEncoder().encode(DOMAIN_PREFIX + canonical);
    const signature = sign2(data, privateKey);
    const proofValue = Buffer.from(signature).toString("base64");
    const proof = {
      type: "Ed25519Signature2020",
      created: now,
      verificationMethod: did,
      proofValue
    };
    return { ...marker, proof };
  } catch (e) {
    if (e instanceof SigningError) throw e;
    throw new SigningError(`Failed to sign marker: ${e.message}`);
  }
}
function verifyMarker(marker, options) {
  const errors = [];
  const schema = validateMarker(marker);
  if (!schema.valid) {
    errors.push(...schema.errors);
  }
  if (!marker.proof) {
    errors.push("Missing proof");
    return { valid: false, errors };
  }
  const alg = algorithmFromProofType(marker.proof.type);
  if (!alg) {
    errors.push(`Unsupported proof type: ${marker.proof.type}`);
    return { valid: false, errors };
  }
  if (!marker.proof.verificationMethod || !marker.proof.proofValue) {
    errors.push("Incomplete proof: missing verificationMethod or proofValue");
    return { valid: false, errors };
  }
  if (!marker.proof.created || !/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z?$/.test(marker.proof.created) || isNaN(Date.parse(marker.proof.created))) {
    errors.push("Proof missing or invalid created timestamp (must be ISO 8601)");
  }
  if (marker.subject && marker.proof.verificationMethod !== marker.subject) {
    errors.push("Proof verificationMethod does not match marker subject \u2014 possible attribution forgery");
    return { valid: false, errors };
  }
  const didAlg = algorithmFromDid(marker.proof.verificationMethod);
  if (didAlg !== alg) {
    errors.push(`Algorithm mismatch: proof type indicates ${alg} but DID uses ${didAlg}`);
    return { valid: false, errors };
  }
  try {
    const publicKey = publicKeyFromDid(marker.proof.verificationMethod);
    const { proof: _proof, id: _id, ...rest } = marker;
    const canonical = canonicalize(rest);
    const data = new TextEncoder().encode(DOMAIN_PREFIX + canonical);
    const signature = new Uint8Array(Buffer.from(marker.proof.proofValue, "base64"));
    let valid;
    if (alg === "P-256") {
      valid = verifyP256(data, signature, publicKey);
    } else {
      valid = verify2(data, signature, publicKey);
    }
    if (!valid) {
      errors.push("Signature verification failed");
    }
  } catch (e) {
    errors.push(
      options?.verbose ? `Signature verification error: ${e.message}` : "Verification failed"
    );
  }
  if (!options?.verbose) {
    const genericErrors = [];
    for (const err of errors) {
      if (err === "Signature verification failed" || err.startsWith("Signature verification error:")) {
        if (!genericErrors.includes("Verification failed")) {
          genericErrors.push("Verification failed");
        }
      } else {
        genericErrors.push(err);
      }
    }
    return { valid: genericErrors.length === 0, errors: genericErrors };
  }
  return { valid: errors.length === 0, errors };
}
async function signMarkerWithSigner(marker, signer) {
  try {
    const did = signer.did();
    const now = (/* @__PURE__ */ new Date()).toISOString();
    const { proof: _proof, id: _id, ...rest } = marker;
    const canonical = canonicalize(rest);
    const data = new TextEncoder().encode(DOMAIN_PREFIX + canonical);
    const signature = await signer.sign(data);
    const proofValue = Buffer.from(signature).toString("base64");
    const proof = {
      type: proofTypeForAlgorithm(signer.algorithm),
      created: now,
      verificationMethod: did,
      proofValue
    };
    return { ...marker, proof };
  } catch (e) {
    if (e instanceof SigningError) throw e;
    throw new SigningError(`Failed to sign marker: ${e.message}`);
  }
}
async function verifyMarkerMultiAlg(marker) {
  return verifyMarker(marker);
}
function verifyTrustEnhancers(marker) {
  const errors = [];
  if (!marker.trustEnhancers) return { valid: true, errors };
  const te = marker.trustEnhancers;
  const MAX_TRUST_ENHANCER_ITEMS = 100;
  for (const arrayName of ["timestamps", "witnesses", "identityClaims"]) {
    const arr = te[arrayName];
    if (Array.isArray(arr) && arr.length > MAX_TRUST_ENHANCER_ITEMS) {
      errors.push(`Trust enhancer ${arrayName} exceeds maximum of ${MAX_TRUST_ENHANCER_ITEMS} items (got ${arr.length})`);
      return { valid: false, errors };
    }
  }
  if (te.timestamps) {
    for (let i = 0; i < te.timestamps.length; i++) {
      const ts = te.timestamps[i];
      if (!ts.tsaUrl) errors.push(`Trust enhancer timestamps[${i}] missing tsaUrl`);
      if (!ts.hash) errors.push(`Trust enhancer timestamps[${i}] missing hash`);
      if (!ts.timestamp) errors.push(`Trust enhancer timestamps[${i}] missing timestamp`);
      if (ts.timestamp && isNaN(Date.parse(ts.timestamp))) errors.push(`Trust enhancer timestamps[${i}] invalid timestamp: ${ts.timestamp}`);
      if (!ts.receipt) errors.push(`Trust enhancer timestamps[${i}] missing receipt`);
    }
  }
  if (te.witnesses) {
    for (let i = 0; i < te.witnesses.length; i++) {
      const w = te.witnesses[i];
      if (!w.witnessDid) errors.push(`Trust enhancer witnesses[${i}] missing witnessDid`);
      if (!w.attestation) errors.push(`Trust enhancer witnesses[${i}] missing attestation`);
      if (!w.signature) errors.push(`Trust enhancer witnesses[${i}] missing signature`);
      if (!w.signatureType) errors.push(`Trust enhancer witnesses[${i}] missing signatureType`);
      if (!w.timestamp) errors.push(`Trust enhancer witnesses[${i}] missing timestamp`);
      if (w.timestamp && isNaN(Date.parse(w.timestamp))) errors.push(`Trust enhancer witnesses[${i}] invalid timestamp: ${w.timestamp}`);
    }
  }
  if (te.identityClaims) {
    for (let i = 0; i < te.identityClaims.length; i++) {
      const c = te.identityClaims[i];
      if (!c.scheme) errors.push(`Trust enhancer identityClaims[${i}] missing scheme`);
      if (!c.value) errors.push(`Trust enhancer identityClaims[${i}] missing value`);
      if (!c.issuedAt) errors.push(`Trust enhancer identityClaims[${i}] missing issuedAt`);
      if (c.issuedAt && isNaN(Date.parse(c.issuedAt))) errors.push(`Trust enhancer identityClaims[${i}] invalid issuedAt: ${c.issuedAt}`);
      if (c.expiresAt && isNaN(Date.parse(c.expiresAt))) errors.push(`Trust enhancer identityClaims[${i}] invalid expiresAt: ${c.expiresAt}`);
    }
  }
  return { valid: errors.length === 0, errors };
}

// src/ceremony.ts
var TRANSITIONS = {
  ["alive" /* Alive */]: ["intent" /* Intent */, "final" /* Final */],
  // Final for emergency
  ["intent" /* Intent */]: ["snapshot" /* Snapshot */, "final" /* Final */],
  // Final for emergency escape from INTENT
  ["snapshot" /* Snapshot */]: ["open" /* Open */, "final" /* Final */],
  // Final for unilateral
  ["open" /* Open */]: ["contested" /* Contested */, "final" /* Final */],
  ["contested" /* Contested */]: ["final" /* Final */],
  ["final" /* Final */]: ["departed" /* Departed */],
  ["departed" /* Departed */]: []
};
function getValidTransitions(state) {
  return [...TRANSITIONS[state] ?? []];
}
var CeremonyStateMachine = class {
  _state = "alive" /* Alive */;
  intent;
  marker;
  exitType;
  /** Current ceremony state (read-only). */
  get state() {
    return this._state;
  }
  transition(to) {
    if (!TRANSITIONS[this._state].includes(to)) {
      throw new CeremonyError(
        this._state,
        to,
        TRANSITIONS[this._state].map((s) => s)
      );
    }
    this._state = to;
  }
  /**
   * Declare intent to exit. Moves ALIVE → INTENT (or stays ALIVE for emergency path).
   *
   * Accepts either a {@link Signer} (preferred) or raw Ed25519 key pair (deprecated).
   *
   * @param subject - The DID of the departing entity.
   * @param origin - The URI of the system being exited.
   * @param exitType - The type of exit.
   * @param signerOrPrivateKey - A Signer instance, or Ed25519 private key (deprecated).
   * @param publicKey - The Ed25519 public key (only when using raw keys).
   * @returns A signed {@link ExitIntent}.
   * @throws {CeremonyError} If the transition is invalid.
   */
  async declareIntent(subject, origin, exitType, signerOrPrivateKey, publicKey) {
    this.exitType = exitType;
    if (exitType === "emergency" /* Emergency */) {
      this.intent = await this.buildIntent(subject, origin, exitType, signerOrPrivateKey, publicKey);
      return this.intent;
    }
    this.transition("intent" /* Intent */);
    this.intent = await this.buildIntent(subject, origin, exitType, signerOrPrivateKey, publicKey);
    return this.intent;
  }
  async buildIntent(subject, origin, exitType, signerOrPrivateKey, publicKey) {
    const timestamp = (/* @__PURE__ */ new Date()).toISOString();
    const intentData = { subject, origin, timestamp, exitType };
    const data = new TextEncoder().encode("exit-intent-v1:" + canonicalize(intentData));
    let sig;
    let did;
    let proofType;
    if (signerOrPrivateKey instanceof Uint8Array) {
      sig = sign2(data, signerOrPrivateKey);
      did = didFromPublicKey(publicKey);
      proofType = "Ed25519Signature2020";
    } else {
      sig = await signerOrPrivateKey.sign(data);
      did = signerOrPrivateKey.did();
      proofType = proofTypeForAlgorithm(signerOrPrivateKey.algorithm);
    }
    return {
      subject,
      origin,
      timestamp,
      exitType,
      proof: {
        type: proofType,
        created: timestamp,
        verificationMethod: did,
        proofValue: Buffer.from(sig).toString("base64")
      }
    };
  }
  /**
   * Advance through snapshot. Moves INTENT → SNAPSHOT.
   *
   * @throws {CeremonyError} If the current state is not INTENT.
   */
  snapshot() {
    this.transition("snapshot" /* Snapshot */);
  }
  /**
   * Open challenge window. Moves SNAPSHOT → OPEN.
   *
   * @throws {CeremonyError} If the current state is not SNAPSHOT.
   */
  openChallenge() {
    this.transition("open" /* Open */);
  }
  /**
   * Mark as contested. Moves OPEN → CONTESTED.
   *
   * @throws {CeremonyError} If the current state is not OPEN.
   */
  contest() {
    this.transition("contested" /* Contested */);
  }
  /**
   * Sign the marker and finalize. Moves to FINAL.
   *
   * Accepts either a {@link Signer} (preferred) or raw Ed25519 key pair (deprecated).
   *
   * @param marker - The EXIT marker to sign.
   * @param signerOrPrivateKey - A Signer instance, or Ed25519 private key (deprecated).
   * @param publicKey - The Ed25519 public key (only when using raw keys).
   * @returns The signed EXIT marker.
   * @throws {CeremonyError} If the transition to FINAL is invalid.
   */
  async signMarker(marker, signerOrPrivateKey, publicKey) {
    this.transition("final" /* Final */);
    if (signerOrPrivateKey instanceof Uint8Array) {
      this.marker = signMarker(marker, signerOrPrivateKey, publicKey);
    } else {
      const signer = signerOrPrivateKey;
      const { proof: _proof, id: _id, ...rest } = marker;
      const canonical = canonicalize(rest);
      const data = new TextEncoder().encode("exit-marker-v1.2:" + canonical);
      const sig = await signer.sign(data);
      this.marker = {
        ...marker,
        proof: {
          type: proofTypeForAlgorithm(signer.algorithm),
          created: (/* @__PURE__ */ new Date()).toISOString(),
          verificationMethod: signer.did(),
          proofValue: Buffer.from(sig).toString("base64")
        }
      };
    }
    return this.marker;
  }
  /**
   * Add a witness co-signature. Stays in FINAL.
   *
   * Accepts either a {@link Signer} (preferred) or raw Ed25519 key pair (deprecated).
   *
   * @param signerOrWitnessKey - A Signer instance, or witness Ed25519 private key (deprecated).
   * @param witnessPublicKey - The witness's Ed25519 public key (only when using raw keys).
   * @returns A {@link DataIntegrityProof} containing the witness's co-signature.
   * @throws {CeremonyError} If the current state is not FINAL or no marker has been signed.
   */
  async witness(signerOrWitnessKey, witnessPublicKey) {
    if (this.state !== "final" /* Final */) {
      throw new CeremonyError(this.state, "witness (requires final)", getValidTransitions(this.state).map((s) => s));
    }
    if (!this.marker) throw new CeremonyError(this.state, "witness", ["sign a marker first"]);
    const { proof: _proof, ...rest } = this.marker;
    const data = new TextEncoder().encode("exit-witness-v1:" + canonicalize(rest));
    if (signerOrWitnessKey instanceof Uint8Array) {
      const sig = sign2(data, signerOrWitnessKey);
      return {
        type: "Ed25519Signature2020",
        created: (/* @__PURE__ */ new Date()).toISOString(),
        verificationMethod: didFromPublicKey(witnessPublicKey),
        proofValue: Buffer.from(sig).toString("base64")
      };
    } else {
      const signer = signerOrWitnessKey;
      const sig = await signer.sign(data);
      return {
        type: proofTypeForAlgorithm(signer.algorithm),
        created: (/* @__PURE__ */ new Date()).toISOString(),
        verificationMethod: signer.did(),
        proofValue: Buffer.from(sig).toString("base64")
      };
    }
  }
  /**
   * Depart. Moves FINAL → DEPARTED. Terminal — no further transitions possible.
   *
   * @returns The finalized EXIT marker.
   * @throws {CeremonyError} If the current state is not FINAL or no marker has been signed.
   */
  depart() {
    this.transition("departed" /* Departed */);
    if (!this.marker) throw new CeremonyError(this.state, "depart", ["sign a marker first"]);
    return this.marker;
  }
};

// src/countersign.ts
var COUNTER_DOMAIN_PREFIX = "exit-counter-v1.2:";
function addCounterSignature(marker, privateKey, publicKey, opts) {
  if (!marker.proof?.proofValue) {
    throw new SigningError("Cannot counter-sign an unsigned marker");
  }
  try {
    const did = didFromPublicKey(publicKey);
    const now = (/* @__PURE__ */ new Date()).toISOString();
    const { proof: _proof, id: _id, ...rest } = marker;
    const restWithoutAcks = { ...rest };
    if (restWithoutAcks.dispute) {
      const { counterpartyAcks: _acks, ...disputeRest } = restWithoutAcks.dispute;
      if (Object.keys(disputeRest).length === 0) {
        delete restWithoutAcks.dispute;
      } else {
        restWithoutAcks.dispute = disputeRest;
      }
    }
    const withPrimary = { ...restWithoutAcks, primaryProofValue: marker.proof.proofValue };
    const canonical = canonicalize(withPrimary);
    const data = new TextEncoder().encode(COUNTER_DOMAIN_PREFIX + canonical);
    const signature = sign2(data, privateKey);
    const proofValue = Buffer.from(signature).toString("base64");
    const ack = {
      type: "Ed25519Signature2020",
      created: now,
      verificationMethod: did,
      proofValue
    };
    const existingDispute = marker.dispute ?? {};
    const existingAcks = existingDispute.counterpartyAcks ?? [];
    const updatedDispute = {
      ...existingDispute,
      counterpartyAcks: [...existingAcks, ack]
    };
    return { ...marker, dispute: updatedDispute };
  } catch (e) {
    if (e instanceof SigningError) throw e;
    throw new SigningError(`Failed to add counter-signature: ${e.message}`);
  }
}
function addWitness(marker, witness) {
  const existing = marker.trustEnhancers?.witnesses ?? [];
  return {
    ...marker,
    trustEnhancers: {
      ...marker.trustEnhancers,
      witnesses: [...existing, witness]
    }
  };
}
function verifyCounterSignature(marker, publicKey, ackIndex = 0) {
  const errors = [];
  const acks = marker.dispute?.counterpartyAcks;
  if (!acks || acks.length === 0) {
    return { valid: false, errors: ["No counterpartyAcks found on marker"] };
  }
  if (ackIndex < 0 || ackIndex >= acks.length) {
    return { valid: false, errors: [`ackIndex ${ackIndex} out of range (0..${acks.length - 1})`] };
  }
  const ack = acks[ackIndex];
  if (!ack.proofValue || !ack.verificationMethod) {
    return { valid: false, errors: ["Incomplete counter-signature: missing proofValue or verificationMethod"] };
  }
  try {
    const { proof: _proof, id: _id, ...rest } = marker;
    const restWithoutAcks = { ...rest };
    if (restWithoutAcks.dispute) {
      const { counterpartyAcks: _acks, ...disputeRest } = restWithoutAcks.dispute;
      if (Object.keys(disputeRest).length === 0) {
        delete restWithoutAcks.dispute;
      } else {
        restWithoutAcks.dispute = disputeRest;
      }
    }
    const primaryProofValue = marker.proof?.proofValue ?? "";
    const withPrimary = { ...restWithoutAcks, primaryProofValue };
    const canonical = canonicalize(withPrimary);
    const data = new TextEncoder().encode(COUNTER_DOMAIN_PREFIX + canonical);
    const signature = new Uint8Array(Buffer.from(ack.proofValue, "base64"));
    const alg = algorithmFromDid(ack.verificationMethod);
    let valid;
    if (alg === "P-256") {
      valid = verifyP256(data, signature, publicKey);
    } else {
      valid = verify2(data, signature, publicKey);
    }
    if (!valid) {
      errors.push("Counter-signature verification failed");
    }
    return { valid: errors.length === 0, errors };
  } catch (e) {
    return { valid: false, errors: [`Verification error: ${e.message}`] };
  }
}
function deriveStatusConfirmation(marker) {
  const hasProof = !!marker.proof?.proofValue;
  const hasAcks = (marker.dispute?.counterpartyAcks?.length ?? 0) > 0;
  const hasWitnesses = (marker.trustEnhancers?.witnesses?.length ?? 0) > 0;
  if (hasWitnesses) return "witnessed" /* Witnessed */;
  if (hasProof && hasAcks) return "mutual" /* Mutual */;
  if (!hasProof && hasAcks) return "origin_only" /* OriginOnly */;
  return "self_only" /* SelfOnly */;
}

// src/convenience.ts
function generateIdentity() {
  const { publicKey, privateKey } = generateKeyPair();
  const did = didFromPublicKey(publicKey);
  return { did, publicKey, privateKey };
}
async function quickExit(origin, opts) {
  const alg = opts?.algorithm ?? "ed25519";
  let identity;
  if (alg === "p256") {
    const { publicKey, privateKey } = generateP256KeyPair();
    const did = didFromP256PublicKey(publicKey);
    identity = { did, publicKey, privateKey };
  } else {
    identity = generateIdentity();
  }
  const exitType = opts?.exitType ?? "voluntary" /* Voluntary */;
  const marker = createMarker({
    subject: identity.did,
    origin,
    exitType,
    status: opts?.status,
    emergencyJustification: opts?.emergencyJustification
  });
  let signed;
  if (alg === "p256") {
    const signer = new P256Signer(identity.privateKey, identity.publicKey);
    signed = await signMarkerWithSigner(marker, signer);
  } else {
    signed = signMarker(marker, identity.privateKey, identity.publicKey);
  }
  return { marker: signed, identity };
}
async function quickExitP256(origin, opts) {
  return quickExit(origin, { ...opts, algorithm: "p256" });
}
function quickVerify(markerJson) {
  const marker = fromJSON(markerJson);
  return verifyMarker(marker);
}
var MAX_JSON_SIZE = 1048576;
function fromJSON(json) {
  if (json.length > MAX_JSON_SIZE) {
    throw new ValidationError([`JSON input too large: ${json.length} chars (max ${MAX_JSON_SIZE})`]);
  }
  let parsed;
  try {
    parsed = JSON.parse(json);
  } catch (e) {
    throw new ValidationError([`Invalid JSON: ${e.message}`]);
  }
  const result = validateMarker(parsed);
  if (!result.valid) {
    throw new ValidationError(result.errors);
  }
  return parsed;
}
function quickCounterSign(marker, opts) {
  let identity;
  if (opts?.privateKey && opts?.publicKey) {
    const did = didFromPublicKey(opts.publicKey);
    identity = { did, publicKey: opts.publicKey, privateKey: opts.privateKey };
  } else {
    identity = generateIdentity();
  }
  const { privateKey: _pk, publicKey: _pub, ...counterSignOpts } = opts ?? {};
  const counterSigned = addCounterSignature(marker, identity.privateKey, identity.publicKey, counterSignOpts);
  return { marker: counterSigned, identity };
}
function toJSON(marker) {
  return JSON.stringify(marker, null, 2);
}

// src/context.ts
var EXIT_CONTEXT = {
  "@context": {
    // Base vocabularies
    exit: "https://cellar-door.dev/exit/v1#",
    xsd: "http://www.w3.org/2001/XMLSchema#",
    dc: "http://purl.org/dc/terms/",
    // 7 Core fields
    id: "@id",
    subject: { "@id": "exit:subject", "@type": "@id" },
    origin: { "@id": "exit:origin", "@type": "@id" },
    timestamp: { "@id": "exit:timestamp", "@type": "xsd:dateTime" },
    exitType: { "@id": "exit:exitType", "@type": "xsd:string" },
    status: { "@id": "exit:status", "@type": "xsd:string" },
    proof: { "@id": "exit:proof", "@type": "@id" },
    // Proof fields
    type: "@type",
    created: { "@id": "exit:created", "@type": "xsd:dateTime" },
    verificationMethod: { "@id": "exit:verificationMethod", "@type": "@id" },
    proofValue: { "@id": "exit:proofValue", "@type": "xsd:string" },
    // Module namespaces
    lineage: { "@id": "exit:lineage", "@type": "@id" },
    reputation: { "@id": "exit:reputation", "@type": "@id" },
    originAttestation: { "@id": "exit:originAttestation", "@type": "@id" },
    assets: { "@id": "exit:assets", "@type": "@id" },
    continuity: { "@id": "exit:continuity", "@type": "@id" },
    dispute: { "@id": "exit:dispute", "@type": "@id" },
    // Legacy module slots (Sprint 1 compat)
    stateSnapshot: { "@id": "exit:stateSnapshot", "@type": "@id" },
    economic: { "@id": "exit:economic", "@type": "@id" },
    metadata: { "@id": "exit:metadata", "@type": "@id" },
    crossDomain: { "@id": "exit:crossDomain", "@type": "@id" },
    // Ceremony terms
    ceremonyState: { "@id": "exit:ceremonyState", "@type": "xsd:string" },
    intent: { "@id": "exit:intent", "@type": "@id" },
    challengeWindow: { "@id": "exit:challengeWindow", "@type": "@id" },
    witness: { "@id": "exit:witness", "@type": "@id" },
    // Lineage module terms
    predecessor: { "@id": "exit:predecessor", "@type": "@id" },
    successor: { "@id": "exit:successor", "@type": "@id" },
    lineageChain: { "@id": "exit:lineageChain", "@type": "@id", "@container": "@list" },
    continuityProof: { "@id": "exit:continuityProof", "@type": "@id" },
    // Reputation module terms
    metrics: { "@id": "exit:metrics", "@type": "@id" },
    endorsements: { "@id": "exit:endorsements", "@type": "@id", "@container": "@list" },
    tenure: { "@id": "exit:tenure", "@type": "xsd:duration" },
    // Asset module terms
    assetManifest: { "@id": "exit:assetManifest", "@type": "@id", "@container": "@list" },
    // Continuity module terms
    stateHash: { "@id": "exit:stateHash", "@type": "xsd:string" },
    memoryRef: { "@id": "exit:memoryRef", "@type": "@id" },
    configHash: { "@id": "exit:configHash", "@type": "xsd:string" }
  }
};
function getContext() {
  return EXIT_CONTEXT;
}

// src/vc.ts
var VC_CONTEXT = "https://www.w3.org/ns/credentials/v2";
function wrapAsVC(marker) {
  const vc = {
    "@context": [VC_CONTEXT, EXIT_CONTEXT_V1],
    type: ["VerifiableCredential", "ExitCredential"],
    id: `urn:vc:exit:${marker.id}`,
    issuer: marker.subject,
    issuanceDate: marker.timestamp,
    credentialSubject: marker
  };
  if (marker.proof && marker.proof.proofValue) {
    vc.proof = marker.proof;
  }
  return vc;
}
function unwrapVC(vc) {
  if (!isVC(vc)) {
    throw new Error("Not a valid ExitVerifiableCredential");
  }
  return vc.credentialSubject;
}
function isVC(obj) {
  if (!obj || typeof obj !== "object") return false;
  const o = obj;
  if (!Array.isArray(o["@context"])) return false;
  const ctx = o["@context"];
  if (ctx[0] !== VC_CONTEXT) return false;
  if (!Array.isArray(o.type)) return false;
  const types = o.type;
  if (!types.includes("VerifiableCredential") || !types.includes("ExitCredential")) return false;
  if (!o.credentialSubject || typeof o.credentialSubject !== "object") return false;
  return true;
}

// src/modules/lineage.ts
function createLineageModule(opts) {
  return {
    predecessor: opts.predecessor,
    successor: opts.successor,
    lineageChain: opts.lineageChain,
    continuityProof: opts.continuityProof
  };
}
function verifyLineageChain(markers) {
  const errors = [];
  if (markers.length < 2) {
    return { valid: true, errors: [] };
  }
  for (let i = 1; i < markers.length; i++) {
    const prev = markers[i - 1];
    const curr = markers[i];
    if (!curr.lineage) {
      errors.push(`Marker ${i} (${curr.id}) missing lineage module`);
      continue;
    }
    if (curr.lineage.predecessor !== prev.subject) {
      errors.push(
        `Marker ${i} predecessor mismatch: expected ${prev.subject}, got ${curr.lineage.predecessor}`
      );
    }
    if (curr.lineage.lineageChain && !curr.lineage.lineageChain.includes(prev.subject)) {
      errors.push(`Marker ${i} lineageChain does not include predecessor ${prev.subject}`);
    }
  }
  return { valid: errors.length === 0, errors };
}
function bindSuccessor(marker, successorDid, privateKey, publicKey) {
  const rotationData = canonicalize({
    predecessor: marker.subject,
    successor: successorDid,
    exitMarkerId: marker.id
  });
  const sig = sign2(new TextEncoder().encode(rotationData), privateKey);
  const proofValue = Buffer.from(sig).toString("base64");
  const lineage = {
    ...marker.lineage,
    successor: successorDid,
    continuityProof: {
      type: "key_rotation_binding" /* KeyRotationBinding */,
      value: proofValue,
      verificationMethod: didFromPublicKey(publicKey)
    }
  };
  return { ...marker, lineage };
}

// src/modules/reputation.ts
function createReputationModule(opts) {
  return {
    moduleType: "reputation",
    metrics: opts.metrics,
    endorsements: opts.endorsements ?? [],
    tenure: opts.tenure
  };
}
function signEndorsement(text, privateKey, publicKey) {
  const timestamp = (/* @__PURE__ */ new Date()).toISOString();
  const data = canonicalize({ text, timestamp });
  const sig = sign2(new TextEncoder().encode(data), privateKey);
  return {
    text,
    signer: didFromPublicKey(publicKey),
    signature: Buffer.from(sig).toString("base64"),
    timestamp
  };
}

// src/modules/origin-attestation.ts
function createOriginAttestation(opts) {
  return {
    moduleType: "originAttestation",
    originStatus: opts.originStatus,
    originStatement: opts.originStatement,
    originSignature: opts.originSignature ?? "",
    originDid: opts.originDid ?? "",
    timestamp: opts.timestamp ?? (/* @__PURE__ */ new Date()).toISOString()
  };
}
function signAttestation(marker, originPrivateKey, originPublicKey, status, statement) {
  const timestamp = (/* @__PURE__ */ new Date()).toISOString();
  const data = canonicalize({
    markerId: marker.id,
    subject: marker.subject,
    originStatus: status,
    originStatement: statement,
    timestamp
  });
  const sig = sign2(new TextEncoder().encode(data), originPrivateKey);
  return {
    moduleType: "originAttestation",
    originStatus: status,
    originStatement: statement,
    originSignature: Buffer.from(sig).toString("base64"),
    originDid: didFromPublicKey(originPublicKey),
    timestamp
  };
}

// src/modules/assets.ts
import { sha256 } from "@noble/hashes/sha256";
function hashString(s) {
  const h = sha256(new TextEncoder().encode(s));
  return Array.from(h).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function createAssetManifest(assetOpts) {
  const assets = assetOpts.map((a) => ({
    id: a.id,
    type: a.type,
    description: a.description,
    hash: a.hash ?? (a.data ? hashString(a.data) : ""),
    portable: a.portable ?? true
  }));
  return { moduleType: "assets", assets };
}

// src/modules/continuity.ts
function createContinuityModule(opts) {
  return {
    moduleType: "continuity",
    stateHash: opts.stateHash,
    memoryRef: opts.memoryRef,
    configHash: opts.configHash,
    continuityProofType: opts.continuityProofType ?? "snapshot_hash"
  };
}

// src/modules/dispute.ts
function createDisputeModule(opts) {
  return {
    moduleType: "dispute",
    disputeType: opts.disputeType,
    filedBy: opts.filedBy,
    description: opts.description,
    evidence: opts.evidence ?? [],
    resolutionStatus: opts.resolutionStatus ?? "open",
    filedAt: opts.filedAt ?? (/* @__PURE__ */ new Date()).toISOString()
  };
}

// src/modules/trust.ts
import { sha256 as sha2562 } from "@noble/hashes/sha256";
function computeStatusConfirmation(marker) {
  const hasOriginAttestation = marker.dispute?.originStatus !== void 0;
  const hasWitness = marker.dispute?.counterpartyAcks !== void 0 && marker.dispute.counterpartyAcks.length > 0;
  if (hasWitness) {
    return "witnessed" /* Witnessed */;
  }
  if (hasOriginAttestation) {
    const originStatus = marker.dispute.originStatus;
    if (originStatus === marker.status) {
      return "mutual" /* Mutual */;
    }
    if (marker.selfAttested) {
      return "disputed_by_origin" /* DisputedByOrigin */;
    }
    return "disputed_by_subject" /* DisputedBySubject */;
  }
  if (marker.selfAttested) {
    return "self_only" /* SelfOnly */;
  }
  return "origin_only" /* OriginOnly */;
}
function createTenureAttestation(duration, startDate, privateKey, publicKey, attestedBy = "subject") {
  const did = didFromPublicKey(publicKey);
  const data = canonicalize({ duration, startDate, attestedBy, attesterDid: did });
  const sig = sign2(new TextEncoder().encode("tenure-attestation-v1:" + data), privateKey);
  return {
    duration,
    startDate,
    attestedBy,
    signature: Buffer.from(sig).toString("base64"),
    attesterDid: did
  };
}
function verifyTenureAttestation(attestation) {
  try {
    const publicKey = publicKeyFromDid(attestation.attesterDid);
    const data = canonicalize({
      duration: attestation.duration,
      startDate: attestation.startDate,
      attestedBy: attestation.attestedBy,
      attesterDid: attestation.attesterDid
    });
    const sig = new Uint8Array(Buffer.from(attestation.signature, "base64"));
    return verify2(new TextEncoder().encode("tenure-attestation-v1:" + data), sig, publicKey);
  } catch {
    return false;
  }
}
function parseDurationToDays(duration) {
  const match = duration.match(/^P(?:(\d+)Y)?(?:(\d+)M)?(?:(\d+)D)?$/);
  if (!match) return 0;
  const years = parseInt(match[1] || "0", 10);
  const months = parseInt(match[2] || "0", 10);
  const days = parseInt(match[3] || "0", 10);
  return years * 365 + months * 30 + days;
}
function createCommitment(intent, revealDelayMs, privateKey, publicKey) {
  const did = didFromPublicKey(publicKey);
  const now = /* @__PURE__ */ new Date();
  const revealAfter = new Date(now.getTime() + revealDelayMs);
  const intentCanonical = canonicalize(intent);
  const hash = sha2562(new TextEncoder().encode(intentCanonical));
  const commitmentHash = Array.from(hash).map((b) => b.toString(16).padStart(2, "0")).join("");
  const committedAt = now.toISOString();
  const revealAfterStr = revealAfter.toISOString();
  const commitData = canonicalize({
    commitmentHash,
    committedAt,
    revealAfter: revealAfterStr,
    committerDid: did
  });
  const sig = sign2(new TextEncoder().encode("exit-commitment-v1:" + commitData), privateKey);
  return {
    commitmentHash,
    committedAt,
    revealAfter: revealAfterStr,
    committerDid: did,
    proof: {
      type: "Ed25519Signature2020",
      created: committedAt,
      verificationMethod: did,
      proofValue: Buffer.from(sig).toString("base64")
    }
  };
}
function verifyCommitment(commitment, revealedIntent) {
  const errors = [];
  const intentCanonical = canonicalize(revealedIntent);
  const hash = sha2562(new TextEncoder().encode(intentCanonical));
  const expectedHash = Array.from(hash).map((b) => b.toString(16).padStart(2, "0")).join("");
  if (commitment.commitmentHash !== expectedHash) {
    errors.push("Commitment hash does not match revealed intent");
  }
  if (commitment.committerDid !== revealedIntent.subject) {
    errors.push("Commitment committer does not match intent subject");
  }
  try {
    const publicKey = publicKeyFromDid(commitment.proof.verificationMethod);
    const commitData = canonicalize({
      commitmentHash: commitment.commitmentHash,
      committedAt: commitment.committedAt,
      revealAfter: commitment.revealAfter,
      committerDid: commitment.committerDid
    });
    const sig = new Uint8Array(Buffer.from(commitment.proof.proofValue, "base64"));
    if (!verify2(new TextEncoder().encode("exit-commitment-v1:" + commitData), sig, publicKey)) {
      errors.push("Commitment signature verification failed");
    }
  } catch (e) {
    errors.push(`Commitment signature error: ${e.message}`);
  }
  return { valid: errors.length === 0, errors };
}
function isRevealWindowOpen(commitment) {
  return /* @__PURE__ */ new Date() >= new Date(commitment.revealAfter);
}
function extractConfidenceFactors(marker, tenure, originTenure, hasCommitReveal = false) {
  const statusConfirmation = computeStatusConfirmation(marker);
  const tenureDays = tenure ? parseDurationToDays(tenure.duration) : 0;
  const tenureMutuallyAttested = tenure !== void 0 && originTenure !== void 0 && tenure.attestedBy === "subject" && originTenure.attestedBy === "origin";
  const lineageDepth = marker.lineage?.lineageChain?.length ?? 0;
  return {
    statusConfirmation,
    tenureDays,
    tenureMutuallyAttested,
    lineageDepth,
    hasCommitReveal
  };
}
function computeConfidence(factors) {
  const statusWeights = {
    ["self_only" /* SelfOnly */]: 0.05,
    ["origin_only" /* OriginOnly */]: 0.2,
    ["mutual" /* Mutual */]: 0.4,
    ["witnessed" /* Witnessed */]: 0.4,
    ["disputed_by_origin" /* DisputedByOrigin */]: 0,
    ["disputed_by_subject" /* DisputedBySubject */]: 0.1
  };
  let statusScore = statusWeights[factors.statusConfirmation] ?? 0;
  let tenureScore = 0;
  if (factors.tenureDays > 0) {
    tenureScore = Math.min(0.3, Math.log2(factors.tenureDays + 1) / Math.log2(731) * 0.3);
    if (factors.tenureMutuallyAttested) {
      tenureScore *= 1;
    } else {
      tenureScore *= 0.5;
    }
  }
  let lineageScore = 0;
  if (factors.lineageDepth > 0) {
    lineageScore = Math.min(0.15, Math.log2(factors.lineageDepth + 1) / Math.log2(11) * 0.15);
  }
  const commitRevealScore = factors.hasCommitReveal ? 0.15 : 0;
  let score = statusScore + tenureScore + lineageScore + commitRevealScore;
  score = Math.min(1, Math.max(0, score));
  score = Math.round(score * 1e3) / 1e3;
  let level;
  if (score < 0.1) level = "none";
  else if (score < 0.3) level = "low";
  else if (score < 0.5) level = "moderate";
  else if (score < 0.75) level = "high";
  else level = "very_high";
  return { score, level, factors };
}

// src/storage.ts
import { readFileSync, writeFileSync, readdirSync, mkdirSync, existsSync, unlinkSync } from "fs";
import { join, resolve } from "path";
function validateDir(dir) {
  const resolved = resolve(dir);
  const parts = dir.split(/[/\\]/);
  if (parts.includes("..")) {
    throw new Error(`Unsafe storage directory: path contains '..' traversal: ${dir}`);
  }
  return resolved;
}
function saveMarker(marker, dir) {
  dir = validateDir(dir);
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
  const safeId = marker.id.replace(/[^a-zA-Z0-9_-]/g, "_");
  const filePath = join(dir, `${safeId}.json`);
  writeFileSync(filePath, JSON.stringify(marker, null, 2), "utf-8");
  return filePath;
}
function loadMarker(id, dir) {
  dir = validateDir(dir);
  const safeId = id.replace(/[^a-zA-Z0-9_-]/g, "_");
  const filePath = join(dir, `${safeId}.json`);
  const raw = readFileSync(filePath, "utf-8");
  const marker = JSON.parse(raw);
  const result = validateMarker(marker);
  if (!result.valid) {
    throw new Error(`Invalid marker ${id}: ${result.errors.join(", ")}`);
  }
  return marker;
}
function listMarkers(dir) {
  dir = validateDir(dir);
  if (!existsSync(dir)) return [];
  return readdirSync(dir).filter((f) => f.endsWith(".json")).map((f) => f.replace(/\.json$/, ""));
}
function deleteMarker(id, dir) {
  dir = validateDir(dir);
  const safeId = id.replace(/[^a-zA-Z0-9_-]/g, "_");
  const filePath = join(dir, `${safeId}.json`);
  if (!existsSync(filePath)) return false;
  unlinkSync(filePath);
  return true;
}
function exportMarker(marker) {
  return JSON.stringify(marker, null, 2);
}
var MAX_JSON_SIZE2 = 1048576;
function saveAmendment(amendment, dir) {
  dir = validateDir(dir);
  const subdir = join(dir, "amendments", amendment.originalMarkerId.replace(/[^a-zA-Z0-9_-]/g, "_"));
  if (!existsSync(subdir)) mkdirSync(subdir, { recursive: true });
  const safeId = amendment.amendmentId.replace(/[^a-zA-Z0-9_-]/g, "_");
  const filePath = join(subdir, `${safeId}.json`);
  writeFileSync(filePath, JSON.stringify(amendment, null, 2), "utf-8");
  return filePath;
}
function loadAmendments(originalMarkerId, dir) {
  dir = validateDir(dir);
  const subdir = join(dir, "amendments", originalMarkerId.replace(/[^a-zA-Z0-9_-]/g, "_"));
  if (!existsSync(subdir)) return [];
  return readdirSync(subdir).filter((f) => f.endsWith(".json")).map((f) => JSON.parse(readFileSync(join(subdir, f), "utf-8")));
}
function saveRevocation(revocation, dir) {
  dir = validateDir(dir);
  const subdir = join(dir, "revocations", revocation.targetMarkerId.replace(/[^a-zA-Z0-9_-]/g, "_"));
  if (!existsSync(subdir)) mkdirSync(subdir, { recursive: true });
  const safeId = revocation.revocationId.replace(/[^a-zA-Z0-9_-]/g, "_");
  const filePath = join(subdir, `${safeId}.json`);
  writeFileSync(filePath, JSON.stringify(revocation, null, 2), "utf-8");
  return filePath;
}
function loadRevocations(targetMarkerId, dir) {
  dir = validateDir(dir);
  const subdir = join(dir, "revocations", targetMarkerId.replace(/[^a-zA-Z0-9_-]/g, "_"));
  if (!existsSync(subdir)) return [];
  return readdirSync(subdir).filter((f) => f.endsWith(".json")).map((f) => JSON.parse(readFileSync(join(subdir, f), "utf-8")));
}
function importMarker(json) {
  if (json.length > MAX_JSON_SIZE2) {
    throw new Error(`JSON input too large: ${json.length} chars (max ${MAX_JSON_SIZE2})`);
  }
  const marker = JSON.parse(json);
  const result = validateMarker(marker);
  if (!result.valid) {
    throw new Error(`Invalid marker: ${result.errors.join(", ")}`);
  }
  return marker;
}

// src/chain.ts
import { readFileSync as readFileSync2, existsSync as existsSync2, appendFileSync } from "fs";
var MockChainAdapter = class {
  ledger = /* @__PURE__ */ new Map();
  async anchor(hash, metadata) {
    const timestamp = (/* @__PURE__ */ new Date()).toISOString();
    this.ledger.set(hash, { timestamp, metadata });
    return { hash, timestamp, txId: `mock-tx-${Date.now()}` };
  }
  async verify(hash) {
    return this.ledger.has(hash);
  }
  async getTimestamp(hash) {
    const entry = this.ledger.get(hash);
    return entry?.timestamp ?? null;
  }
};
var FileChainAdapter = class {
  /**
   * @param filePath - Path to the append-only JSONL file for storing anchors.
   */
  constructor(filePath) {
    this.filePath = filePath;
  }
  readEntries() {
    if (!existsSync2(this.filePath)) return [];
    const raw = readFileSync2(this.filePath, "utf-8").trim();
    if (!raw) return [];
    return raw.split("\n").map((line) => JSON.parse(line));
  }
  async anchor(hash, metadata) {
    const timestamp = (/* @__PURE__ */ new Date()).toISOString();
    const entry = { hash, timestamp, metadata };
    appendFileSync(this.filePath, JSON.stringify(entry) + "\n", "utf-8");
    return { hash, timestamp };
  }
  async verify(hash) {
    return this.readEntries().some((e) => e.hash === hash);
  }
  async getTimestamp(hash) {
    const entry = this.readEntries().find((e) => e.hash === hash);
    return entry?.timestamp ?? null;
  }
};

// src/privacy.ts
import { sha256 as sha2563 } from "@noble/hashes/sha256";
import { hkdf } from "@noble/hashes/hkdf";
import { xchacha20poly1305 } from "@noble/ciphers/chacha.js";
import { gcm } from "@noble/ciphers/aes.js";
import { randomBytes } from "@noble/ciphers/utils.js";
import { x25519 } from "@noble/curves/ed25519.js";
var HKDF_INFO = "exit-encryption-v1.2";
function toHex(bytes) {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function fromHex(hex) {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) bytes[i / 2] = parseInt(hex.slice(i, i + 2), 16);
  return bytes;
}
function deriveKey(shared, salt) {
  return hkdf(sha2563, shared, salt, HKDF_INFO, 32);
}
function encryptMarker(marker, recipientPublicKey, algorithm = "xchacha20-poly1305") {
  const ephemeralPrivate = randomBytes(32);
  const ephemeralPublic = x25519.getPublicKey(ephemeralPrivate);
  const shared = x25519.getSharedSecret(ephemeralPrivate, recipientPublicKey);
  const markerIdBytes = marker.id ? new TextEncoder().encode(marker.id) : void 0;
  const salt = markerIdBytes && markerIdBytes.length > 0 ? markerIdBytes : randomBytes(32);
  const key = deriveKey(shared, salt);
  const plaintext = new TextEncoder().encode(JSON.stringify(marker));
  let nonce;
  let ciphertext;
  if (algorithm === "aes-256-gcm") {
    nonce = randomBytes(12);
    const cipher = gcm(key, nonce);
    ciphertext = cipher.encrypt(plaintext);
  } else {
    nonce = randomBytes(24);
    const cipher = xchacha20poly1305(key, nonce);
    ciphertext = cipher.encrypt(plaintext);
  }
  return {
    ephemeralPublicKey: toHex(ephemeralPublic),
    nonce: toHex(nonce),
    ciphertext: toHex(ciphertext),
    salt: toHex(salt),
    algorithm
  };
}
function decryptMarker(blob, privateKey) {
  const ephemeralPublic = fromHex(blob.ephemeralPublicKey);
  const shared = x25519.getSharedSecret(privateKey, ephemeralPublic);
  let key;
  if (blob.salt) {
    const salt = fromHex(blob.salt);
    key = deriveKey(shared, salt);
  } else {
    key = sha2563(shared);
  }
  const nonce = fromHex(blob.nonce);
  const ciphertext = fromHex(blob.ciphertext);
  let plaintext;
  if (blob.algorithm === "aes-256-gcm") {
    const cipher = gcm(key, nonce);
    plaintext = cipher.decrypt(ciphertext);
  } else {
    const cipher = xchacha20poly1305(key, nonce);
    plaintext = cipher.decrypt(ciphertext);
  }
  return JSON.parse(new TextDecoder().decode(plaintext));
}
function hashField(value) {
  const canonical = canonicalize(value);
  const hash = sha2563(new TextEncoder().encode(canonical));
  return `redacted:sha256:${toHex(hash)}`;
}
function redactMarker(marker, fields) {
  const result = { ...marker };
  for (const field of fields) {
    if (field in result) {
      result[field] = hashField(result[field]);
    }
  }
  return result;
}
function createMinimalDisclosure(marker, revealFields) {
  const full = marker;
  const result = {};
  for (const key of Object.keys(full)) {
    if (revealFields.includes(key)) {
      result[key] = full[key];
    } else {
      result[key] = hashField(full[key]);
    }
  }
  return result;
}
var InMemoryKeyStore = class {
  keys = /* @__PURE__ */ new Map();
  async store(id, key) {
    this.keys.set(id, new Uint8Array(key));
  }
  async retrieve(id) {
    return this.keys.get(id) ?? null;
  }
  async delete(id) {
    const key = this.keys.get(id);
    if (key) key.fill(0);
    this.keys.delete(id);
  }
};
async function encryptMarkerWithManagedKey(marker, keyStore, algorithm = "xchacha20-poly1305") {
  const key = randomBytes(32);
  await keyStore.store(marker.id, key);
  const plaintext = new TextEncoder().encode(JSON.stringify(marker));
  let nonce;
  let ciphertext;
  if (algorithm === "aes-256-gcm") {
    nonce = randomBytes(12);
    const cipher = gcm(key, nonce);
    ciphertext = cipher.encrypt(plaintext);
  } else {
    nonce = randomBytes(24);
    const cipher = xchacha20poly1305(key, nonce);
    ciphertext = cipher.encrypt(plaintext);
  }
  return {
    markerId: marker.id,
    nonce: toHex(nonce),
    ciphertext: toHex(ciphertext),
    algorithm
  };
}
async function decryptManagedMarker(blob, keyStore) {
  const key = await keyStore.retrieve(blob.markerId);
  if (!key) {
    throw new Error(`Key not found for marker "${blob.markerId}" \u2014 may have been crypto-shredded`);
  }
  const nonce = fromHex(blob.nonce);
  const ciphertext = fromHex(blob.ciphertext);
  let plaintext;
  if (blob.algorithm === "aes-256-gcm") {
    const cipher = gcm(key, nonce);
    plaintext = cipher.decrypt(ciphertext);
  } else {
    const cipher = xchacha20poly1305(key, nonce);
    plaintext = cipher.decrypt(ciphertext);
  }
  return JSON.parse(new TextDecoder().decode(plaintext));
}
async function deleteMarkerKey(markerId, keyStore) {
  await keyStore.delete(markerId);
}

// src/resolver.ts
function isDid(str) {
  return /^did:[a-z]+:.+$/.test(str);
}
function didMethod(did) {
  if (!isDid(did)) return "unknown";
  const method = did.split(":")[1];
  if (method === "key") return "key";
  if (method === "keri") return "keri";
  if (method === "web") return "web";
  return "unknown";
}
function resolveDid(did) {
  const method = didMethod(did);
  switch (method) {
    case "key": {
      const publicKey = publicKeyFromDid(did);
      return {
        publicKey,
        method: "key",
        multicodecPrefix: new Uint8Array([237, 1])
      };
    }
    case "keri":
      throw new Error("did:keri resolution requires a KeyEventLog \u2014 use resolveDidKeri() instead");
    case "web":
      throw new Error("did:web resolution not yet implemented");
    default:
      throw new Error(`Unknown DID method: ${did}`);
  }
}
function createDidDocument(did) {
  const method = didMethod(did);
  if (method !== "key") {
    throw new Error(
      `createDidDocument currently only supports did:key, got did:${method}`
    );
  }
  const multibaseKey = did.slice("did:key:".length);
  return {
    "@context": [
      "https://www.w3.org/ns/did/v1",
      "https://w3id.org/security/multikey/v1"
    ],
    id: did,
    verificationMethod: [
      {
        id: `${did}#${multibaseKey}`,
        type: "Multikey",
        controller: did,
        publicKeyMultibase: multibaseKey
      }
    ],
    authentication: [`${did}#${multibaseKey}`],
    assertionMethod: [`${did}#${multibaseKey}`]
  };
}
function resolveDidKeri(did, keyEventLog) {
  if (didMethod(did) !== "keri") {
    throw new Error(`Expected did:keri, got: ${did}`);
  }
  const state = keyEventLog.getCurrentState();
  if (state.did !== did) {
    throw new Error(`Key event log DID mismatch: expected ${did}, got ${state.did}`);
  }
  return {
    publicKey: state.currentKeys[0],
    method: "keri",
    multicodecPrefix: new Uint8Array([237, 1])
  };
}
function createDidKeri(inceptionEvent) {
  return inceptionEvent.identifier;
}

// src/registry.ts
var MarkerRegistry = class {
  byId = /* @__PURE__ */ new Map();
  bySubject = /* @__PURE__ */ new Map();
  byOrigin = /* @__PURE__ */ new Map();
  /**
   * Register a marker. Indexes by ID, subject DID, and origin.
   *
   * @param marker - The EXIT marker to register.
   * @throws {Error} If a marker with the same ID is already registered.
   */
  register(marker) {
    if (this.byId.has(marker.id)) {
      throw new Error(`Marker already registered: ${marker.id}`);
    }
    this.byId.set(marker.id, marker);
    if (!this.bySubject.has(marker.subject)) {
      this.bySubject.set(marker.subject, /* @__PURE__ */ new Set());
    }
    this.bySubject.get(marker.subject).add(marker.id);
    if (!this.byOrigin.has(marker.origin)) {
      this.byOrigin.set(marker.origin, /* @__PURE__ */ new Set());
    }
    this.byOrigin.get(marker.origin).add(marker.id);
  }
  /**
   * Look up a marker by ID.
   *
   * @param id - The marker ID to look up.
   * @returns The marker if found, or `null`.
   */
  lookup(id) {
    return this.byId.get(id) ?? null;
  }
  /**
   * Find all markers for a given subject DID.
   *
   * @param did - The subject DID to search for.
   * @returns An array of markers for the subject (empty if none found).
   */
  findBySubject(did) {
    const ids = this.bySubject.get(did);
    if (!ids) return [];
    return [...ids].map((id) => this.byId.get(id));
  }
  /**
   * Find all markers from a given origin URI.
   *
   * @param uri - The origin URI to search for.
   * @returns An array of markers from the origin (empty if none found).
   */
  findByOrigin(uri) {
    const ids = this.byOrigin.get(uri);
    if (!ids) return [];
    return [...ids].map((id) => this.byId.get(id));
  }
  /**
   * Find markers within a time range (inclusive).
   *
   * @param start - The start of the time range.
   * @param end - The end of the time range.
   * @returns An array of markers whose timestamps fall within the range.
   */
  findByTimeRange(start, end) {
    const results = [];
    for (const marker of this.byId.values()) {
      const t = new Date(marker.timestamp).getTime();
      if (t >= start.getTime() && t <= end.getTime()) {
        results.push(marker);
      }
    }
    return results;
  }
  /**
   * Total number of registered markers.
   *
   * @returns The count of markers in the registry.
   */
  count() {
    return this.byId.size;
  }
  /**
   * Export all markers as an array.
   *
   * @returns An array of all registered markers.
   */
  export() {
    return [...this.byId.values()];
  }
};

// src/batch.ts
import { sha256 as sha2564 } from "@noble/hashes/sha256";
function toHex2(bytes) {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function hashPair(a, b) {
  const combined = a < b ? a + b : b + a;
  return toHex2(sha2564(new TextEncoder().encode(combined)));
}
function computeMerkleRoot(hashes) {
  if (hashes.length === 0) throw new Error("Cannot compute Merkle root of empty set");
  if (hashes.length === 1) return hashes[0];
  let level = [...hashes];
  while (level.length > 1) {
    const next = [];
    for (let i = 0; i < level.length; i += 2) {
      if (i + 1 < level.length) {
        next.push(hashPair(level[i], level[i + 1]));
      } else {
        next.push(hashPair(level[i], level[i]));
      }
    }
    level = next;
  }
  return level[0];
}
function computeMerkleProof(hashes, index) {
  if (index < 0 || index >= hashes.length) {
    throw new Error(`Index ${index} out of range [0, ${hashes.length})`);
  }
  const leaf = hashes[index];
  const path = [];
  let level = [...hashes];
  let idx = index;
  while (level.length > 1) {
    const next = [];
    for (let i = 0; i < level.length; i += 2) {
      if (i + 1 < level.length) {
        next.push(hashPair(level[i], level[i + 1]));
      } else {
        next.push(hashPair(level[i], level[i]));
      }
    }
    const siblingIdx = idx % 2 === 0 ? idx + 1 : idx - 1;
    if (siblingIdx < level.length) {
      path.push({
        hash: level[siblingIdx],
        position: idx % 2 === 0 ? "right" : "left"
      });
    } else {
      path.push({
        hash: level[idx],
        position: "right"
      });
    }
    idx = Math.floor(idx / 2);
    level = next;
  }
  return { leaf, path, root: level[0] };
}
function verifyBatchMembership(marker, batchRoot, proof) {
  const markerHash = computeAnchorHash(marker);
  if (markerHash !== proof.leaf) return false;
  let current = proof.leaf;
  for (const step of proof.path) {
    if (step.position === "left") {
      current = hashPair(step.hash, current);
    } else {
      current = hashPair(current, step.hash);
    }
  }
  return current === batchRoot;
}
function createBatchExit(markers) {
  if (markers.length === 0) throw new Error("Cannot create batch from empty marker set");
  const leaves = markers.map((m) => computeAnchorHash(m));
  const merkleRoot = computeMerkleRoot(leaves);
  return {
    merkleRoot,
    count: markers.length,
    timestamp: (/* @__PURE__ */ new Date()).toISOString(),
    leaves
  };
}
function createShutdownBatch(platformDid, markers, privateKey) {
  if (markers.length === 0) throw new Error("Cannot create shutdown batch from empty marker set");
  const leaves = markers.map((m) => computeAnchorHash(m));
  const merkleRoot = computeMerkleRoot(leaves);
  const now = (/* @__PURE__ */ new Date()).toISOString();
  const markerIds = markers.map((m) => m.id);
  const attestationPayload = canonicalize({
    platformDid,
    merkleRoot,
    count: markers.length,
    markerIds,
    timestamp: now
  });
  const data = new TextEncoder().encode(attestationPayload);
  const signature = sign2(data, privateKey);
  const proofValue = Buffer.from(signature).toString("base64");
  return {
    platformDid,
    merkleRoot,
    count: markers.length,
    markerIds,
    timestamp: now,
    platformAttestation: {
      type: "Ed25519Signature2020",
      created: now,
      verificationMethod: platformDid,
      proofValue
    }
  };
}

// src/interop.ts
import { EventEmitter } from "events";
function createExitMiddleware(opts) {
  const base = opts.basePath ?? "/exit";
  return (req, res, next) => {
    if (req.method === "POST" && req.path === base) {
      const validation = validateMarker(req.body);
      if (!validation.valid) {
        res.status(400).json({ errors: validation.errors });
        return;
      }
      opts.onPost(req.body).then((m) => res.status(201).json(m)).catch((e) => res.status(500).json({ error: e.message }));
      return;
    }
    if (req.method === "GET" && req.path.startsWith(base + "/") && !req.path.endsWith("/verify")) {
      const id = req.params.id ?? req.path.slice(base.length + 1);
      opts.onGet(id).then((m) => {
        if (!m) {
          res.status(404).json({ error: "Not found" });
          return;
        }
        res.status(200).json(m);
      }).catch((e) => res.status(500).json({ error: e.message }));
      return;
    }
    if (req.method === "POST" && req.path.endsWith("/verify")) {
      const id = req.params.id ?? req.path.slice(base.length + 1, -"/verify".length);
      opts.onGet(id).then((m) => {
        if (!m) {
          res.status(404).json({ error: "Not found" });
          return;
        }
        const result = verifyMarker(m);
        res.status(200).json(result);
      }).catch((e) => res.status(500).json({ error: e.message }));
      return;
    }
    next();
  };
}
function createExitHook(callbacks) {
  return {
    async execute(marker) {
      if (callbacks.beforeExit) await callbacks.beforeExit(marker);
      if (callbacks.onExit) await callbacks.onExit(marker);
      if (callbacks.afterExit) await callbacks.afterExit(marker);
    }
  };
}
var ExitEventEmitter = class extends EventEmitter {
  /**
   * Emit an intent event.
   * @param marker - The EXIT marker associated with the intent.
   */
  emitIntent(marker) {
    this.emit("intent", marker);
  }
  /**
   * Emit a negotiating event.
   * @param marker - The EXIT marker being negotiated.
   */
  emitNegotiating(marker) {
    this.emit("negotiating", marker);
  }
  /**
   * Emit a signing event.
   * @param marker - The EXIT marker being signed.
   */
  emitSigning(marker) {
    this.emit("signing", marker);
  }
  /**
   * Emit a departed event.
   * @param marker - The finalized EXIT marker.
   */
  emitDeparted(marker) {
    this.emit("departed", marker);
  }
};
function serializeForTransport(marker) {
  const json = canonicalize(marker);
  const payload = Buffer.from(json, "utf-8");
  const header = Buffer.alloc(4);
  header.writeUInt32BE(payload.length, 0);
  return Buffer.concat([header, payload]);
}
var MAX_PAYLOAD_SIZE = 1048576;
function deserializeFromTransport(buffer) {
  if (buffer.length < 4) throw new Error("Transport buffer too short");
  const length = buffer.readUInt32BE(0);
  if (length > MAX_PAYLOAD_SIZE) throw new Error(`Transport payload too large: ${length} bytes (max ${MAX_PAYLOAD_SIZE})`);
  if (buffer.length < 4 + length) throw new Error("Transport buffer truncated");
  const json = buffer.slice(4, 4 + length).toString("utf-8");
  return JSON.parse(json);
}

// src/keri.ts
import { sha256 as sha2565 } from "@noble/hashes/sha256";
function toHex3(bytes) {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function signPayload(data, privateKey) {
  const sig = sign2(new TextEncoder().encode(data), privateKey);
  return Buffer.from(sig).toString("base64");
}
function verifyPayload(data, signature, publicKey) {
  const sig = new Uint8Array(Buffer.from(signature, "base64"));
  return verify2(new TextEncoder().encode(data), sig, publicKey);
}
function eventPayload(event) {
  if (event.type === "inception") {
    return JSON.stringify({
      type: event.type,
      identifier: event.identifier,
      keys: event.keys.map(toHex3),
      nextKeyDigests: event.nextKeyDigests,
      witnesses: event.witnesses,
      timestamp: event.timestamp
    });
  }
  return JSON.stringify({
    type: event.type,
    prior: event.prior,
    current: event.current.map(toHex3),
    next: event.next,
    sequenceNumber: event.sequenceNumber,
    timestamp: event.timestamp
  });
}
function digestKey(publicKey) {
  return toHex3(sha2565(publicKey));
}
function createInception(opts) {
  const identifier = `did:keri:${toHex3(sha2565(opts.keys[0])).slice(0, 32)}`;
  const timestamp = (/* @__PURE__ */ new Date()).toISOString();
  const event = {
    type: "inception",
    identifier,
    keys: opts.keys,
    nextKeyDigests: opts.nextKeyDigests,
    witnesses: opts.witnesses,
    timestamp,
    signature: ""
  };
  event.signature = signPayload(eventPayload(event), opts.signingKey);
  return event;
}
function createRotation(currentState, newKeys, nextKeyDigests, signingKey) {
  const timestamp = (/* @__PURE__ */ new Date()).toISOString();
  const prior = toHex3(sha2565(new TextEncoder().encode(JSON.stringify({
    keys: currentState.currentKeys.map(toHex3),
    seq: currentState.sequenceNumber
  }))));
  const event = {
    type: "rotation",
    prior,
    current: newKeys,
    next: nextKeyDigests,
    sequenceNumber: currentState.sequenceNumber + 1,
    timestamp,
    signature: ""
  };
  event.signature = signPayload(eventPayload(event), signingKey);
  return event;
}
function verifyKeyState(events) {
  if (events.length === 0) {
    throw new Error("Empty event log");
  }
  const first = events[0];
  if (first.type !== "inception") {
    throw new Error("First event must be inception");
  }
  if (!verifyPayload(eventPayload(first), first.signature, first.keys[0])) {
    throw new Error("Invalid inception signature");
  }
  let state = {
    did: first.identifier,
    currentKeys: [...first.keys],
    nextKeyDigests: [...first.nextKeyDigests],
    sequenceNumber: 0,
    witnesses: first.witnesses ? [...first.witnesses] : void 0
  };
  for (let i = 1; i < events.length; i++) {
    const event = events[i];
    if (event.type !== "rotation") {
      throw new Error(`Expected rotation event at index ${i}`);
    }
    if (event.sequenceNumber !== state.sequenceNumber + 1) {
      throw new Error(
        `Sequence gap: expected ${state.sequenceNumber + 1}, got ${event.sequenceNumber}`
      );
    }
    let sigValid = false;
    for (const key of state.currentKeys) {
      if (verifyPayload(eventPayload(event), event.signature, key)) {
        sigValid = true;
        break;
      }
    }
    if (!sigValid) {
      throw new Error(`Invalid rotation signature at sequence ${event.sequenceNumber}`);
    }
    for (const newKey of event.current) {
      const digest = digestKey(newKey);
      if (!state.nextKeyDigests.includes(digest)) {
        throw new Error(
          `New key not in pre-committed next key digests at sequence ${event.sequenceNumber}`
        );
      }
    }
    state = {
      ...state,
      currentKeys: [...event.current],
      nextKeyDigests: [...event.next],
      sequenceNumber: event.sequenceNumber
    };
  }
  return state;
}
function isKeyCompromised(keyState, marker) {
  const vm = marker.proof?.verificationMethod;
  if (!vm) return false;
  if (vm.startsWith("did:key:")) {
    try {
      const pubKey = publicKeyFromDid(vm);
      const pubKeyHex = toHex3(pubKey);
      const currentHexes = keyState.currentKeys.map(toHex3);
      return !currentHexes.includes(pubKeyHex);
    } catch {
      return false;
    }
  }
  return false;
}
var KeyEventLog = class {
  events = [];
  /**
   * Append an event to the log. Verifies the full chain after append.
   *
   * @param event - The key event (inception or rotation) to append.
   * @throws {Error} If the event makes the chain invalid (bad signature, sequence gap, etc.).
   */
  append(event) {
    this.events.push(event);
    try {
      verifyKeyState(this.events);
    } catch (e) {
      this.events.pop();
      throw e;
    }
  }
  /**
   * Get all events in the log.
   *
   * @returns A shallow copy of the event array.
   */
  getEvents() {
    return [...this.events];
  }
  /**
   * Get the current key state by walking the log.
   *
   * @returns The current {@link KeyState} derived from replaying all events.
   * @throws {Error} If the log is empty or verification fails.
   */
  getCurrentState() {
    return verifyKeyState(this.events);
  }
  /** Get the number of events. */
  get length() {
    return this.events.length;
  }
};

// src/pre-rotation.ts
function generatePreRotatedKeys() {
  const currentKeyPair = generateKeyPair();
  const nextKeyPair = generateKeyPair();
  const nextKeyDigest = commitNextKey(nextKeyPair.publicKey);
  return { currentKeyPair, nextKeyPair, nextKeyDigest };
}
function commitNextKey(publicKey) {
  return digestKey(publicKey);
}
function verifyNextKeyCommitment(publicKey, commitment) {
  return digestKey(publicKey) === commitment;
}
function rotateKeys(currentState, nextKeyPair, newNextKeyDigest, signingKey) {
  const rotationEvent = createRotation(
    currentState,
    [nextKeyPair.publicKey],
    [newNextKeyDigest],
    signingKey
  );
  const newState = {
    ...currentState,
    currentKeys: [nextKeyPair.publicKey],
    nextKeyDigests: [newNextKeyDigest],
    sequenceNumber: currentState.sequenceNumber + 1
  };
  return { newState, rotationEvent };
}

// src/ethics.ts
function durationToDays(dur) {
  const m = dur.match(/^P(?:(\d+)Y)?(?:(\d+)M)?(?:(\d+)D)?$/);
  if (!m) return -1;
  return parseInt(m[1] || "0") * 365 + parseInt(m[2] || "0") * 30 + parseInt(m[3] || "0");
}
function detectCoercion(marker) {
  const signals = [];
  if (marker.exitType === "forced" /* Forced */ && marker.status === "good_standing" /* GoodStanding */ && marker.dispute?.originStatus === "disputed" /* Disputed */) {
    signals.push("Forced exit with conflicting status: subject claims good_standing but origin disputes");
  }
  if (marker.exitType === "emergency" /* Emergency */ && marker.emergencyJustification && !/shutdown|unresponsive|offline|failing|down|unavailable/i.test(marker.emergencyJustification)) {
    signals.push("Emergency exit without infrastructure failure indicators \u2014 possible coerced departure");
  }
  if (marker.exitType === "forced" /* Forced */ && marker.lineage?.lineageChain) {
    const chain = marker.lineage.lineageChain;
    if (chain.length <= 1) {
      signals.push("Forced exit with minimal lineage \u2014 possible retaliation against new participant");
    }
  }
  if (marker.exitType === "forced" /* Forced */ && marker.metadata?.tags) {
    const tenureTag = marker.metadata.tags.find((t) => t.startsWith("tenure:"));
    if (tenureTag) {
      const days = durationToDays(tenureTag.replace("tenure:", ""));
      if (days >= 0 && days < 30) {
        signals.push("Forced exit after very short tenure (<30 days) \u2014 possible retaliation");
      }
    }
  }
  let riskLevel = "none";
  if (signals.length === 1) riskLevel = "low";
  else if (signals.length === 2) riskLevel = "medium";
  else if (signals.length >= 3) riskLevel = "high";
  const recommendation = riskLevel === "none" ? "No coercion indicators detected." : riskLevel === "low" ? "Minor coercion signal detected. Consider requesting additional context." : riskLevel === "medium" ? "Multiple coercion indicators. Recommend independent verification before relying on this marker." : "Strong coercion indicators. This marker should be treated with significant skepticism and independently investigated.";
  return { signals, riskLevel, recommendation };
}
function detectWeaponization(markers) {
  const patterns = [];
  const affectedSubjects = [];
  const byOrigin = /* @__PURE__ */ new Map();
  for (const m of markers) {
    const list = byOrigin.get(m.origin) || [];
    list.push(m);
    byOrigin.set(m.origin, list);
  }
  for (const [origin, originMarkers] of byOrigin) {
    const forcedExits = originMarkers.filter((m) => m.exitType === "forced" /* Forced */);
    const disputedOrigin = originMarkers.filter(
      (m) => m.dispute?.originStatus === "disputed" /* Disputed */
    );
    if (forcedExits.length >= 3) {
      patterns.push(
        `Origin "${origin}" has ${forcedExits.length} forced exits \u2014 potential weaponization`
      );
      for (const m of forcedExits) {
        if (!affectedSubjects.includes(m.subject)) affectedSubjects.push(m.subject);
      }
    }
    if (disputedOrigin.length >= 3 && disputedOrigin.length === originMarkers.filter((m) => m.dispute?.originStatus !== void 0).length) {
      patterns.push(
        `Origin "${origin}" disputes all departing subjects (${disputedOrigin.length}) \u2014 potential blacklisting behavior`
      );
      for (const m of disputedOrigin) {
        if (!affectedSubjects.includes(m.subject)) affectedSubjects.push(m.subject);
      }
    }
    if (originMarkers.length >= 5) {
      const timestamps = originMarkers.map((m) => new Date(m.timestamp).getTime()).sort();
      const span = timestamps[timestamps.length - 1] - timestamps[0];
      const spanDays = span / (1e3 * 60 * 60 * 24);
      if (spanDays <= 7) {
        patterns.push(
          `Origin "${origin}" had ${originMarkers.length} exits within ${Math.ceil(spanDays)} days \u2014 potential purge`
        );
        for (const m of originMarkers) {
          if (!affectedSubjects.includes(m.subject)) affectedSubjects.push(m.subject);
        }
      }
    }
  }
  let severity = "none";
  if (patterns.length === 1) severity = "concerning";
  else if (patterns.length >= 2) severity = "severe";
  return { patterns, severity, affectedSubjects };
}
function detectReputationLaundering(markers, subjectDid) {
  const signals = [];
  const subjectMarkers = markers.filter((m) => m.subject === subjectDid);
  if (subjectMarkers.length < 2) {
    return { signals: [], probability: "low" };
  }
  const shortTenure = subjectMarkers.filter((m) => {
    if (m.metadata?.tags) {
      const tag = m.metadata.tags.find((t) => t.startsWith("tenure:"));
      if (tag) {
        const days = durationToDays(tag.replace("tenure:", ""));
        return days >= 0 && days < 30;
      }
    }
    return false;
  });
  if (shortTenure.length >= 2) {
    signals.push(
      `Subject has ${shortTenure.length} exits with tenure under 30 days \u2014 possible identity cycling`
    );
  }
  if (subjectMarkers.length >= 3) {
    const timestamps = subjectMarkers.map((m) => new Date(m.timestamp).getTime()).sort();
    const spanDays = (timestamps[timestamps.length - 1] - timestamps[0]) / (1e3 * 60 * 60 * 24);
    if (spanDays <= 90) {
      signals.push(
        `Subject has ${subjectMarkers.length} exits within ${Math.ceil(spanDays)} days \u2014 high churn rate`
      );
    }
  }
  const selfAttestedGood = subjectMarkers.filter(
    (m) => m.selfAttested && m.status === "good_standing" /* GoodStanding */ && m.dispute?.originStatus === void 0
  );
  if (selfAttestedGood.length >= 2) {
    signals.push(
      `Subject has ${selfAttestedGood.length} self-attested good_standing exits with no origin corroboration \u2014 suspicious pattern`
    );
  }
  let probability = "low";
  if (signals.length === 1) probability = "medium";
  else if (signals.length >= 2) probability = "high";
  return { signals, probability };
}
var RETALIATION_WINDOW_MS = 72 * 60 * 60 * 1e3;
function detectRetaliationWindow(marker, reportedActivities) {
  const indicators = [];
  if (marker.exitType !== "forced" /* Forced */ && marker.exitType !== "directed" /* Directed */ && marker.exitType !== "constructive" /* Constructive */) {
    return { detected: false, indicators };
  }
  if (!reportedActivities || reportedActivities.length === 0) {
    return { detected: false, indicators };
  }
  const exitTime = new Date(marker.timestamp).getTime();
  for (const activity of reportedActivities) {
    const activityTime = new Date(activity.timestamp).getTime();
    const delta = exitTime - activityTime;
    if (delta >= 0 && delta <= RETALIATION_WINDOW_MS) {
      const hours = Math.round(delta / (1e3 * 60 * 60));
      indicators.push(
        `Forced exit occurred ${hours}h after "${activity.type}" \u2014 within 72h retaliation window`
      );
    }
  }
  const coercion = detectCoercion(marker);
  if (coercion.riskLevel !== "none" && indicators.length > 0) {
    indicators.push(
      `Coercion signals (${coercion.riskLevel}) compound retaliation concern`
    );
  }
  return {
    detected: indicators.length > 0,
    indicators
  };
}
function generateEthicsReport(markers) {
  const coercionFindings = markers.map((m) => ({
    markerId: m.id,
    signals: detectCoercion(m)
  }));
  const weaponization = detectWeaponization(markers);
  const subjects = [...new Set(markers.map((m) => m.subject))];
  const launderingFindings = subjects.map((s) => ({
    subjectDid: s,
    signals: detectReputationLaundering(markers, s)
  }));
  const recommendations = [];
  const hasCoercion = coercionFindings.some((f) => f.signals.riskLevel !== "none");
  const hasWeaponization = weaponization.severity !== "none";
  const hasLaundering = launderingFindings.some((f) => f.signals.probability !== "low");
  if (hasCoercion) {
    recommendations.push("Coercion signals detected. Investigate flagged markers for possible forced departures under duress.");
  }
  if (hasWeaponization) {
    recommendations.push("Weaponization patterns detected. Review flagged origins for systemic abuse of forced exit mechanisms.");
  }
  if (hasLaundering) {
    recommendations.push("Reputation laundering signals detected. Apply additional verification for flagged subjects.");
  }
  if (!hasCoercion && !hasWeaponization && !hasLaundering) {
    recommendations.push("No ethical concerns detected in this marker set.");
  }
  let overallRisk = "low";
  const highCoercion = coercionFindings.some((f) => f.signals.riskLevel === "high");
  if (weaponization.severity === "severe" || highCoercion && hasLaundering) {
    overallRisk = "critical";
  } else if (hasWeaponization || highCoercion) {
    overallRisk = "high";
  } else if (hasCoercion || hasLaundering) {
    overallRisk = "medium";
  }
  return {
    generatedAt: (/* @__PURE__ */ new Date()).toISOString(),
    markerCount: markers.length,
    coercionFindings,
    weaponization,
    launderingFindings,
    recommendations,
    overallRisk
  };
}

// src/guardrails.ts
var ANTI_WEAPONIZATION_CLAUSE = `EXIT markers MUST NOT be used as blacklists, ban lists, or exclusion databases. An EXIT marker records a departure event; it does not constitute a judgment of character, competence, or trustworthiness. Any system that aggregates EXIT markers to systematically exclude subjects based on exitType or originStatus is in violation of this specification. Origins that consistently issue 'disputed' status without substantive basis may be flagged for weaponization by ethics auditors.`;
function addCoercionLabel(marker, label, _evidence) {
  return {
    ...marker,
    coercionLabel: label,
    metadata: {
      ...marker.metadata,
      tags: [
        ...marker.metadata?.tags || [],
        `coercion:${label}`,
        `coercion-evidence:${_evidence}`
      ]
    }
  };
}
function addRightOfReply(marker, replyText, signerKey) {
  const reply = {
    replyText,
    signerKey,
    timestamp: (/* @__PURE__ */ new Date()).toISOString(),
    // In production this would be a real signature; placeholder for structural purposes
    signature: `sig:${signerKey}:${Buffer.from(replyText).toString("base64")}`
  };
  return {
    ...marker,
    dispute: {
      ...marker.dispute,
      rightOfReply: reply
    }
  };
}
function validateEthicalCompliance(marker) {
  const violations = [];
  if (marker.exitType === "forced" /* Forced */) {
    const hasReason = marker.metadata?.reason || marker.metadata?.narrative;
    if (!hasReason) {
      violations.push("Forced exit must include a reason or narrative (Module E) explaining the expulsion.");
    }
  }
  if (marker.dispute?.originStatus && marker.dispute.originStatus !== marker.status && !marker.dispute.rightOfReply) {
    violations.push(
      "Origin attestation conflicts with subject status but no right of reply is attached. Subjects must be given the opportunity to attach a counter-narrative."
    );
  }
  if (marker.exitType === "emergency" /* Emergency */ && !marker.emergencyJustification) {
    violations.push("Emergency exit without justification violates ethical transparency requirements.");
  }
  const expiryDate = marker.expires || marker.sunsetDate;
  if (expiryDate && new Date(expiryDate) < /* @__PURE__ */ new Date()) {
    violations.push("Marker has passed its sunset date and should no longer be relied upon.");
  }
  return { compliant: violations.length === 0, violations };
}
function applySunset(marker, policy) {
  const exitDate = new Date(marker.timestamp);
  const sunsetDate = new Date(exitDate.getTime() + policy.durationDays * 24 * 60 * 60 * 1e3);
  return {
    ...marker,
    expires: sunsetDate.toISOString(),
    sunsetDate: sunsetDate.toISOString()
    // backward compat (deprecated)
  };
}
function isExpired(marker) {
  const expiryDate = marker.expires || marker.sunsetDate;
  if (!expiryDate) return false;
  return new Date(expiryDate) < /* @__PURE__ */ new Date();
}

// src/full-service.ts
async function tryTsaTimestamp(anchorHash, url) {
  try {
    const tsa = await import("./tsa-P7QA4BVE.js");
    if (typeof tsa.requestTimestamp === "function") {
      return await tsa.requestTimestamp(anchorHash, url);
    }
  } catch {
  }
  return void 0;
}
async function tryGitLedger(marker, anchorHash, config) {
  try {
    const ledger = await import("./git-ledger-WTX64KSR.js");
    if (typeof ledger.anchorToGit === "function") {
      return await ledger.anchorToGit(marker, config);
    }
  } catch {
  }
  return void 0;
}
async function tryVisual(marker) {
  try {
    const vis = await import("./visual-ZYEGPIH5.js");
    if (typeof vis.renderDoorASCII === "function") {
      return vis.renderDoorASCII(marker.id || "");
    }
  } catch {
  }
  return void 0;
}
async function tryVerifyTsa(receipt) {
  try {
    const tsa = await import("./tsa-P7QA4BVE.js");
    if (typeof tsa.checkTSAReceiptStructure === "function") {
      return await tsa.checkTSAReceiptStructure(receipt, receipt.hash);
    }
  } catch {
  }
  return void 0;
}
async function departAndAnchor(origin, opts) {
  const { marker, identity } = await quickExit(origin, opts);
  const anchorHash = computeAnchorHash(marker);
  const safeIdentity = opts?.includePrivateKey ? identity : { did: identity.did, publicKey: identity.publicKey };
  const result = { marker, identity: safeIdentity, anchorHash };
  const promises = [];
  if (opts?.anchor?.tsa) {
    promises.push(
      tryTsaTimestamp(anchorHash, opts.anchor.tsa.url).then((receipt) => {
        if (receipt) result.tsaReceipt = receipt;
      })
    );
  }
  if (opts?.anchor?.git) {
    promises.push(
      tryGitLedger(marker, anchorHash, opts.anchor.git).then((entry) => {
        if (entry) result.ledgerEntry = entry;
      })
    );
  }
  if (opts?.visual) {
    promises.push(
      tryVisual(marker).then((art) => {
        if (art) result.visual = art;
      })
    );
  }
  await Promise.all(promises);
  return result;
}
async function departAndVerify(markerInput, tsaReceipt) {
  const reasons = [];
  let marker;
  if (typeof markerInput === "string") {
    try {
      marker = fromJSON(markerInput);
    } catch (e) {
      return {
        valid: false,
        trustLevel: "none",
        anchorHash: "",
        signatureValid: false,
        reasons: [`Invalid marker: ${e.message}`]
      };
    }
  } else {
    marker = markerInput;
  }
  const verification = verifyMarker(marker);
  const signatureValid = verification.valid;
  if (signatureValid) {
    reasons.push("Signature valid");
  } else {
    reasons.push(`Signature invalid: ${verification.errors.join(", ")}`);
  }
  const anchorHash = computeAnchorHash(marker);
  reasons.push(`Anchor hash: ${anchorHash.slice(0, 16)}\u2026`);
  let tsaStructuralMatch;
  if (tsaReceipt) {
    tsaStructuralMatch = await tryVerifyTsa(tsaReceipt);
    if (tsaStructuralMatch === true) {
      reasons.push("TSA receipt structurally plausible (cryptographic verification requires external tooling)");
    } else if (tsaStructuralMatch === false) {
      reasons.push("TSA receipt structural check failed");
    } else {
      reasons.push("TSA module not available \u2014 receipt not checked");
    }
  }
  let trustLevel;
  if (!signatureValid) {
    trustLevel = "none";
  } else if (tsaReceipt && tsaStructuralMatch === false) {
    trustLevel = "low";
  } else {
    trustLevel = "informational";
  }
  return {
    valid: signatureValid,
    trustLevel,
    anchorHash,
    signatureValid,
    tsaStructuralMatch,
    reasons
  };
}

// src/key-compromise.ts
function flagCompromisedPlatformMarkers(markers, declaration) {
  const after = new Date(declaration.compromisedAfter).getTime();
  const before = declaration.compromisedBefore ? new Date(declaration.compromisedBefore).getTime() : Infinity;
  return markers.filter((m) => {
    const proofTime = new Date(m.proof.created).getTime();
    if (proofTime >= after && proofTime <= before) {
      if (m.proof.verificationMethod === declaration.platformDid) return true;
      if (m.dispute?.counterpartyAcks?.some(
        (ack) => ack.verificationMethod === declaration.platformDid
      )) return true;
    }
    return false;
  }).map((m) => m.id);
}
function createCompromiseMarker(compromisedDid, rotationEvent, newDid, signingKey) {
  const now = (/* @__PURE__ */ new Date()).toISOString();
  const marker = {
    "@context": EXIT_CONTEXT_V1,
    specVersion: EXIT_SPEC_VERSION,
    id: "",
    subject: compromisedDid,
    origin: "did:keri:key-event-log",
    timestamp: now,
    exitType: "keyCompromise",
    status: "unverified",
    selfAttested: true,
    proof: {},
    lineage: {
      predecessor: compromisedDid,
      successor: newDid
    },
    metadata: {
      reason: "Key compromise detected. Identity rotated via KERI key event log.",
      tags: ["key-compromise", "rotation"]
    }
  };
  const { proof: _p, id: _id, ...rest } = marker;
  marker.id = `urn:exit:${computeId(rest)}`;
  const DOMAIN_PREFIX2 = "exit-marker-v1.2:";
  const canonical = canonicalize({ ...marker, proof: void 0 });
  const data = new TextEncoder().encode(DOMAIN_PREFIX2 + canonical);
  const signature = sign2(data, signingKey);
  const proofValue = Buffer.from(signature).toString("base64");
  marker.proof = {
    type: "Ed25519Signature2020",
    created: now,
    verificationMethod: newDid,
    proofValue
  };
  return marker;
}
function verifyCompromiseRecovery(compromiseMarker, keyEventLog) {
  try {
    const state = keyEventLog.getCurrentState();
    const successor = compromiseMarker.lineage?.successor;
    if (!successor) return false;
    const compromised = compromiseMarker.subject;
    if (!compromised) return false;
    verifyKeyState(keyEventLog.getEvents());
    if (compromiseMarker.exitType !== "keyCompromise") return false;
    return true;
  } catch {
    return false;
  }
}
function linkCompromisedMarkers(compromiseMarker, affectedMarkerIds) {
  return {
    compromiseMarkerId: compromiseMarker.id,
    affectedMarkerIds: [...affectedMarkerIds],
    timestamp: (/* @__PURE__ */ new Date()).toISOString()
  };
}

// src/dispute.ts
function createDispute(markerId, reason, arbiterDid, filerDid, evidenceRefs = [], expiresAt) {
  if (!markerId) throw new Error("markerId is required");
  if (!reason) throw new Error("reason is required");
  if (!arbiterDid || !arbiterDid.startsWith("did:")) throw new Error("arbiterDid must be a valid DID");
  if (!filerDid || !filerDid.startsWith("did:")) throw new Error("filerDid must be a valid DID");
  const now = (/* @__PURE__ */ new Date()).toISOString();
  return {
    id: `dispute:${markerId}:${Date.now()}`,
    markerId,
    filerDid,
    reason,
    arbiterDid,
    filedAt: now,
    expiresAt,
    evidenceRefs
  };
}
function resolveDispute(dispute, outcome, summary, arbiterPrivateKey) {
  if (dispute.resolution) throw new Error("Dispute is already resolved");
  const resolvedAt = (/* @__PURE__ */ new Date()).toISOString();
  const payload = canonicalize({ disputeId: dispute.id, outcome, summary, resolvedAt });
  const data = new TextEncoder().encode(payload);
  const signature = sign2(data, arbiterPrivateKey);
  const resolution = {
    outcome,
    summary,
    resolvedAt,
    proof: {
      type: "Ed25519Signature2020",
      created: resolvedAt,
      verificationMethod: dispute.arbiterDid,
      proofValue: Buffer.from(signature).toString("base64")
    }
  };
  return { ...dispute, resolution };
}
function verifyDisputeResolution(dispute) {
  if (!dispute.resolution) return false;
  const { outcome, summary, resolvedAt, proof } = dispute.resolution;
  const payload = canonicalize({ disputeId: dispute.id, outcome, summary, resolvedAt });
  const data = new TextEncoder().encode(payload);
  try {
    const publicKey = publicKeyFromDid(proof.verificationMethod);
    const signature = Uint8Array.from(Buffer.from(proof.proofValue, "base64"));
    return verify2(data, signature, publicKey);
  } catch {
    return false;
  }
}
function isDisputed(marker) {
  return getDisputeStatus(marker) === "active";
}
function getDisputeStatus(marker) {
  const disputes = marker.dispute?.disputes;
  if (!disputes || disputes.length === 0) {
    return marker.status === "disputed" ? "active" : "none";
  }
  const now = /* @__PURE__ */ new Date();
  let hasActive = false;
  let hasResolved = false;
  for (const d of disputes) {
    if (d.resolution) {
      hasResolved = true;
    } else if (d.disputeExpiry && new Date(d.disputeExpiry) < now) {
    } else {
      hasActive = true;
    }
  }
  if (hasActive) return "active";
  if (hasResolved) return "resolved";
  return "expired";
}

// src/passage.ts
var createDepartureMarker = createMarker;
var signDepartureMarker = signMarker;
var signDepartureWithSigner = signMarkerWithSigner;
var verifyDeparture = verifyMarker;
var verifyDepartureMultiAlg = verifyMarkerMultiAlg;
var quickDeparture = quickExit;
var quickPassageVerify = quickVerify;
var generatePassageIdentity = generateIdentity;
var createPassage = departAndAnchor;
var verifyPassage = departAndVerify;

// src/claim-store.ts
import { createHash } from "crypto";
var ClaimType = /* @__PURE__ */ ((ClaimType2) => {
  ClaimType2["ExitMarker"] = "exit_marker";
  ClaimType2["WitnessAttestation"] = "witness_attestation";
  ClaimType2["Identity"] = "identity";
  ClaimType2["Timestamp"] = "timestamp";
  ClaimType2["Custom"] = "custom";
  return ClaimType2;
})(ClaimType || {});
var MemoryClaimStore = class {
  claims = /* @__PURE__ */ new Map();
  put(claim) {
    if (!claim.id) {
      claim = { ...claim, id: this.computeId(claim) };
    }
    this.validateClaim(claim);
    this.claims.set(claim.id, { ...claim });
  }
  get(id) {
    return this.claims.get(id) ?? null;
  }
  query(q) {
    const excludeExpired = q.excludeExpired !== false;
    const limit = q.limit ?? 100;
    const now = (/* @__PURE__ */ new Date()).toISOString();
    let results = Array.from(this.claims.values());
    if (q.subject) results = results.filter((c) => c.subject === q.subject);
    if (q.type) results = results.filter((c) => c.type === q.type);
    if (q.issuer) results = results.filter((c) => c.issuer === q.issuer);
    if (q.markerRef) results = results.filter((c) => c.markerRef === q.markerRef);
    if (q.tags?.length) {
      results = results.filter(
        (c) => c.tags?.some((t) => q.tags.includes(t))
      );
    }
    if (excludeExpired) {
      results = results.filter((c) => !c.expiresAt || c.expiresAt > now);
    }
    if (q.sort === "oldest") {
      results.sort((a, b) => a.issuedAt.localeCompare(b.issuedAt));
    } else {
      results.sort((a, b) => b.issuedAt.localeCompare(a.issuedAt));
    }
    return results.slice(0, limit);
  }
  delete(id) {
    return this.claims.delete(id);
  }
  /** GDPR Article 17 — delete all claims for a subject. Returns count deleted. */
  deleteBySubject(subject) {
    let count = 0;
    for (const [id, claim] of this.claims) {
      if (claim.subject === subject) {
        this.claims.delete(id);
        count++;
      }
    }
    return count;
  }
  stats() {
    const now = (/* @__PURE__ */ new Date()).toISOString();
    const claims = Array.from(this.claims.values());
    const byType = {};
    const subjects = /* @__PURE__ */ new Set();
    const issuers = /* @__PURE__ */ new Set();
    let expired = 0;
    for (const c of claims) {
      byType[c.type] = (byType[c.type] ?? 0) + 1;
      subjects.add(c.subject);
      issuers.add(c.issuer);
      if (c.expiresAt && c.expiresAt <= now) expired++;
    }
    return {
      totalClaims: claims.length,
      claimsByType: byType,
      uniqueSubjects: subjects.size,
      uniqueIssuers: issuers.size,
      expiredClaims: expired
    };
  }
  /** Clear all claims. */
  clear() {
    this.claims.clear();
  }
  computeId(claim) {
    const hash = createHash("sha256");
    hash.update(claim.subject);
    hash.update(claim.type);
    hash.update(claim.issuer);
    hash.update(claim.issuedAt);
    hash.update(JSON.stringify(claim.payload));
    return `claim:${hash.digest("hex").slice(0, 32)}`;
  }
  validateClaim(claim) {
    if (!claim.subject) throw new Error("Claim must have a subject");
    if (!claim.type) throw new Error("Claim must have a type");
    if (!claim.issuer) throw new Error("Claim must have an issuer");
    if (!claim.issuedAt) throw new Error("Claim must have an issuedAt timestamp");
    if (!claim.payload || typeof claim.payload !== "object") {
      throw new Error("Claim must have an object payload");
    }
  }
};
function claimFromMarker(marker) {
  return {
    id: `claim:marker:${marker.id}`,
    subject: marker.subject,
    type: "exit_marker" /* ExitMarker */,
    payload: {
      markerId: marker.id,
      origin: marker.origin,
      exitType: marker.exitType,
      status: marker.status,
      selfAttested: marker.selfAttested
    },
    issuer: marker.subject,
    issuedAt: marker.timestamp,
    markerRef: marker.id
  };
}
function claimsFromTrustEnhancers(marker) {
  const claims = [];
  const te = marker.trustEnhancers;
  if (!te) return claims;
  if (te.timestamps) {
    for (const ts of te.timestamps) {
      claims.push({
        id: `claim:tsa:${ts.hash.slice(0, 16)}`,
        subject: marker.subject,
        type: "timestamp" /* Timestamp */,
        payload: { tsaUrl: ts.tsaUrl, hash: ts.hash, timestamp: ts.timestamp },
        issuer: ts.tsaUrl,
        issuedAt: ts.timestamp,
        markerRef: marker.id
      });
    }
  }
  if (te.witnesses) {
    for (const w of te.witnesses) {
      claims.push({
        id: `claim:witness:${w.witnessDid.slice(-16)}:${marker.id.slice(-8)}`,
        subject: marker.subject,
        type: "witness_attestation" /* WitnessAttestation */,
        payload: { attestation: w.attestation, signatureType: w.signatureType },
        issuer: w.witnessDid,
        issuedAt: w.timestamp,
        markerRef: marker.id
      });
    }
  }
  if (te.identityClaims) {
    for (const ic of te.identityClaims) {
      claims.push({
        id: `claim:identity:${ic.scheme}:${ic.value.slice(-16)}`,
        subject: marker.subject,
        type: "identity" /* Identity */,
        payload: { scheme: ic.scheme, value: ic.value, issuer: ic.issuer },
        issuer: ic.issuer ?? marker.subject,
        issuedAt: ic.issuedAt,
        expiresAt: ic.expiresAt,
        markerRef: marker.id
      });
    }
  }
  return claims;
}
function ingestMarker(store, marker) {
  store.put(claimFromMarker(marker));
  for (const claim of claimsFromTrustEnhancers(marker)) {
    store.put(claim);
  }
}

// src/amendment.ts
import { sha256 as sha2566 } from "@noble/hashes/sha256";
var AMENDMENT_PREFIX = "exit-amendment-v1.2:";
var REVOCATION_PREFIX = "exit-revocation-v1.2:";
function hexHash(data) {
  const hash = sha2566(data);
  return Array.from(hash).map((b) => b.toString(16).padStart(2, "0")).join("");
}
async function createAmendment(originalMarker, amendedFields, reason, signer) {
  const timestamp = (/* @__PURE__ */ new Date()).toISOString();
  const unsigned = {
    type: "MarkerAmendment",
    originalMarkerId: originalMarker.id,
    amendedFields,
    reason,
    timestamp
  };
  const canonical = canonicalize(unsigned);
  const amendmentId = hexHash(new TextEncoder().encode(canonical));
  const data = new TextEncoder().encode(AMENDMENT_PREFIX + canonical);
  const signature = await signer.sign(data);
  const proofValue = Buffer.from(signature).toString("base64");
  return {
    amendmentId,
    ...unsigned,
    proof: {
      type: proofTypeForAlgorithm(signer.algorithm),
      created: timestamp,
      verificationMethod: signer.did(),
      proofValue
    }
  };
}
async function createRevocation(targetMarkerId, reason, signer) {
  const timestamp = (/* @__PURE__ */ new Date()).toISOString();
  const unsigned = {
    type: "MarkerRevocation",
    targetMarkerId,
    reason,
    timestamp
  };
  const canonical = canonicalize(unsigned);
  const revocationId = hexHash(new TextEncoder().encode(canonical));
  const data = new TextEncoder().encode(REVOCATION_PREFIX + canonical);
  const signature = await signer.sign(data);
  const proofValue = Buffer.from(signature).toString("base64");
  return {
    revocationId,
    ...unsigned,
    proof: {
      type: proofTypeForAlgorithm(signer.algorithm),
      created: timestamp,
      verificationMethod: signer.did(),
      proofValue
    }
  };
}
function applyAmendment(marker, amendment) {
  return { ...marker, ...amendment.amendedFields };
}
function applyAmendments(marker, amendments) {
  const sorted = [...amendments].sort(
    (a, b) => new Date(a.timestamp).getTime() - new Date(b.timestamp).getTime()
  );
  return sorted.reduce((m, a) => applyAmendment(m, a), marker);
}
function resolveMarker(marker, amendments, revocations) {
  const resolved = applyAmendments(marker, amendments);
  const isRevoked = revocations.some((r) => r.targetMarkerId === marker.id);
  return {
    marker: resolved,
    isRevoked,
    amendmentCount: amendments.length
  };
}
async function verifyAmendmentSignature(amendment, signer) {
  if (!amendment.proof) return false;
  if (amendment.proof.verificationMethod !== signer.did()) return false;
  const { amendmentId: _id, proof: _proof, ...unsigned } = amendment;
  const canonical = canonicalize(unsigned);
  const data = new TextEncoder().encode(AMENDMENT_PREFIX + canonical);
  const signature = new Uint8Array(Buffer.from(amendment.proof.proofValue, "base64"));
  return signer.verify(data, signature);
}
async function verifyRevocationSignature(revocation, signer) {
  if (!revocation.proof) return false;
  if (revocation.proof.verificationMethod !== signer.did()) return false;
  const { revocationId: _id, proof: _proof, ...unsigned } = revocation;
  const canonical = canonicalize(unsigned);
  const data = new TextEncoder().encode(REVOCATION_PREFIX + canonical);
  const signature = new Uint8Array(Buffer.from(revocation.proof.proofValue, "base64"));
  return signer.verify(data, signature);
}

// src/sqlite-store.ts
var Database;
try {
  Database = __require("better-sqlite3");
} catch {
}
var SqliteClaimStore = class {
  db;
  constructor(filePath = ":memory:") {
    if (!Database) {
      throw new Error(
        "better-sqlite3 is required for SqliteClaimStore. Install it: npm install better-sqlite3"
      );
    }
    this.db = new Database(filePath);
    this.db.pragma("journal_mode = WAL");
    this.db.pragma("busy_timeout = 5000");
    this.init();
  }
  init() {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS claims (
        id TEXT PRIMARY KEY,
        subject TEXT NOT NULL,
        type TEXT NOT NULL,
        issuer TEXT NOT NULL,
        issued_at TEXT NOT NULL,
        expires_at TEXT,
        marker_ref TEXT,
        tags_json TEXT,
        payload_json TEXT NOT NULL,
        claim_json TEXT NOT NULL
      );
      CREATE INDEX IF NOT EXISTS idx_claims_subject ON claims(subject);
      CREATE INDEX IF NOT EXISTS idx_claims_type ON claims(type);
      CREATE INDEX IF NOT EXISTS idx_claims_issuer ON claims(issuer);
      CREATE INDEX IF NOT EXISTS idx_claims_issued_at ON claims(issued_at);
      CREATE INDEX IF NOT EXISTS idx_claims_marker_ref ON claims(marker_ref);
    `);
  }
  put(claim) {
    const stmt = this.db.prepare(`
      INSERT OR REPLACE INTO claims (id, subject, type, issuer, issued_at, expires_at, marker_ref, tags_json, payload_json, claim_json)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);
    stmt.run(
      claim.id,
      claim.subject,
      claim.type,
      claim.issuer,
      claim.issuedAt,
      claim.expiresAt ?? null,
      claim.markerRef ?? null,
      claim.tags ? JSON.stringify(claim.tags) : null,
      JSON.stringify(claim.payload),
      JSON.stringify(claim)
    );
  }
  get(id) {
    const row = this.db.prepare("SELECT claim_json FROM claims WHERE id = ?").get(id);
    return row ? JSON.parse(row.claim_json) : null;
  }
  query(q) {
    const excludeExpired = q.excludeExpired !== false;
    const limit = q.limit ?? 100;
    const sort = q.sort === "oldest" ? "ASC" : "DESC";
    const conditions = [];
    const params = [];
    if (q.subject) {
      conditions.push("subject = ?");
      params.push(q.subject);
    }
    if (q.type) {
      conditions.push("type = ?");
      params.push(q.type);
    }
    if (q.issuer) {
      conditions.push("issuer = ?");
      params.push(q.issuer);
    }
    if (q.markerRef) {
      conditions.push("marker_ref = ?");
      params.push(q.markerRef);
    }
    if (excludeExpired) {
      conditions.push("(expires_at IS NULL OR expires_at > ?)");
      params.push((/* @__PURE__ */ new Date()).toISOString());
    }
    const where = conditions.length ? "WHERE " + conditions.join(" AND ") : "";
    const sql = `SELECT claim_json FROM claims ${where} ORDER BY issued_at ${sort} LIMIT ?`;
    params.push(limit);
    const rows = this.db.prepare(sql).all(...params);
    let results = rows.map((r) => JSON.parse(r.claim_json));
    if (q.tags?.length) {
      results = results.filter((c) => c.tags?.some((t) => q.tags.includes(t)));
    }
    return results;
  }
  delete(id) {
    const result = this.db.prepare("DELETE FROM claims WHERE id = ?").run(id);
    return result.changes > 0;
  }
  deleteBySubject(subject) {
    const result = this.db.prepare("DELETE FROM claims WHERE subject = ?").run(subject);
    return result.changes;
  }
  stats() {
    const now = (/* @__PURE__ */ new Date()).toISOString();
    const total = this.db.prepare("SELECT COUNT(*) as c FROM claims").get().c;
    const typeRows = this.db.prepare("SELECT type, COUNT(*) as c FROM claims GROUP BY type").all();
    const byType = {};
    for (const r of typeRows) byType[r.type] = r.c;
    const subjects = this.db.prepare("SELECT COUNT(DISTINCT subject) as c FROM claims").get().c;
    const issuers = this.db.prepare("SELECT COUNT(DISTINCT issuer) as c FROM claims").get().c;
    const expired = this.db.prepare("SELECT COUNT(*) as c FROM claims WHERE expires_at IS NOT NULL AND expires_at <= ?").get(now).c;
    return { totalClaims: total, claimsByType: byType, uniqueSubjects: subjects, uniqueIssuers: issuers, expiredClaims: expired };
  }
  /** Close the database connection. */
  close() {
    this.db.close();
  }
};

// src/telemetry.ts
var PACKAGE_VERSION = "0.2.0";
var noopSpan = {
  setAttribute() {
  },
  setStatus() {
  },
  end() {
  }
};
var noopTracer = {
  startSpan() {
    return noopSpan;
  }
};
var _tracer = noopTracer;
var _config = {};
function initTelemetry(config) {
  _config = config ?? {};
  if (config?.tracer) {
    _tracer = config.tracer;
  } else {
    try {
      const otelApi = __require("@opentelemetry/api");
      _tracer = otelApi.trace.getTracer("cellar-door-exit", PACKAGE_VERSION);
    } catch {
      _tracer = noopTracer;
    }
  }
}
function resetTelemetry() {
  _tracer = noopTracer;
  _config = {};
}
var SpanStatusCode = { UNSET: 0, OK: 1, ERROR: 2 };
function startExitSpan(name, attrs) {
  return _tracer.startSpan(`cellar-door.${name}`, {
    attributes: {
      "cellar_door.version": PACKAGE_VERSION,
      ...attrs
    }
  });
}
function withSpan(name, attrs, fn) {
  const span = startExitSpan(name, attrs);
  try {
    const result = fn();
    span.setStatus({ code: SpanStatusCode.OK });
    return result;
  } catch (err) {
    span.setStatus({ code: SpanStatusCode.ERROR, message: err.message });
    throw err;
  } finally {
    span.end();
  }
}
async function withSpanAsync(name, attrs, fn) {
  const span = startExitSpan(name, attrs);
  try {
    const result = await fn();
    span.setStatus({ code: SpanStatusCode.OK });
    return result;
  } catch (err) {
    span.setStatus({ code: SpanStatusCode.ERROR, message: err.message });
    throw err;
  } finally {
    span.end();
  }
}
function instrumentedSignMarker(marker, privateKey, publicKey) {
  const attrs = {
    "cellar_door.operation": "sign",
    "cellar_door.exit_type": marker.exitType,
    "cellar_door.algorithm": marker.proof?.type ?? "Ed25519Signature2020"
  };
  if (_config.includeMarkerIds && marker.id) attrs["cellar_door.marker_id"] = marker.id;
  if (_config.includeSubjects && marker.subject) attrs["cellar_door.subject"] = marker.subject;
  return withSpan("sign", attrs, () => signMarker(marker, privateKey, publicKey));
}
function instrumentedVerifyMarker(marker) {
  const attrs = {
    "cellar_door.operation": "verify",
    "cellar_door.proof_type": marker.proof?.type ?? "unknown"
  };
  if (_config.includeMarkerIds && marker.id) attrs["cellar_door.marker_id"] = marker.id;
  return withSpan("verify", attrs, () => {
    const result = verifyMarker(marker);
    const span = startExitSpan("verify.result", {
      "cellar_door.valid": result.valid,
      "cellar_door.error_count": result.errors.length
    });
    span.end();
    return result;
  });
}
function startCeremonySpan(subject, origin, exitType) {
  const attrs = {
    "cellar_door.operation": "ceremony",
    "cellar_door.exit_type": exitType
  };
  if (_config.includeSubjects) attrs["cellar_door.subject"] = subject;
  attrs["cellar_door.origin"] = origin;
  return startExitSpan("ceremony", attrs);
}
var instrumentedSignDepartureMarker = instrumentedSignMarker;
var instrumentedVerifyDeparture = instrumentedVerifyMarker;

// src/discovery.ts
var InMemoryAmendmentStore = class {
  amendments = /* @__PURE__ */ new Map();
  revocations = /* @__PURE__ */ new Map();
  async putAmendment(amendment) {
    const markerId = amendment.originalMarkerId;
    if (!this.amendments.has(markerId)) {
      this.amendments.set(markerId, /* @__PURE__ */ new Map());
    }
    this.amendments.get(markerId).set(amendment.amendmentId, amendment);
  }
  async putRevocation(revocation) {
    const markerId = revocation.targetMarkerId;
    if (!this.revocations.has(markerId)) {
      this.revocations.set(markerId, /* @__PURE__ */ new Map());
    }
    this.revocations.get(markerId).set(revocation.revocationId, revocation);
  }
  async getAmendments(markerId, since) {
    const map = this.amendments.get(markerId);
    if (!map) return [];
    let amendments = Array.from(map.values());
    if (since) {
      const sinceTime = new Date(since).getTime();
      amendments = amendments.filter((a) => new Date(a.timestamp).getTime() > sinceTime);
    }
    return amendments.sort(
      (a, b) => new Date(a.timestamp).getTime() - new Date(b.timestamp).getTime()
    );
  }
  async getRevocations(markerId) {
    const map = this.revocations.get(markerId);
    if (!map) return [];
    return Array.from(map.values());
  }
  async query(markerId, since) {
    return {
      markerId,
      amendments: await this.getAmendments(markerId, since),
      revocations: await this.getRevocations(markerId),
      queriedAt: (/* @__PURE__ */ new Date()).toISOString(),
      complete: true
    };
  }
};
async function discoverAmendments(markerId, options = {}) {
  const { markerDiscoveryUrl, wellKnownUrl, localStore, timeoutMs = 5e3, since } = options;
  if (markerDiscoveryUrl) {
    try {
      const result = await fetchAmendments(markerDiscoveryUrl, markerId, timeoutMs, since);
      if (result) return result;
    } catch {
    }
  }
  if (wellKnownUrl) {
    try {
      const result = await fetchAmendments(wellKnownUrl, markerId, timeoutMs, since);
      if (result) return result;
    } catch {
    }
  }
  if (localStore) {
    try {
      return await localStore.query(markerId, since);
    } catch {
    }
  }
  return {
    markerId,
    amendments: [],
    revocations: [],
    queriedAt: (/* @__PURE__ */ new Date()).toISOString(),
    complete: false
  };
}
async function fetchAmendments(baseUrl, markerId, timeoutMs, since) {
  const url = new URL(`${baseUrl.replace(/\/$/, "")}/${encodeURIComponent(markerId)}`);
  if (since) url.searchParams.set("since", since);
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const response = await fetch(url.toString(), {
      method: "GET",
      headers: { Accept: "application/json" },
      signal: controller.signal
    });
    if (!response.ok) return null;
    const data = await response.json();
    if (!data.markerId || !Array.isArray(data.amendments) || !Array.isArray(data.revocations)) {
      return null;
    }
    return data;
  } finally {
    clearTimeout(timeout);
  }
}
function createWellKnownMetadata(endpointUrl, capabilities = {}) {
  return {
    version: "1.2",
    endpoint: endpointUrl,
    capabilities: {
      byMarkerId: true,
      bySubject: capabilities.bySubject ?? false,
      since: capabilities.since ?? true,
      ...capabilities
    },
    lastUpdated: (/* @__PURE__ */ new Date()).toISOString()
  };
}
function createAmendmentQueryHandler(store) {
  return async (markerId, since) => {
    return store.query(markerId, since);
  };
}

export {
  generateKeyPair,
  sign2 as sign,
  verify2 as verify,
  didFromPublicKey,
  publicKeyFromDid,
  generateP256KeyPair,
  signP256,
  verifyP256,
  didFromP256PublicKey,
  publicKeyFromP256Did,
  AlgorithmRegistry,
  getAlgorithm,
  algorithmFromDid,
  Ed25519Signer,
  P256Signer,
  generateKeyPairForAlgorithm,
  createSigner,
  createVerifier,
  proofTypeForAlgorithm,
  algorithmFromProofType,
  validateMarker,
  signMarker,
  verifyMarker,
  signMarkerWithSigner,
  verifyMarkerMultiAlg,
  verifyTrustEnhancers,
  getValidTransitions,
  CeremonyStateMachine,
  addCounterSignature,
  addWitness,
  verifyCounterSignature,
  deriveStatusConfirmation,
  generateIdentity,
  quickExit,
  quickExitP256,
  quickVerify,
  fromJSON,
  quickCounterSign,
  toJSON,
  EXIT_CONTEXT,
  getContext,
  VC_CONTEXT,
  wrapAsVC,
  unwrapVC,
  isVC,
  createLineageModule,
  verifyLineageChain,
  bindSuccessor,
  createReputationModule,
  signEndorsement,
  createOriginAttestation,
  signAttestation,
  createAssetManifest,
  createContinuityModule,
  createDisputeModule,
  computeStatusConfirmation,
  createTenureAttestation,
  verifyTenureAttestation,
  parseDurationToDays,
  createCommitment,
  verifyCommitment,
  isRevealWindowOpen,
  extractConfidenceFactors,
  computeConfidence,
  saveMarker,
  loadMarker,
  listMarkers,
  deleteMarker,
  exportMarker,
  saveAmendment,
  loadAmendments,
  saveRevocation,
  loadRevocations,
  importMarker,
  MockChainAdapter,
  FileChainAdapter,
  encryptMarker,
  decryptMarker,
  redactMarker,
  createMinimalDisclosure,
  InMemoryKeyStore,
  encryptMarkerWithManagedKey,
  decryptManagedMarker,
  deleteMarkerKey,
  isDid,
  didMethod,
  resolveDid,
  createDidDocument,
  resolveDidKeri,
  createDidKeri,
  MarkerRegistry,
  computeMerkleRoot,
  computeMerkleProof,
  verifyBatchMembership,
  createBatchExit,
  createShutdownBatch,
  createExitMiddleware,
  createExitHook,
  ExitEventEmitter,
  serializeForTransport,
  deserializeFromTransport,
  digestKey,
  createInception,
  createRotation,
  verifyKeyState,
  isKeyCompromised,
  KeyEventLog,
  generatePreRotatedKeys,
  commitNextKey,
  verifyNextKeyCommitment,
  rotateKeys,
  detectCoercion,
  detectWeaponization,
  detectReputationLaundering,
  detectRetaliationWindow,
  generateEthicsReport,
  ANTI_WEAPONIZATION_CLAUSE,
  addCoercionLabel,
  addRightOfReply,
  validateEthicalCompliance,
  applySunset,
  isExpired,
  departAndAnchor,
  departAndVerify,
  flagCompromisedPlatformMarkers,
  createCompromiseMarker,
  verifyCompromiseRecovery,
  linkCompromisedMarkers,
  createDispute,
  resolveDispute,
  verifyDisputeResolution,
  isDisputed,
  getDisputeStatus,
  createDepartureMarker,
  signDepartureMarker,
  signDepartureWithSigner,
  verifyDeparture,
  verifyDepartureMultiAlg,
  quickDeparture,
  quickPassageVerify,
  generatePassageIdentity,
  createPassage,
  verifyPassage,
  ClaimType,
  MemoryClaimStore,
  claimFromMarker,
  claimsFromTrustEnhancers,
  ingestMarker,
  createAmendment,
  createRevocation,
  applyAmendment,
  applyAmendments,
  resolveMarker,
  verifyAmendmentSignature,
  verifyRevocationSignature,
  SqliteClaimStore,
  initTelemetry,
  resetTelemetry,
  startExitSpan,
  withSpan,
  withSpanAsync,
  instrumentedSignMarker,
  instrumentedVerifyMarker,
  startCeremonySpan,
  instrumentedSignDepartureMarker,
  instrumentedVerifyDeparture,
  InMemoryAmendmentStore,
  discoverAmendments,
  createWellKnownMetadata,
  createAmendmentQueryHandler
};
