// cellar-door-exit/dist/cli.cjs
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// src/types.ts
var ExitType, ExitStatus, EXIT_CONTEXT_V1, EXIT_SPEC_VERSION;
var init_types = __esm({
  "src/types.ts"() {
    "use strict";
    ExitType = /* @__PURE__ */ ((ExitType2) => {
      ExitType2["Voluntary"] = "voluntary";
      ExitType2["Forced"] = "forced";
      ExitType2["Emergency"] = "emergency";
      ExitType2["KeyCompromise"] = "keyCompromise";
      ExitType2["PlatformShutdown"] = "platform_shutdown";
      ExitType2["Directed"] = "directed";
      ExitType2["Constructive"] = "constructive";
      ExitType2["Acquisition"] = "acquisition";
      return ExitType2;
    })(ExitType || {});
    ExitStatus = /* @__PURE__ */ ((ExitStatus4) => {
      ExitStatus4["GoodStanding"] = "good_standing";
      ExitStatus4["Disputed"] = "disputed";
      ExitStatus4["Unverified"] = "unverified";
      return ExitStatus4;
    })(ExitStatus || {});
    EXIT_CONTEXT_V1 = "https://cellar-door.dev/exit/v1";
    EXIT_SPEC_VERSION = "1.2";
  }
});

// src/errors.ts
var ExitError, ValidationError, SigningError;
var init_errors = __esm({
  "src/errors.ts"() {
    "use strict";
    ExitError = class extends Error {
      code;
      /**
       * @param code - The structured error code.
       * @param message - Human-readable error message.
       */
      constructor(code, message) {
        super(message);
        this.name = "ExitError";
        this.code = code;
        Object.setPrototypeOf(this, new.target.prototype);
      }
    };
    ValidationError = class extends ExitError {
      errors;
      /**
       * @param errors - Array of validation error messages.
       * @param message - Optional custom message (defaults to a summary of errors).
       */
      constructor(errors, message) {
        const msg = message ?? `Marker validation failed: ${errors.join("; ")}`;
        super("VALIDATION_FAILED", msg);
        this.name = "ValidationError";
        this.errors = errors;
      }
    };
    SigningError = class extends ExitError {
      constructor(message) {
        super("SIGNING_FAILED", message);
        this.name = "SigningError";
      }
    };
  }
});

// src/marker.ts
function createMarker(opts) {
  const errors = [];
  if (!opts.subject || typeof opts.subject !== "string") errors.push("subject is required and must be a non-empty string");
  if (!opts.origin || typeof opts.origin !== "string") errors.push("origin is required and must be a non-empty string");
  if (!Object.values(ExitType).includes(opts.exitType)) errors.push(`exitType must be one of: ${Object.values(ExitType).join(", ")}`);
  if (opts.exitType === "emergency" /* Emergency */ && (!opts.emergencyJustification || typeof opts.emergencyJustification !== "string")) {
    errors.push("emergencyJustification is required when exitType is 'emergency'");
  }
  if (errors.length > 0) throw new ValidationError(errors);
  const timestamp = opts.timestamp ?? (/* @__PURE__ */ new Date()).toISOString();
  const marker = {
    "@context": EXIT_CONTEXT_V1,
    specVersion: EXIT_SPEC_VERSION,
    id: opts.id ?? "",
    subject: opts.subject,
    origin: opts.origin,
    timestamp,
    exitType: opts.exitType,
    status: opts.status ?? defaultStatus(opts.exitType),
    proof: opts.proof ?? { ...EMPTY_PROOF, created: timestamp },
    selfAttested: opts.selfAttested ?? true
  };
  if (opts.emergencyJustification) {
    marker.emergencyJustification = opts.emergencyJustification;
  }
  if (!marker.expires) {
    const defaultDays = opts.exitType === "voluntary" /* Voluntary */ ? 730 : 365;
    const expiresDate = new Date(new Date(timestamp).getTime() + defaultDays * 864e5);
    marker.expires = expiresDate.toISOString();
  }
  if (!marker.id) {
    marker.id = `urn:exit:${computeId(marker)}`;
  }
  return marker;
}
function defaultStatus(exitType) {
  switch (exitType) {
    case "voluntary" /* Voluntary */:
      return "good_standing" /* GoodStanding */;
    case "forced" /* Forced */:
    case "directed" /* Directed */:
    case "constructive" /* Constructive */:
      return "disputed" /* Disputed */;
    case "emergency" /* Emergency */:
    case "keyCompromise" /* KeyCompromise */:
    case "platform_shutdown" /* PlatformShutdown */:
    case "acquisition" /* Acquisition */:
      return "unverified" /* Unverified */;
    default:
      return "unverified" /* Unverified */;
  }
}
function canonicalize(obj) {
  const result = (0, import_canonicalize.default)(obj);
  if (result === void 0) return "null";
  return result;
}
function computeId(marker) {
  const { proof: _proof, id: _id, ...rest } = marker;
  const canonical = canonicalize(rest);
  const hash = (0, import_sha256.sha256)(new TextEncoder().encode(canonical));
  return Array.from(hash).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function addModule(marker, key, module2) {
  const updated = { ...marker, [key]: module2 };
  updated.id = `urn:exit:${computeId(updated)}`;
  return updated;
}
var import_sha256, import_canonicalize, EMPTY_PROOF;
var init_marker = __esm({
  "src/marker.ts"() {
    "use strict";
    import_sha256 = require("@noble/hashes/sha256");
    import_canonicalize = __toESM(require("canonicalize"), 1);
    init_types();
    init_errors();
    EMPTY_PROOF = {
      type: "Ed25519Signature2020",
      created: "",
      verificationMethod: "",
      proofValue: ""
    };
  }
});

// src/anchor.ts
function toHex(bytes) {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function computeAnchorHash(marker) {
  const canonical = canonicalize(marker);
  return toHex((0, import_sha2564.sha256)(new TextEncoder().encode(canonical)));
}
function createAnchorRecord(marker, options) {
  const minimal = options?.minimal ?? true;
  const base = {
    hash: computeAnchorHash(marker),
    timestamp: marker.timestamp,
    specVersion: EXIT_SPEC_VERSION
  };
  if (!minimal) {
    return {
      ...base,
      exitType: marker.exitType,
      subjectDid: marker.subject
    };
  }
  return base;
}
var import_sha2564;
var init_anchor = __esm({
  "src/anchor.ts"() {
    "use strict";
    import_sha2564 = require("@noble/hashes/sha256");
    init_marker();
    init_types();
  }
});

// src/git-ledger.ts
var import_node_child_process, import_node_util, import_promises, import_node_path2, execFile;
var init_git_ledger = __esm({
  "src/git-ledger.ts"() {
    "use strict";
    import_node_child_process = require("child_process");
    import_node_util = require("util");
    import_promises = require("fs/promises");
    import_node_path2 = require("path");
    init_anchor();
    execFile = (0, import_node_util.promisify)(import_node_child_process.execFile);
  }
});

// src/tsa.ts
var import_node_crypto2, SHA256_OID;
var init_tsa = __esm({
  "src/tsa.ts"() {
    "use strict";
    import_node_crypto2 = require("crypto");
    init_anchor();
    SHA256_OID = Buffer.from([
      96,
      134,
      72,
      1,
      101,
      3,
      4,
      2,
      1
    ]);
  }
});

// src/visual.ts
var init_visual = __esm({
  "src/visual.ts"() {
    "use strict";
  }
});

// src/cli.ts
var import_commander = require("commander");
var import_node_fs3 = require("fs");

// src/index.ts
init_types();

// src/crypto.ts
var import_node_crypto = require("crypto");
var ed = __toESM(require("@noble/ed25519"), 1);
var import_sha512 = require("@noble/hashes/sha512");
var import_nist = require("@noble/curves/nist.js");
if (!globalThis.crypto) {
  globalThis.crypto = import_node_crypto.webcrypto;
}
ed.etc.sha512Sync = (...m) => {
  const h = import_sha512.sha512.create();
  for (const msg of m) h.update(msg);
  return h.digest();
};
var ED25519_MULTICODEC = new Uint8Array([237, 1]);
var P256_MULTICODEC = new Uint8Array([128, 36]);
function generateKeyPair() {
  const privateKey = ed.utils.randomPrivateKey();
  const publicKey = ed.getPublicKey(privateKey);
  return { publicKey, privateKey };
}
function sign2(data, privateKey) {
  return ed.sign(data, privateKey);
}
function verify2(data, signature, publicKey) {
  try {
    return ed.verify(signature, data, publicKey);
  } catch {
    return false;
  }
}
function base58btcEncode(bytes) {
  const ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
  let num = 0n;
  for (const b of bytes) num = num * 256n + BigInt(b);
  let encoded = "";
  while (num > 0n) {
    encoded = ALPHABET[Number(num % 58n)] + encoded;
    num /= 58n;
  }
  for (const b of bytes) {
    if (b === 0) encoded = "1" + encoded;
    else break;
  }
  return encoded || "1";
}
function base58btcDecode(str) {
  const ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
  let num = 0n;
  for (const c of str) {
    const idx = ALPHABET.indexOf(c);
    if (idx === -1) throw new Error(`Invalid base58 character: ${c}`);
    num = num * 58n + BigInt(idx);
  }
  const hex = num.toString(16).padStart(2, "0");
  const paddedHex = hex.length % 2 ? "0" + hex : hex;
  const bytes = [];
  for (let i = 0; i < paddedHex.length; i += 2) {
    bytes.push(parseInt(paddedHex.slice(i, i + 2), 16));
  }
  let leadingZeros = 0;
  for (const c of str) {
    if (c === "1") leadingZeros++;
    else break;
  }
  const result = new Uint8Array(leadingZeros + bytes.length);
  result.set(new Uint8Array(bytes), leadingZeros);
  return result;
}
function didFromPublicKey(publicKey) {
  const multicodec = new Uint8Array(ED25519_MULTICODEC.length + publicKey.length);
  multicodec.set(ED25519_MULTICODEC);
  multicodec.set(publicKey, ED25519_MULTICODEC.length);
  return `did:key:z${base58btcEncode(multicodec)}`;
}
function publicKeyFromDid(did) {
  if (!did.startsWith("did:key:z")) {
    throw new Error("Invalid did:key format: must start with 'did:key:z'");
  }
  const encoded = did.slice("did:key:z".length);
  const decoded = base58btcDecode(encoded);
  if (decoded[0] === 237 && decoded[1] === 1) {
    const key = decoded.slice(2);
    if (key.length !== 32) {
      throw new Error(`Invalid Ed25519 public key length: expected 32 bytes, got ${key.length}`);
    }
    return key;
  }
  if (decoded[0] === 128 && decoded[1] === 36) {
    const key = decoded.slice(2);
    if (key.length !== 33) {
      throw new Error(`Invalid P-256 public key length: expected 33 bytes, got ${key.length}`);
    }
    return key;
  }
  throw new Error("Invalid multicodec prefix: expected Ed25519 (0xed01) or P-256 (0x8024)");
}
function generateP256KeyPair() {
  const privateKey = import_nist.p256.utils.randomSecretKey();
  const publicKey = import_nist.p256.getPublicKey(privateKey, true);
  return { publicKey, privateKey };
}
function signP256(data, privateKey) {
  const sig = import_nist.p256.sign(data, privateKey, { prehash: true });
  return Uint8Array.from(sig);
}
function verifyP256(data, signature, publicKey) {
  try {
    return import_nist.p256.verify(signature, data, publicKey, { prehash: true });
  } catch {
    return false;
  }
}
function didFromP256PublicKey(publicKey) {
  const multicodec = new Uint8Array(P256_MULTICODEC.length + publicKey.length);
  multicodec.set(P256_MULTICODEC);
  multicodec.set(publicKey, P256_MULTICODEC.length);
  return `did:key:z${base58btcEncode(multicodec)}`;
}
function publicKeyFromP256Did(did) {
  if (!did.startsWith("did:key:z")) {
    throw new Error("Invalid did:key format: must start with 'did:key:z'");
  }
  const encoded = did.slice("did:key:z".length);
  const decoded = base58btcDecode(encoded);
  if (decoded[0] !== 128 || decoded[1] !== 36) {
    throw new Error("Invalid multicodec prefix: expected P-256 (0x8024)");
  }
  return decoded.slice(2);
}
var AlgorithmRegistry = /* @__PURE__ */ new Map();
AlgorithmRegistry.set("ed25519", {
  generateKeyPair,
  sign: sign2,
  verify: verify2,
  didFromPublicKey,
  publicKeyFromDid
});
AlgorithmRegistry.set("p256", {
  generateKeyPair: generateP256KeyPair,
  sign: signP256,
  verify: verifyP256,
  didFromPublicKey: didFromP256PublicKey,
  publicKeyFromDid: publicKeyFromP256Did
});
function algorithmFromDid(did) {
  if (!did.startsWith("did:key:z")) {
    throw new Error("Invalid did:key format");
  }
  const encoded = did.slice("did:key:z".length);
  const decoded = base58btcDecode(encoded);
  if (decoded[0] === 237 && decoded[1] === 1) return "Ed25519";
  if (decoded[0] === 128 && decoded[1] === 36) return "P-256";
  throw new Error("Unknown multicodec prefix");
}

// src/signer.ts
function algorithmFromProofType(proofType) {
  if (proofType === "Ed25519Signature2020") return "Ed25519";
  if (proofType === "EcdsaP256Signature2019") return "P-256";
  return null;
}

// src/index.ts
init_marker();

// src/proof.ts
init_marker();

// src/validate.ts
init_types();
var ISO_8601_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z?$/;
var DID_KEY_RE = /^did:key:z[1-9A-HJ-NP-Za-km-z]+$/;
var MAX_URI_LENGTH = 2048;
var MAX_LABEL_LENGTH = 256;
var MAX_STRING_LENGTH = 4096;
var HEX_HASH_RE = /^[0-9a-f]{64}$/i;
function isValidISO8601(s) {
  return ISO_8601_RE.test(s) && !isNaN(Date.parse(s));
}
function isValidDid(s) {
  return DID_KEY_RE.test(s) || s.startsWith("did:");
}
function containsControlChars(s) {
  return /[\x00-\x08\x0b\x0c\x0e-\x1f]/.test(s);
}
function validateMarker(marker) {
  const errors = [];
  const m = marker;
  if (!m || typeof m !== "object") {
    return { valid: false, errors: ["Marker must be an object"] };
  }
  try {
    const serialized = JSON.stringify(m);
    if (new TextEncoder().encode(serialized).byteLength > 8192) {
      return { valid: false, errors: ["Marker exceeds maximum size of 8192 bytes"] };
    }
  } catch {
    return { valid: false, errors: ["Marker cannot be serialized to JSON"] };
  }
  if (m["@context"] !== EXIT_CONTEXT_V1) {
    errors.push(`Invalid @context: expected "${EXIT_CONTEXT_V1}"`);
  }
  if (typeof m.specVersion !== "string" || !m.specVersion) {
    errors.push("Missing or invalid required field: specVersion");
  } else if (m.specVersion !== EXIT_SPEC_VERSION) {
    errors.push(`Unsupported specVersion: expected "${EXIT_SPEC_VERSION}", got "${m.specVersion}"`);
  }
  if (typeof m.id !== "string" || !m.id) {
    errors.push("Missing or invalid required field: id");
  }
  if (typeof m.subject !== "string" || !m.subject) {
    errors.push("Missing or invalid required field: subject");
  }
  if (typeof m.origin !== "string" || !m.origin) {
    errors.push("Missing or invalid required field: origin");
  }
  if (typeof m.timestamp !== "string" || !m.timestamp) {
    errors.push("Missing or invalid required field: timestamp");
  } else if (!isValidISO8601(m.timestamp)) {
    errors.push("Invalid timestamp: must be ISO-8601 format");
  }
  if (!Object.values(ExitType).includes(m.exitType)) {
    errors.push(`Invalid exitType: must be one of ${Object.values(ExitType).join(", ")}`);
  }
  if (!Object.values(ExitStatus).includes(m.status)) {
    errors.push(`Invalid status: must be one of ${Object.values(ExitStatus).join(", ")}`);
  }
  if (!m.proof || typeof m.proof !== "object") {
    errors.push("Missing or invalid required field: proof");
  } else {
    const p = m.proof;
    if (typeof p.type !== "string" || !p.type) errors.push("Proof missing type");
    if (typeof p.proofValue !== "string") errors.push("Proof missing proofValue");
    if (typeof p.verificationMethod !== "string") errors.push("Proof missing verificationMethod");
  }
  for (const field of ["subject", "origin", "id"]) {
    if (typeof m[field] === "string" && containsControlChars(m[field])) {
      errors.push(`${field} contains invalid control characters`);
    }
  }
  if (typeof m.subject === "string" && m.subject.startsWith("did:") && !isValidDid(m.subject)) {
    errors.push("Invalid subject DID format");
  }
  if (m.exitType === "emergency" /* Emergency */) {
    if (typeof m.emergencyJustification !== "string" || !m.emergencyJustification) {
      errors.push("emergencyJustification is required when exitType is 'emergency'");
    }
  }
  if (m.legalHold !== void 0) {
    if (typeof m.legalHold !== "object" || m.legalHold === null) {
      errors.push("legalHold must be an object");
    } else {
      const lh = m.legalHold;
      if (typeof lh.holdType !== "string" || !lh.holdType) errors.push("legalHold.holdType is required");
      if (typeof lh.authority !== "string" || !lh.authority) errors.push("legalHold.authority is required");
      if (typeof lh.reference !== "string" || !lh.reference) errors.push("legalHold.reference is required");
      if (typeof lh.dateIssued !== "string" || !lh.dateIssued) {
        errors.push("legalHold.dateIssued is required");
      } else if (!isValidISO8601(lh.dateIssued)) {
        errors.push("legalHold.dateIssued must be ISO-8601 format");
      }
      if (typeof lh.acknowledged !== "boolean") errors.push("legalHold.acknowledged must be a boolean");
    }
  }
  if (m.expires !== void 0) {
    if (typeof m.expires !== "string" || !isValidISO8601(m.expires)) {
      errors.push("expires must be a valid ISO-8601 date string");
    }
  }
  if (m.sequenceNumber !== void 0) {
    if (typeof m.sequenceNumber !== "number" || !Number.isInteger(m.sequenceNumber) || m.sequenceNumber < 0) {
      errors.push("sequenceNumber must be a non-negative integer");
    }
  }
  if (m.completenessAttestation !== void 0) {
    if (typeof m.completenessAttestation !== "object" || m.completenessAttestation === null) {
      errors.push("completenessAttestation must be an object");
    }
  }
  if (m.disputeExpiry !== void 0) {
    if (typeof m.disputeExpiry !== "string" || !isValidISO8601(m.disputeExpiry)) {
      errors.push("disputeExpiry must be a valid ISO-8601 date string");
    }
  }
  if (m.resolution !== void 0) {
    if (typeof m.resolution !== "string" || !m.resolution) {
      errors.push("resolution must be a non-empty string");
    }
  }
  if (m.arbiterDid !== void 0) {
    if (typeof m.arbiterDid !== "string" || !m.arbiterDid.startsWith("did:")) {
      errors.push('arbiterDid must be a string starting with "did:"');
    }
  }
  if (m.lineage !== void 0 && (typeof m.lineage !== "object" || m.lineage === null)) {
    errors.push("Module A (lineage) must be an object");
  }
  if (m.stateSnapshot !== void 0) {
    const ss = m.stateSnapshot;
    if (typeof ss !== "object" || ss === null) {
      errors.push("Module B (stateSnapshot) must be an object");
    } else if (typeof ss.stateHash !== "string" || !ss.stateHash) {
      errors.push("Module B (stateSnapshot) requires stateHash");
    }
  }
  if (m.trustEnhancers !== void 0) {
    if (typeof m.trustEnhancers !== "object" || m.trustEnhancers === null) {
      errors.push("trustEnhancers must be an object");
    } else {
      const te = m.trustEnhancers;
      if (te.timestamps !== void 0) {
        if (!Array.isArray(te.timestamps)) {
          errors.push("trustEnhancers.timestamps must be an array");
        } else {
          for (let i = 0; i < te.timestamps.length; i++) {
            const ts = te.timestamps[i];
            if (typeof ts.tsaUrl !== "string" || !ts.tsaUrl) {
              errors.push(`trustEnhancers.timestamps[${i}].tsaUrl is required`);
            }
            if (typeof ts.hash !== "string" || !/^[0-9a-f]{64}$/i.test(ts.hash)) {
              errors.push(`trustEnhancers.timestamps[${i}].hash must be a 64-char hex SHA-256`);
            }
            if (typeof ts.timestamp !== "string" || !isValidISO8601(ts.timestamp)) {
              errors.push(`trustEnhancers.timestamps[${i}].timestamp must be ISO-8601`);
            }
            if (typeof ts.receipt !== "string" || !ts.receipt) {
              errors.push(`trustEnhancers.timestamps[${i}].receipt is required`);
            }
          }
        }
      }
      if (te.witnesses !== void 0) {
        if (!Array.isArray(te.witnesses)) {
          errors.push("trustEnhancers.witnesses must be an array");
        } else {
          for (let i = 0; i < te.witnesses.length; i++) {
            const w = te.witnesses[i];
            if (typeof w.witnessDid !== "string" || !w.witnessDid) {
              errors.push(`trustEnhancers.witnesses[${i}].witnessDid is required`);
            }
            if (typeof w.attestation !== "string" || !w.attestation) {
              errors.push(`trustEnhancers.witnesses[${i}].attestation is required`);
            }
            if (typeof w.timestamp !== "string" || !isValidISO8601(w.timestamp)) {
              errors.push(`trustEnhancers.witnesses[${i}].timestamp must be ISO-8601`);
            }
            if (typeof w.signature !== "string" || !w.signature) {
              errors.push(`trustEnhancers.witnesses[${i}].signature is required`);
            }
            if (typeof w.signatureType !== "string" || !w.signatureType) {
              errors.push(`trustEnhancers.witnesses[${i}].signatureType is required`);
            }
          }
        }
      }
      if (te.identityClaims !== void 0) {
        if (!Array.isArray(te.identityClaims)) {
          errors.push("trustEnhancers.identityClaims must be an array");
        } else {
          for (let i = 0; i < te.identityClaims.length; i++) {
            const ic = te.identityClaims[i];
            if (typeof ic.scheme !== "string" || !ic.scheme) {
              errors.push(`trustEnhancers.identityClaims[${i}].scheme is required`);
            }
            if (typeof ic.value !== "string" || !ic.value) {
              errors.push(`trustEnhancers.identityClaims[${i}].value is required`);
            }
            if (typeof ic.issuedAt !== "string" || !isValidISO8601(ic.issuedAt)) {
              errors.push(`trustEnhancers.identityClaims[${i}].issuedAt must be ISO-8601`);
            }
            if (ic.expiresAt !== void 0 && (typeof ic.expiresAt !== "string" || !isValidISO8601(ic.expiresAt))) {
              errors.push(`trustEnhancers.identityClaims[${i}].expiresAt must be ISO-8601`);
            }
          }
        }
      }
    }
  }
  const warnings = [];
  for (const field of ["id", "origin"]) {
    if (typeof m[field] === "string" && m[field].length > MAX_URI_LENGTH) {
      warnings.push(`${field} exceeds recommended maximum length of ${MAX_URI_LENGTH} characters`);
    }
  }
  for (const field of ["subject"]) {
    if (typeof m[field] === "string" && m[field].length > MAX_URI_LENGTH) {
      warnings.push(`${field} exceeds recommended maximum length of ${MAX_URI_LENGTH} characters`);
    }
  }
  if (typeof m.coercionLabel === "string" && m.coercionLabel.length > MAX_LABEL_LENGTH) {
    warnings.push(`coercionLabel exceeds recommended maximum length of ${MAX_LABEL_LENGTH} characters`);
  }
  for (const field of ["reason", "narrative", "emergencyJustification", "resolution"]) {
    if (typeof m[field] === "string" && m[field].length > MAX_STRING_LENGTH) {
      warnings.push(`${field} exceeds recommended maximum length of ${MAX_STRING_LENGTH} characters`);
    }
  }
  if (typeof m.timestamp === "string" && ISO_8601_RE.test(m.timestamp) && !m.timestamp.endsWith("Z")) {
    warnings.push("timestamp missing 'Z' suffix \u2014 treating as UTC (B7)");
  }
  if (m.selfAttested !== void 0 && typeof m.selfAttested !== "boolean") {
    warnings.push("selfAttested should be a boolean");
  }
  if (m.coercionLabel !== void 0) {
    if (typeof m.coercionLabel !== "string") {
      warnings.push("coercionLabel should be a string");
    }
  }
  if (m.preRotationCommitment !== void 0) {
    if (typeof m.preRotationCommitment !== "string") {
      warnings.push("preRotationCommitment should be a string");
    } else if (!HEX_HASH_RE.test(m.preRotationCommitment)) {
      warnings.push("preRotationCommitment should be a 64-character hex SHA-256 hash");
    }
  }
  if (m.expires === void 0) {
    warnings.push("Missing required 'expires' field (\xA78.5). Implementations MUST populate with default: 730 days for voluntary exits, 365 days for involuntary exits.");
  }
  if (m.sunsetDate !== void 0) {
    if (typeof m.sunsetDate !== "string" || !isValidISO8601(m.sunsetDate)) {
      warnings.push("sunsetDate should be a valid ISO 8601 date string");
    }
  }
  if (m.completenessAttestation !== void 0 && typeof m.completenessAttestation === "object" && m.completenessAttestation !== null) {
    const ca = m.completenessAttestation;
    if (ca.attestedAt !== void 0 && (typeof ca.attestedAt !== "string" || !isValidISO8601(ca.attestedAt))) {
      warnings.push("completenessAttestation.attestedAt should be a valid ISO 8601 date");
    }
    if (ca.markerCount !== void 0 && (typeof ca.markerCount !== "number" || !Number.isInteger(ca.markerCount) || ca.markerCount < 0)) {
      warnings.push("completenessAttestation.markerCount should be a non-negative integer");
    }
    if (ca.signature !== void 0 && typeof ca.signature !== "string") {
      warnings.push("completenessAttestation.signature should be a string");
    }
  }
  if (m.intent !== void 0 && typeof m.intent === "object" && m.intent !== null) {
    const intent = m.intent;
    if (typeof intent.subject === "string" && typeof m.subject === "string" && intent.subject !== m.subject) {
      warnings.push("ExitIntent subject does not match marker subject");
    }
  }
  const result = { valid: errors.length === 0, errors };
  if (warnings.length > 0) result.warnings = warnings;
  return result;
}

// src/proof.ts
init_errors();
var DOMAIN_PREFIX = "exit-marker-v1.2:";
function signMarker(marker, privateKey, publicKey) {
  try {
    const did = didFromPublicKey(publicKey);
    const now = (/* @__PURE__ */ new Date()).toISOString();
    const { proof: _proof, id: _id, ...rest } = marker;
    const canonical = canonicalize(rest);
    const data = new TextEncoder().encode(DOMAIN_PREFIX + canonical);
    const signature = sign2(data, privateKey);
    const proofValue = Buffer.from(signature).toString("base64");
    const proof = {
      type: "Ed25519Signature2020",
      created: now,
      verificationMethod: did,
      proofValue
    };
    return { ...marker, proof };
  } catch (e) {
    if (e instanceof SigningError) throw e;
    throw new SigningError(`Failed to sign marker: ${e.message}`);
  }
}
function verifyMarker(marker, options) {
  const errors = [];
  const schema = validateMarker(marker);
  if (!schema.valid) {
    errors.push(...schema.errors);
  }
  if (!marker.proof) {
    errors.push("Missing proof");
    return { valid: false, errors };
  }
  const alg = algorithmFromProofType(marker.proof.type);
  if (!alg) {
    errors.push(`Unsupported proof type: ${marker.proof.type}`);
    return { valid: false, errors };
  }
  if (!marker.proof.verificationMethod || !marker.proof.proofValue) {
    errors.push("Incomplete proof: missing verificationMethod or proofValue");
    return { valid: false, errors };
  }
  if (!marker.proof.created || !/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z?$/.test(marker.proof.created) || isNaN(Date.parse(marker.proof.created))) {
    errors.push("Proof missing or invalid created timestamp (must be ISO 8601)");
  }
  if (marker.subject && marker.proof.verificationMethod !== marker.subject) {
    errors.push("Proof verificationMethod does not match marker subject \u2014 possible attribution forgery");
    return { valid: false, errors };
  }
  const didAlg = algorithmFromDid(marker.proof.verificationMethod);
  if (didAlg !== alg) {
    errors.push(`Algorithm mismatch: proof type indicates ${alg} but DID uses ${didAlg}`);
    return { valid: false, errors };
  }
  try {
    const publicKey = publicKeyFromDid(marker.proof.verificationMethod);
    const { proof: _proof, id: _id, ...rest } = marker;
    const canonical = canonicalize(rest);
    const data = new TextEncoder().encode(DOMAIN_PREFIX + canonical);
    const signature = new Uint8Array(Buffer.from(marker.proof.proofValue, "base64"));
    let valid;
    if (alg === "P-256") {
      valid = verifyP256(data, signature, publicKey);
    } else {
      valid = verify2(data, signature, publicKey);
    }
    if (!valid) {
      errors.push("Signature verification failed");
    }
  } catch (e) {
    errors.push(
      options?.verbose ? `Signature verification error: ${e.message}` : "Verification failed"
    );
  }
  if (!options?.verbose) {
    const genericErrors = [];
    for (const err of errors) {
      if (err === "Signature verification failed" || err.startsWith("Signature verification error:")) {
        if (!genericErrors.includes("Verification failed")) {
          genericErrors.push("Verification failed");
        }
      } else {
        genericErrors.push(err);
      }
    }
    return { valid: genericErrors.length === 0, errors: genericErrors };
  }
  return { valid: errors.length === 0, errors };
}

// src/ceremony.ts
init_types();
init_marker();
init_errors();
var TRANSITIONS = {
  ["alive" /* Alive */]: ["intent" /* Intent */, "final" /* Final */],
  // Final for emergency
  ["intent" /* Intent */]: ["snapshot" /* Snapshot */, "final" /* Final */],
  // Final for emergency escape from INTENT
  ["snapshot" /* Snapshot */]: ["open" /* Open */, "final" /* Final */],
  // Final for unilateral
  ["open" /* Open */]: ["contested" /* Contested */, "final" /* Final */],
  ["contested" /* Contested */]: ["final" /* Final */],
  ["final" /* Final */]: ["departed" /* Departed */],
  ["departed" /* Departed */]: []
};

// src/index.ts
init_errors();

// src/convenience.ts
init_marker();
init_types();
init_errors();

// src/countersign.ts
init_marker();
init_types();
init_errors();

// src/vc.ts
init_types();

// src/modules/lineage.ts
init_types();
init_marker();

// src/modules/reputation.ts
init_marker();

// src/modules/origin-attestation.ts
init_marker();

// src/modules/assets.ts
var import_sha2562 = require("@noble/hashes/sha256");

// src/modules/trust.ts
init_types();
init_marker();
var import_sha2563 = require("@noble/hashes/sha256");

// src/index.ts
init_anchor();

// src/storage.ts
var import_node_fs = require("fs");
var import_node_path = require("path");
function validateDir(dir) {
  const resolved = (0, import_node_path.resolve)(dir);
  const parts = dir.split(/[/\\]/);
  if (parts.includes("..")) {
    throw new Error(`Unsafe storage directory: path contains '..' traversal: ${dir}`);
  }
  return resolved;
}
function saveMarker(marker, dir) {
  dir = validateDir(dir);
  if (!(0, import_node_fs.existsSync)(dir)) (0, import_node_fs.mkdirSync)(dir, { recursive: true });
  const safeId = marker.id.replace(/[^a-zA-Z0-9_-]/g, "_");
  const filePath = (0, import_node_path.join)(dir, `${safeId}.json`);
  (0, import_node_fs.writeFileSync)(filePath, JSON.stringify(marker, null, 2), "utf-8");
  return filePath;
}
function listMarkers(dir) {
  dir = validateDir(dir);
  if (!(0, import_node_fs.existsSync)(dir)) return [];
  return (0, import_node_fs.readdirSync)(dir).filter((f) => f.endsWith(".json")).map((f) => f.replace(/\.json$/, ""));
}

// src/chain.ts
var import_node_fs2 = require("fs");

// src/privacy.ts
var import_sha2565 = require("@noble/hashes/sha256");
var import_hkdf = require("@noble/hashes/hkdf");
var import_chacha = require("@noble/ciphers/chacha.js");
var import_aes = require("@noble/ciphers/aes.js");
var import_utils = require("@noble/ciphers/utils.js");
var import_ed25519 = require("@noble/curves/ed25519.js");
init_marker();
function toHex2(bytes) {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function hashField(value) {
  const canonical = canonicalize(value);
  const hash = (0, import_sha2565.sha256)(new TextEncoder().encode(canonical));
  return `redacted:sha256:${toHex2(hash)}`;
}
function redactMarker(marker, fields) {
  const result = { ...marker };
  for (const field of fields) {
    if (field in result) {
      result[field] = hashField(result[field]);
    }
  }
  return result;
}

// src/batch.ts
var import_sha2566 = require("@noble/hashes/sha256");
init_anchor();
init_marker();

// src/interop.ts
var import_node_events = require("events");
init_marker();

// src/keri.ts
var import_sha2567 = require("@noble/hashes/sha256");

// src/ethics.ts
init_types();
var RETALIATION_WINDOW_MS = 72 * 60 * 60 * 1e3;

// src/guardrails.ts
init_types();
init_types();

// src/index.ts
init_git_ledger();
init_tsa();

// src/full-service.ts
init_anchor();

// src/key-compromise.ts
init_marker();
init_types();

// src/dispute.ts
init_marker();

// src/index.ts
init_visual();

// src/passage.ts
init_marker();

// src/claim-store.ts
var import_node_crypto3 = require("crypto");

// src/amendment.ts
var import_sha2568 = require("@noble/hashes/sha256");
init_marker();

// src/sqlite-store.ts
var Database;
try {
  Database = require("better-sqlite3");
} catch {
}

// src/cli.ts
var EXIT_VALIDATION_ERROR = 1;
var EXIT_FILE_ERROR = 2;
var EXIT_PARSE_ERROR = 3;
var EXIT_CRYPTO_ERROR = 4;
var EXIT_STORAGE_ERROR = 5;
var EXIT_ARG_ERROR = 6;
function toHex3(bytes) {
  return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
}
function fromHexOrBase64(s) {
  if (/^[0-9a-fA-F]+$/.test(s)) {
    const bytes = new Uint8Array(s.length / 2);
    for (let i = 0; i < s.length; i += 2) bytes[i / 2] = parseInt(s.slice(i, i + 2), 16);
    return bytes;
  }
  try {
    return Uint8Array.from(atob(s), (c) => c.charCodeAt(0));
  } catch {
    throw new Error(`Key data is neither valid hex nor base64. Got: "${s.slice(0, 20)}..."`);
  }
}
function parseExitType(s) {
  const map = {
    voluntary: "voluntary" /* Voluntary */,
    forced: "forced" /* Forced */,
    emergency: "emergency" /* Emergency */
  };
  if (!map[s]) {
    throw new import_commander.InvalidArgumentError(
      `Invalid exit type "${s}". Valid types: voluntary, forced, emergency`
    );
  }
  return map[s];
}
function parseExitStatus(s) {
  const map = {
    good_standing: "good_standing" /* GoodStanding */,
    disputed: "disputed" /* Disputed */,
    unverified: "unverified" /* Unverified */
  };
  if (!map[s]) {
    throw new import_commander.InvalidArgumentError(
      `Invalid status "${s}". Valid statuses: good_standing, disputed, unverified`
    );
  }
  return map[s];
}
function readMarkerFile(file, context) {
  let raw;
  try {
    if (!(0, import_node_fs3.existsSync)(file)) {
      error(
        `File not found: ${file}`,
        `Attempted to ${context}, but the file does not exist.`,
        [
          `Check that the path is correct`,
          `Use an absolute path or a path relative to the current directory`,
          `Run 'exit list' to see stored markers`
        ],
        EXIT_FILE_ERROR
      );
    }
    raw = (0, import_node_fs3.readFileSync)(file, "utf-8");
  } catch (e) {
    if (e.exitCode) throw e;
    error(
      `Cannot read file: ${file}`,
      `Attempted to ${context}, but the file could not be read.`,
      [
        `Check file permissions (current user needs read access)`,
        `Ensure the file is not locked by another process`,
        e.message ? `System error: ${e.message}` : null
      ].filter(Boolean),
      EXIT_FILE_ERROR
    );
  }
  try {
    return JSON.parse(raw);
  } catch (e) {
    error(
      `Invalid JSON in file: ${file}`,
      `Attempted to ${context}, but the file does not contain valid JSON.`,
      [
        `Ensure the file contains a valid EXIT marker in JSON format`,
        `Check for trailing commas, unquoted keys, or truncated content`,
        `You can create a new marker with: exit create --origin <uri>`,
        e.message ? `Parse error: ${e.message}` : null
      ].filter(Boolean),
      EXIT_PARSE_ERROR
    );
  }
  return null;
}
function readKeyFile(keyPath) {
  let raw;
  try {
    if (!(0, import_node_fs3.existsSync)(keyPath)) {
      error(
        `Key file not found: ${keyPath}`,
        `Attempted to read a private key, but the file does not exist.`,
        [
          `Check the path to your key file`,
          `Generate a new keypair with: exit keygen`,
          `Save the privateKey field to a file and reference it with --key`
        ],
        EXIT_FILE_ERROR
      );
    }
    raw = (0, import_node_fs3.readFileSync)(keyPath, "utf-8").trim();
  } catch (e) {
    if (e.exitCode) throw e;
    error(
      `Cannot read key file: ${keyPath}`,
      `The key file exists but could not be read.`,
      [
        `Check file permissions`,
        e.message ? `System error: ${e.message}` : null
      ].filter(Boolean),
      EXIT_FILE_ERROR
    );
    return null;
  }
  try {
    return fromHexOrBase64(raw);
  } catch (e) {
    error(
      `Invalid key format in: ${keyPath}`,
      `The key file was read but its contents are not valid hex or base64.`,
      [
        `Key files should contain a single line of hex (64 chars for Ed25519) or base64`,
        `Generate a valid keypair with: exit keygen`,
        e.message ? `Detail: ${e.message}` : null
      ].filter(Boolean),
      EXIT_CRYPTO_ERROR
    );
    return null;
  }
}
function error(title, context, suggestions, exitCode) {
  console.error();
  console.error(`  \u2717 Error: ${title}`);
  console.error();
  console.error(`  ${context}`);
  console.error();
  if (suggestions.length > 0) {
    console.error(`  Suggestions:`);
    for (const s of suggestions) {
      console.error(`    \u2192 ${s}`);
    }
    console.error();
  }
  const err = new Error(title);
  err.exitCode = exitCode;
  throw err;
}
function withErrorHandling(fn) {
  return (...args) => {
    try {
      fn(...args);
    } catch (e) {
      if (e.exitCode !== void 0) {
        process.exit(e.exitCode);
      }
      console.error();
      console.error(`  \u2717 Unexpected error: ${e.message || e}`);
      console.error();
      console.error(`  This may be a bug in cellar-door-exit.`);
      console.error(`  Please report it with the full error output.`);
      if (e.stack) {
        console.error();
        console.error(`  Stack trace:`);
        console.error(`  ${e.stack}`);
      }
      console.error();
      process.exit(1);
    }
  };
}
var program = new import_commander.Command();
program.name("exit").description(
  "cellar-door-exit \u2014 Verifiable EXIT markers for agents.\n\nCreate, sign, verify, inspect, anchor, store, and manage EXIT markers\nthat provide authenticated declarations of departure with continuity\npreservation across contexts."
).version("0.1.0").addHelpText("after", `
Examples:
  $ exit keygen                                    Generate a new Ed25519 keypair
  $ exit create --origin example.com --sign        Create and sign a marker
  $ exit verify marker.json                        Verify a marker file
  $ exit inspect marker.json                       Human-readable marker display
  $ exit store marker.json                         Save marker to local storage
  $ exit list                                      List stored markers
  $ exit anchor marker.json                        Compute anchor hash
  $ exit redact marker.json --fields subject       Redact fields from a marker

Documentation: https://cellar-door.dev/exit/v1
`);
program.command("create").description("Create a new EXIT marker.").requiredOption("--origin <uri>", "Origin URI \u2014 the system/context being exited").option("--subject <did>", "Subject DID (who is exiting). Generates a keypair if omitted").option("--type <type>", "Exit type: voluntary, forced, emergency (default: voluntary)", "voluntary").option("--status <status>", "Standing: good_standing, disputed, unverified (default: good_standing)", "good_standing").option("--reason <text>", "Human-readable reason for departure").option("--sign", "Sign the marker (uses --key if provided, otherwise generates a keypair)").option("--key <path>", "Path to private key file (hex or base64 encoded)").option("--show-keys", "Print generated private key material (DANGER: do not use in production logs)").addHelpText("after", `
Examples:
  $ exit create --origin example.com
    Create an unsigned marker with a generated DID

  $ exit create --origin example.com --sign
    Create and sign a marker with a generated keypair

  $ exit create --origin example.com --sign --key ./my.key --subject did:key:z6Mk...
    Create and sign with an existing key

  $ exit create --origin example.com --type emergency --reason "System shutting down"
    Create an emergency exit marker with a reason

  $ exit create --origin example.com --status disputed --reason "Banned unfairly"
    Create a marker with disputed standing
`).action(withErrorHandling(async (opts) => {
  let privateKey;
  let publicKey;
  let subject = opts.subject;
  if (opts.key) {
    privateKey = readKeyFile(opts.key);
    try {
      const ed2 = await import("@noble/ed25519");
      publicKey = ed2.getPublicKey(privateKey);
    } catch (e) {
      error(
        `Failed to derive public key from private key`,
        `The private key in "${opts.key}" could not be used to derive a public key.`,
        [
          `Ensure the key file contains a valid 32-byte Ed25519 private key`,
          `Generate a new keypair with: exit keygen`,
          e.message ? `Crypto error: ${e.message}` : null
        ].filter(Boolean),
        EXIT_CRYPTO_ERROR
      );
    }
    if (!subject) subject = didFromPublicKey(publicKey);
  } else if (opts.sign || !subject) {
    const kp = generateKeyPair();
    privateKey = kp.privateKey;
    publicKey = kp.publicKey;
    if (!subject) subject = didFromPublicKey(publicKey);
    const keypairInfo = {
      did: didFromPublicKey(publicKey),
      publicKey: toHex3(publicKey)
    };
    if (opts.showKeys) {
      keypairInfo.privateKey = toHex3(privateKey);
    } else {
      keypairInfo.privateKey = "[REDACTED \u2014 use --show-keys to display]";
    }
    process.stderr.write(
      JSON.stringify({ _generated_keypair: keypairInfo }) + "\n"
    );
  }
  const exitType = parseExitType(opts.type);
  const status = parseExitStatus(opts.status);
  let marker;
  try {
    marker = createMarker({ subject, origin: opts.origin, exitType, status });
  } catch (e) {
    error(
      `Failed to create EXIT marker`,
      `The marker could not be created with the provided parameters.`,
      [
        `Check that --origin is a valid URI`,
        `Check that --subject (if provided) is a valid DID`,
        e.message ? `Detail: ${e.message}` : null
      ].filter(Boolean),
      EXIT_VALIDATION_ERROR
    );
    return;
  }
  if (opts.reason) {
    const meta = { reason: opts.reason };
    marker = addModule(marker, "metadata", meta);
  }
  if (opts.sign && privateKey && publicKey) {
    try {
      marker = signMarker(marker, privateKey, publicKey);
    } catch (e) {
      error(
        `Failed to sign EXIT marker`,
        `The marker was created but signing failed.`,
        [
          `Ensure the private key is a valid 32-byte Ed25519 key`,
          `If using --key, verify the file contains the correct key`,
          e.message ? `Signing error: ${e.message}` : null
        ].filter(Boolean),
        EXIT_CRYPTO_ERROR
      );
    }
  }
  process.stdout.write(JSON.stringify(marker, null, 2) + "\n");
}));
program.command("verify <file>").description("Verify an EXIT marker from a JSON file.").addHelpText("after", `
Examples:
  $ exit verify marker.json
    Verify a marker and display the result

  $ exit verify ./exit-markers/abc123.json
    Verify a specific stored marker

Exit codes:
  0  Marker is valid
  1  Marker is invalid (verification failed)
  2  File not found or unreadable
  3  File is not valid JSON
`).action(withErrorHandling((file) => {
  const marker = readMarkerFile(file, "verify the EXIT marker");
  let result;
  try {
    result = verifyMarker(marker);
  } catch (e) {
    error(
      `Verification error`,
      `An error occurred while verifying the marker in "${file}".`,
      [
        `The file may contain a malformed or incomplete marker`,
        `Ensure all required fields are present (id, subject, origin, timestamp, exitType, status, proof)`,
        e.message ? `Detail: ${e.message}` : null
      ].filter(Boolean),
      EXIT_VALIDATION_ERROR
    );
    return;
  }
  if (result.valid) {
    console.log("\u2713 VALID");
    console.log(`  Subject: ${marker.subject}`);
    console.log(`  Origin:  ${marker.origin}`);
    console.log(`  Type:    ${marker.exitType}`);
    console.log(`  Status:  ${marker.status}`);
  } else {
    console.log("\u2717 INVALID");
    console.log();
    for (const err of result.errors) {
      console.log(`  \u2717 ${err}`);
    }
    console.log();
    console.log("  Suggestions:");
    console.log("    \u2192 Ensure the marker was signed with the correct key");
    console.log("    \u2192 Check that no fields were modified after signing");
    console.log("    \u2192 Re-create and re-sign the marker if needed: exit create --origin <uri> --sign");
    process.exit(EXIT_VALIDATION_ERROR);
  }
}));
program.command("inspect <file>").description("Human-readable display of an EXIT marker.").addHelpText("after", `
Examples:
  $ exit inspect marker.json
    Display a formatted view of the marker with all modules

  $ exit inspect ./exit-markers/abc123.json
    Inspect a specific stored marker
`).action(withErrorHandling((file) => {
  const marker = readMarkerFile(file, "inspect the EXIT marker");
  console.log("\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551                    EXIT MARKER                          \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D");
  console.log();
  console.log(`  ID:        ${marker.id}`);
  console.log(`  Subject:   ${marker.subject}`);
  console.log(`  Origin:    ${marker.origin}`);
  console.log(`  Timestamp: ${marker.timestamp}`);
  console.log(`  Type:      ${marker.exitType}`);
  console.log(`  Status:    ${marker.status}`);
  console.log();
  console.log("  \u2500\u2500 Proof \u2500\u2500");
  console.log(`  Algorithm:    ${marker.proof.type}`);
  console.log(`  Created:      ${marker.proof.created}`);
  console.log(`  Verification: ${marker.proof.verificationMethod}`);
  console.log(`  Signature:    ${marker.proof.proofValue ? marker.proof.proofValue.slice(0, 32) + "..." : "(unsigned)"}`);
  console.log();
  if (marker.lineage) {
    console.log("  \u2500\u2500 Module A: Lineage \u2500\u2500");
    if (marker.lineage.predecessor) console.log(`  Predecessor: ${marker.lineage.predecessor}`);
    if (marker.lineage.successor) console.log(`  Successor:   ${marker.lineage.successor}`);
    if (marker.lineage.lineageChain) console.log(`  Chain:       ${marker.lineage.lineageChain.join(" \u2192 ")}`);
    if (marker.lineage.continuityProof) console.log(`  Proof type:  ${marker.lineage.continuityProof.type}`);
    console.log();
  }
  if (marker.stateSnapshot) {
    console.log("  \u2500\u2500 Module B: State Snapshot \u2500\u2500");
    console.log(`  Hash:     ${marker.stateSnapshot.stateHash}`);
    if (marker.stateSnapshot.stateLocation) console.log(`  Location: ${marker.stateSnapshot.stateLocation}`);
    console.log();
  }
  if (marker.dispute) {
    console.log("  \u2500\u2500 Module C: Dispute \u2500\u2500");
    if (marker.dispute.disputes) console.log(`  Active disputes: ${marker.dispute.disputes.length}`);
    if (marker.dispute.challengeWindow) {
      console.log(`  Challenge: ${marker.dispute.challengeWindow.opens} \u2192 ${marker.dispute.challengeWindow.closes}`);
    }
    console.log();
  }
  if (marker.economic) {
    console.log("  \u2500\u2500 Module D: Economic \u2500\u2500");
    if (marker.economic.assetManifest) console.log(`  Assets: ${marker.economic.assetManifest.length}`);
    if (marker.economic.exitFee) console.log(`  Exit fee: ${marker.economic.exitFee.amount}`);
    console.log();
  }
  if (marker.metadata) {
    console.log("  \u2500\u2500 Module E: Metadata \u2500\u2500");
    if (marker.metadata.reason) console.log(`  Reason:    ${marker.metadata.reason}`);
    if (marker.metadata.narrative) console.log(`  Narrative: ${marker.metadata.narrative}`);
    if (marker.metadata.tags) console.log(`  Tags:      ${marker.metadata.tags.join(", ")}`);
    console.log();
  }
  if (marker.crossDomain) {
    console.log("  \u2500\u2500 Module F: Cross-Domain \u2500\u2500");
    if (marker.crossDomain.anchors) {
      for (const a of marker.crossDomain.anchors) {
        console.log(`  ${a.chain}: ${a.txHash}`);
      }
    }
    console.log();
  }
  let result;
  try {
    result = verifyMarker(marker);
  } catch {
    console.log("  Verification: \u26A0 Could not verify (marker may be malformed)");
    return;
  }
  console.log(`  Verification: ${result.valid ? "\u2713 VALID" : "\u2717 INVALID"}`);
  if (!result.valid) {
    for (const err of result.errors) console.log(`    - ${err}`);
  }
}));
program.command("keygen").description("Generate a new Ed25519 keypair for signing EXIT markers.").addHelpText("after", `
Examples:
  $ exit keygen
    Print a new keypair as JSON (did, publicKey, privateKey)

  $ exit keygen > keypair.json
    Save the keypair to a file

  $ exit keygen | jq -r .privateKey > my.key
    Extract just the private key for use with --key

Security:
  The private key is printed to stdout. Store it securely and never share it.
  Anyone with the private key can sign EXIT markers as your identity.
`).option("--show-keys", "Print private key to stdout (DANGER: do not use in production logs)").action(withErrorHandling((opts) => {
  const kp = generateKeyPair();
  const did = didFromPublicKey(kp.publicKey);
  if (opts.showKeys) {
    console.log(
      JSON.stringify(
        {
          did,
          publicKey: toHex3(kp.publicKey),
          privateKey: toHex3(kp.privateKey)
        },
        null,
        2
      )
    );
  } else {
    const keyFileName = `${did.replace(/[^a-zA-Z0-9]/g, "_")}.key`;
    (0, import_node_fs3.writeFileSync)(keyFileName, toHex3(kp.privateKey), { mode: 384 });
    console.log(
      JSON.stringify(
        {
          did,
          publicKey: toHex3(kp.publicKey),
          privateKeyFile: keyFileName,
          privateKey: "[REDACTED \u2014 saved to file. Use --show-keys to print to stdout instead]"
        },
        null,
        2
      )
    );
    process.stderr.write(`\u26A0 Private key written to ${keyFileName} (mode 0600)
`);
  }
}));
program.command("anchor <file>").description("Compute and display the anchor hash for a marker (for cross-domain anchoring).").addHelpText("after", `
Examples:
  $ exit anchor marker.json
    Compute the anchor record (hash + metadata) for on-chain anchoring

  $ exit anchor marker.json > anchor.json
    Save the anchor record for submission to a blockchain or registry
`).action(withErrorHandling((file) => {
  const marker = readMarkerFile(file, "compute the anchor hash");
  try {
    const record = createAnchorRecord(marker);
    console.log(JSON.stringify(record, null, 2));
  } catch (e) {
    error(
      `Failed to compute anchor hash`,
      `The anchor hash could not be computed for the marker in "${file}".`,
      [
        `Ensure the marker is complete and well-formed`,
        `Try verifying the marker first: exit verify ${file}`,
        e.message ? `Detail: ${e.message}` : null
      ].filter(Boolean),
      EXIT_VALIDATION_ERROR
    );
  }
}));
program.command("store <file>").description("Save a marker to local file-based storage.").option("--dir <path>", "Storage directory (default: ./exit-markers)", "./exit-markers").addHelpText("after", `
Examples:
  $ exit store marker.json
    Save to ./exit-markers/

  $ exit store marker.json --dir /var/lib/exit-markers
    Save to a custom directory

  $ exit list --dir /var/lib/exit-markers
    List markers in the custom directory
`).action(withErrorHandling((file, opts) => {
  const marker = readMarkerFile(file, "store the EXIT marker");
  try {
    const path = saveMarker(marker, opts.dir);
    console.log(`Saved to ${path}`);
  } catch (e) {
    error(
      `Failed to save marker`,
      `The marker from "${file}" could not be saved to "${opts.dir}".`,
      [
        `Check that you have write permissions to the directory`,
        `Ensure the directory path is valid`,
        `Try creating the directory manually: mkdir -p ${opts.dir}`,
        e.message ? `Storage error: ${e.message}` : null
      ].filter(Boolean),
      EXIT_STORAGE_ERROR
    );
  }
}));
program.command("list").description("List stored EXIT markers.").option("--dir <path>", "Storage directory (default: ./exit-markers)", "./exit-markers").addHelpText("after", `
Examples:
  $ exit list
    List all markers in ./exit-markers/

  $ exit list --dir /var/lib/exit-markers
    List markers in a custom directory
`).action(withErrorHandling((opts) => {
  let ids;
  try {
    ids = listMarkers(opts.dir);
  } catch (e) {
    if (!(0, import_node_fs3.existsSync)(opts.dir)) {
      console.log(`No markers found. Storage directory does not exist: ${opts.dir}`);
      console.log();
      console.log(`  To create your first marker:`);
      console.log(`    $ exit create --origin example.com --sign > marker.json`);
      console.log(`    $ exit store marker.json --dir ${opts.dir}`);
      return;
    }
    error(
      `Failed to list markers`,
      `Could not read the storage directory "${opts.dir}".`,
      [
        `Check that you have read permissions on the directory`,
        e.message ? `System error: ${e.message}` : null
      ].filter(Boolean),
      EXIT_STORAGE_ERROR
    );
    return;
  }
  if (ids.length === 0) {
    console.log("No markers found.");
    console.log();
    console.log("  To create and store a marker:");
    console.log("    $ exit create --origin example.com --sign > marker.json");
    console.log(`    $ exit store marker.json --dir ${opts.dir}`);
  } else {
    for (const id of ids) console.log(id);
  }
}));
program.command("redact <file>").description("Create a privacy-preserving redacted version of a marker.").requiredOption("--fields <fields>", "Comma-separated list of fields to redact (e.g., subject,origin)").addHelpText("after", `
Examples:
  $ exit redact marker.json --fields subject
    Redact the subject field (replace with hash)

  $ exit redact marker.json --fields subject,origin,metadata
    Redact multiple fields

  $ exit redact marker.json --fields subject > redacted.json
    Save the redacted marker to a new file

Redactable fields:
  subject, origin, metadata, lineage, stateSnapshot, dispute, economic, crossDomain
`).action(withErrorHandling((file, opts) => {
  const marker = readMarkerFile(file, "redact the EXIT marker");
  const fields = opts.fields.split(",").map((f) => f.trim()).filter(Boolean);
  if (fields.length === 0) {
    error(
      `No fields specified for redaction`,
      `The --fields option was provided but no valid field names were found.`,
      [
        `Provide a comma-separated list of fields: --fields subject,origin`,
        `Redactable fields: subject, origin, metadata, lineage, stateSnapshot, dispute, economic, crossDomain`
      ],
      EXIT_ARG_ERROR
    );
  }
  try {
    const redacted = redactMarker(marker, fields);
    process.stdout.write(JSON.stringify(redacted, null, 2) + "\n");
  } catch (e) {
    error(
      `Redaction failed`,
      `Could not redact the specified fields from the marker in "${file}".`,
      [
        `Check that the field names are valid: ${fields.join(", ")}`,
        `Redactable fields: subject, origin, metadata, lineage, stateSnapshot, dispute, economic, crossDomain`,
        e.message ? `Detail: ${e.message}` : null
      ].filter(Boolean),
      EXIT_VALIDATION_ERROR
    );
  }
}));
program.parse();
