// cellar-door-exit/dist/index.js
import {
  ANTI_WEAPONIZATION_CLAUSE,
  AlgorithmRegistry,
  CeremonyStateMachine,
  ClaimType,
  EXIT_CONTEXT,
  Ed25519Signer,
  ExitEventEmitter,
  FileChainAdapter,
  InMemoryAmendmentStore,
  InMemoryKeyStore,
  KeyEventLog,
  MarkerRegistry,
  MemoryClaimStore,
  MockChainAdapter,
  P256Signer,
  SqliteClaimStore,
  VC_CONTEXT,
  addCoercionLabel,
  addCounterSignature,
  addRightOfReply,
  addWitness,
  algorithmFromDid,
  algorithmFromProofType,
  applyAmendment,
  applyAmendments,
  applySunset,
  bindSuccessor,
  claimFromMarker,
  claimsFromTrustEnhancers,
  commitNextKey,
  computeConfidence,
  computeMerkleProof,
  computeMerkleRoot,
  computeStatusConfirmation,
  createAmendment,
  createAmendmentQueryHandler,
  createAssetManifest,
  createBatchExit,
  createCommitment,
  createCompromiseMarker,
  createContinuityModule,
  createDepartureMarker,
  createDidDocument,
  createDidKeri,
  createDispute,
  createDisputeModule,
  createExitHook,
  createExitMiddleware,
  createInception,
  createLineageModule,
  createMinimalDisclosure,
  createOriginAttestation,
  createPassage,
  createReputationModule,
  createRevocation,
  createRotation,
  createShutdownBatch,
  createSigner,
  createTenureAttestation,
  createVerifier,
  createWellKnownMetadata,
  decryptManagedMarker,
  decryptMarker,
  deleteMarker,
  deleteMarkerKey,
  departAndAnchor,
  departAndVerify,
  deriveStatusConfirmation,
  deserializeFromTransport,
  detectCoercion,
  detectReputationLaundering,
  detectRetaliationWindow,
  detectWeaponization,
  didFromP256PublicKey,
  didFromPublicKey,
  didMethod,
  digestKey,
  discoverAmendments,
  encryptMarker,
  encryptMarkerWithManagedKey,
  exportMarker,
  extractConfidenceFactors,
  flagCompromisedPlatformMarkers,
  fromJSON,
  generateEthicsReport,
  generateIdentity,
  generateKeyPair,
  generateKeyPairForAlgorithm,
  generateP256KeyPair,
  generatePassageIdentity,
  generatePreRotatedKeys,
  getAlgorithm,
  getContext,
  getDisputeStatus,
  getValidTransitions,
  importMarker,
  ingestMarker,
  initTelemetry,
  instrumentedSignDepartureMarker,
  instrumentedSignMarker,
  instrumentedVerifyDeparture,
  instrumentedVerifyMarker,
  isDid,
  isDisputed,
  isExpired,
  isKeyCompromised,
  isRevealWindowOpen,
  isVC,
  linkCompromisedMarkers,
  listMarkers,
  loadAmendments,
  loadMarker,
  loadRevocations,
  parseDurationToDays,
  proofTypeForAlgorithm,
  publicKeyFromDid,
  publicKeyFromP256Did,
  quickCounterSign,
  quickDeparture,
  quickExit,
  quickExitP256,
  quickPassageVerify,
  quickVerify,
  redactMarker,
  resetTelemetry,
  resolveDid,
  resolveDidKeri,
  resolveDispute,
  resolveMarker,
  rotateKeys,
  saveAmendment,
  saveMarker,
  saveRevocation,
  serializeForTransport,
  sign,
  signAttestation,
  signDepartureMarker,
  signDepartureWithSigner,
  signEndorsement,
  signMarker,
  signMarkerWithSigner,
  signP256,
  startCeremonySpan,
  startExitSpan,
  toJSON,
  unwrapVC,
  validateEthicalCompliance,
  validateMarker,
  verify,
  verifyAmendmentSignature,
  verifyBatchMembership,
  verifyCommitment,
  verifyCompromiseRecovery,
  verifyCounterSignature,
  verifyDeparture,
  verifyDepartureMultiAlg,
  verifyDisputeResolution,
  verifyKeyState,
  verifyLineageChain,
  verifyMarker,
  verifyMarkerMultiAlg,
  verifyNextKeyCommitment,
  verifyP256,
  verifyPassage,
  verifyRevocationSignature,
  verifyTenureAttestation,
  verifyTrustEnhancers,
  withSpan,
  withSpanAsync,
  wrapAsVC
} from "./chunk-H5F44PDI.js";
import {
  anchorToGit,
  initLedger,
  listLedgerEntries,
  verifyLedgerEntry
} from "./chunk-CFRFGOQV.js";
import {
  anchorWithTSA,
  buildTimestampRequest,
  checkTSAReceiptStructure,
  checkTSRStatus,
  extractTimestampFromTSR,
  requestTimestamp
} from "./chunk-KKNMKJSS.js";
import {
  CeremonyError,
  CeremonyRole,
  CeremonyState,
  CoercionLabel,
  ContinuityProofType,
  EXIT_CONTEXT_V1,
  EXIT_SPEC_VERSION,
  ExitError,
  ExitStatus,
  ExitType,
  SigningError,
  StatusConfirmation,
  StorageError,
  SuccessorTrustLevel,
  ValidationError,
  VerificationError,
  addModule,
  canonicalize,
  computeAnchorHash,
  computeId,
  createAnchorRecord,
  createMarker,
  createMinimalAnchor,
  verifyAnchorRecord
} from "./chunk-PXQJCDFL.js";
import {
  hashToColors,
  renderDoorASCII,
  renderDoorSVG,
  shortHash
} from "./chunk-LY5BBBIO.js";
import "./chunk-3RG5ZIWI.js";
export {
  ANTI_WEAPONIZATION_CLAUSE,
  AlgorithmRegistry,
  CoercionLabel as COERCION_LABELS,
  CeremonyError,
  CeremonyRole,
  CeremonyState,
  CeremonyStateMachine,
  ClaimType,
  CoercionLabel,
  ContinuityProofType,
  EXIT_CONTEXT,
  EXIT_CONTEXT_V1,
  EXIT_SPEC_VERSION,
  Ed25519Signer,
  ExitError,
  ExitEventEmitter,
  ExitStatus,
  ExitType,
  FileChainAdapter,
  InMemoryAmendmentStore,
  InMemoryKeyStore,
  KeyEventLog,
  MarkerRegistry,
  MemoryClaimStore,
  MockChainAdapter,
  P256Signer,
  SigningError,
  SqliteClaimStore,
  StatusConfirmation,
  StorageError,
  SuccessorTrustLevel,
  VC_CONTEXT,
  ValidationError,
  VerificationError,
  addCoercionLabel,
  addCounterSignature,
  addModule,
  addRightOfReply,
  addWitness,
  algorithmFromDid,
  algorithmFromProofType,
  anchorToGit,
  anchorWithTSA,
  applyAmendment,
  applyAmendments,
  applySunset,
  bindSuccessor,
  buildTimestampRequest,
  canonicalize,
  checkTSAReceiptStructure,
  checkTSRStatus,
  claimFromMarker,
  claimsFromTrustEnhancers,
  commitNextKey,
  computeAnchorHash,
  computeConfidence,
  computeId,
  computeMerkleProof,
  computeMerkleRoot,
  computeStatusConfirmation,
  createAmendment,
  createAmendmentQueryHandler,
  createAnchorRecord,
  createAssetManifest,
  createBatchExit,
  createCommitment,
  createCompromiseMarker,
  createContinuityModule,
  createDepartureMarker,
  createDidDocument,
  createDidKeri,
  createDispute,
  createDisputeModule,
  createExitHook,
  createExitMiddleware,
  createInception,
  createLineageModule,
  createMarker,
  createMinimalAnchor,
  createMinimalDisclosure,
  createOriginAttestation,
  createPassage,
  createReputationModule,
  createRevocation,
  createRotation,
  createShutdownBatch,
  createSigner,
  createTenureAttestation,
  createVerifier,
  createWellKnownMetadata,
  decryptManagedMarker,
  decryptMarker,
  deleteMarker,
  deleteMarkerKey,
  departAndAnchor,
  departAndVerify,
  deriveStatusConfirmation,
  deserializeFromTransport,
  detectCoercion,
  detectReputationLaundering,
  detectRetaliationWindow,
  detectWeaponization,
  didFromP256PublicKey,
  didFromPublicKey,
  didMethod,
  digestKey,
  discoverAmendments,
  encryptMarker,
  encryptMarkerWithManagedKey,
  exportMarker,
  extractConfidenceFactors,
  extractTimestampFromTSR,
  flagCompromisedPlatformMarkers,
  fromJSON,
  generateEthicsReport,
  generateIdentity,
  generateKeyPair,
  generateKeyPairForAlgorithm,
  generateP256KeyPair,
  generatePassageIdentity,
  generatePreRotatedKeys,
  getAlgorithm,
  getContext,
  getDisputeStatus,
  getValidTransitions,
  hashToColors,
  importMarker,
  ingestMarker,
  initLedger,
  initTelemetry,
  instrumentedSignDepartureMarker,
  instrumentedSignMarker,
  instrumentedVerifyDeparture,
  instrumentedVerifyMarker,
  isDid,
  isDisputed,
  isExpired,
  isKeyCompromised,
  isRevealWindowOpen,
  isVC,
  linkCompromisedMarkers,
  listLedgerEntries,
  listMarkers,
  loadAmendments,
  loadMarker,
  loadRevocations,
  parseDurationToDays,
  proofTypeForAlgorithm,
  publicKeyFromDid,
  publicKeyFromP256Did,
  quickCounterSign,
  quickDeparture,
  quickExit,
  quickExitP256,
  quickPassageVerify,
  quickVerify,
  redactMarker,
  renderDoorASCII,
  renderDoorSVG,
  requestTimestamp,
  resetTelemetry,
  resolveDid,
  resolveDidKeri,
  resolveDispute,
  resolveMarker,
  rotateKeys,
  saveAmendment,
  saveMarker,
  saveRevocation,
  serializeForTransport,
  shortHash,
  sign,
  signAttestation,
  signDepartureMarker,
  signDepartureWithSigner,
  signEndorsement,
  signMarker,
  signMarkerWithSigner,
  signP256,
  startCeremonySpan,
  startExitSpan,
  toJSON,
  unwrapVC,
  validateEthicalCompliance,
  validateMarker,
  verify,
  verifyAmendmentSignature,
  verifyAnchorRecord,
  verifyBatchMembership,
  verifyCommitment,
  verifyCompromiseRecovery,
  verifyCounterSignature,
  verifyDeparture,
  verifyDepartureMultiAlg,
  verifyDisputeResolution,
  verifyKeyState,
  verifyLedgerEntry,
  verifyLineageChain,
  verifyMarker,
  verifyMarkerMultiAlg,
  verifyNextKeyCommitment,
  verifyP256,
  verifyPassage,
  verifyRevocationSignature,
  checkTSAReceiptStructure as verifyTSAReceipt,
  verifyTenureAttestation,
  verifyTrustEnhancers,
  withSpan,
  withSpanAsync,
  wrapAsVC
};
